{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experiments for Inference and Analysis NCBI Yellow Fever"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TBC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Imports and setup environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "`ecutilities` already installed\n",
      "`metagentools` already installed\n"
     ]
    }
   ],
   "source": [
    "# Install required custom packages if not installed yet.\n",
    "import importlib.util\n",
    "if not importlib.util.find_spec('ecutilities'):\n",
    "    print('installing package: `ecutilities`')\n",
    "    ! pip install -qqU ecutilities\n",
    "else:\n",
    "    print('`ecutilities` already installed')\n",
    "if not importlib.util.find_spec('metagentools'):\n",
    "    print('installing package: `metagentools')\n",
    "    ! pip install -qqU metagentools\n",
    "else:\n",
    "    print('`metagentools` already installed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set autoreload mode\n",
      "Tensorflow version: 2.8.2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Import all required packages\n",
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "import sqlite3\n",
    "import time\n",
    "\n",
    "from datetime import datetime\n",
    "from ecutilities.core import files_in_tree\n",
    "from ecutilities.ipython import nb_setup\n",
    "from functools import partial\n",
    "from IPython.display import display, update_display, Markdown, HTML\n",
    "from pandas import HDFStore\n",
    "from pathlib import Path\n",
    "from pprint import pprint\n",
    "from tqdm.notebook import tqdm, trange\n",
    "\n",
    "# Setup the notebook for development\n",
    "nb_setup()\n",
    "\n",
    "os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' # or any {'0', '1', '2'}\n",
    "import tensorflow as tf\n",
    "from tensorflow.python.client import device_lib\n",
    "from tensorflow.keras.models import load_model\n",
    "print(f\"Tensorflow version: {tf.__version__}\\n\")\n",
    "\n",
    "from metagentools.cnn_virus.data import _base_hot_encode, strings_to_tensors\n",
    "from metagentools.cnn_virus.data import split_kmer_into_50mers, combine_prediction_batch\n",
    "from metagentools.cnn_virus.data import FastaFileReader, FastqFileReader, AlnFileReader\n",
    "from metagentools.cnn_virus.data import OriginalLabels\n",
    "from metagentools.cnn_virus.data import string_input_batch_to_tensors, split_kmer_batch_into_50mers\n",
    "from metagentools.cnn_virus.architecture import create_model_original\n",
    "from metagentools.core import ProjectFileSystem, TextFileBaseReader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "List all computing devices available on the machine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Devices:\n",
      "  - CPU  /device:CPU:0                          \n",
      "  - GPU  /device:GPU:0  NVIDIA GeForce GTX 1050 \n"
     ]
    }
   ],
   "source": [
    "devices = device_lib.list_local_devices()\n",
    "print('\\nDevices:')\n",
    "for d in devices:\n",
    "    t = d.device_type\n",
    "    name = d.physical_device_desc\n",
    "    l = [item.split(':', 1) for item in name.split(', ')]\n",
    "    name_attr = dict([x for x in l if len(x)==2])\n",
    "    dev = name_attr.get('name', ' ')\n",
    "    print(f\"  - {t}  {d.name} {dev:25s}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Setup paths to files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Key folders and system information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running linux on local computer\n",
      "Device's home directory: /home/vtec\n",
      "Project file structure:\n",
      " - Root ........ /home/vtec/projects/bio/metagentools \n",
      " - Data Dir .... /home/vtec/projects/bio/metagentools/data \n",
      " - Notebooks ... /home/vtec/projects/bio/metagentools/nbs\n"
     ]
    }
   ],
   "source": [
    "pfs = ProjectFileSystem()\n",
    "pfs.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- `p2model`: path to file with saved original pretrained model\n",
    "- `p2virus_labels` path to file with virus names and labels mapping for original model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "p2model = pfs.data / 'saved/cnn_virus_original/pretrained_model.h5'\n",
    "assert p2model.is_file(), f\"No file found at {p2model.absolute()}\"\n",
    "\n",
    "p2virus_labels = pfs.data / 'CNN_Virus_data/virus_name_mapping'\n",
    "assert p2virus_labels.is_file(), f\"No file found at {p2virus_labels.absolute()}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Load simulated reads and review"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check which simread files are already created:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "simreads\n",
      "  |--yf\n",
      "  |    |--readme.md (0)\n",
      "  |    |--single_1seq_150bp\n",
      "  |    |    |--single_1seq_150bp.fq (1)\n",
      "  |    |    |--single_1seq_150bp.aln (2)\n",
      "  |    |--single_69seq_150bp\n",
      "  |    |    |--single_69seq_150bp.fq (3)\n",
      "  |    |    |--single_69seq_150bp.aln (4)\n",
      "  |    |--paired_1seq_150bp\n",
      "  |    |    |--paired_1seq_150bp2.aln (5)\n",
      "  |    |    |--paired_1seq_150bp2.fq (6)\n",
      "  |    |    |--paired_1seq_150bp1.fq (7)\n",
      "  |    |    |--paired_1seq_150bp1.aln (8)\n",
      "  |    |--paired_69seq_150bp\n",
      "  |    |    |--paired_69seq_150bp1.fq (9)\n",
      "  |    |    |--paired_69seq_150bp2.fq (10)\n",
      "  |    |    |--paired_69seq_150bp1.aln (11)\n",
      "  |    |    |--paired_69seq_150bp2.aln (12)\n"
     ]
    }
   ],
   "source": [
    "files_in_tree(pfs.data / 'ncbi/simreads/yf');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "p2aln = pfs.data / 'ncbi/simreads/yf/single_1seq_150bp/single_1seq_150bp.aln'\n",
    "assert p2aln.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17,000 simulated reads in file single_1seq_150bp.aln from 1 refseq.\n",
      "ART command:  /usr/bin/art_illumina -i /home/vtec/projects/bio/metagentools/data/ncbi/refsequences/yf/yf_1971_angola.fa -ss HS25 -l 150 -f 250 -o /home/vtec/projects/bio/metagentools/data/ncbi/simreads/yf/single_1seq_150bp/single_1seq_150bp -rs 1724163574\n",
      "Reference Sequences:\n",
      " @SQ\t11089:ncbi:1\t1\tAY968064\t11089\tncbi\tAngola_1971\t10237\n"
     ]
    }
   ],
   "source": [
    "aln = AlnFileReader(p2aln)\n",
    "for i, aln_read in enumerate(aln):\n",
    "    pass\n",
    "print(f\"{i+1:,d} simulated reads in file {p2aln.name} from {len(aln.header['reference sequences'])} refseq.\")\n",
    "\n",
    "print('ART command: ',aln.header['command'])\n",
    "print('Reference Sequences:\\n','\\n'.join(aln.header['reference sequences']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prediction Loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Label for Yellow Fever in `118`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Yellow_fever_virus', 118)]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: add functionality to do species search into OriginalLabels\n",
    "[(k,v) for k, v in OriginalLabels()._species2label.items() if 'yellow' in k.lower()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's test:\n",
    "- the generator (`aln.cnn_virus_input_generator`)\n",
    "- the string to tensor transform function (`string_input_batch_to_tensors`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(TensorShape([8, 150, 5]), TensorShape([8, 187]), TensorShape([8, 10]))"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aln.reset_iterator()\n",
    "for batch_idxs, batch_reads in aln.cnn_virus_input_generator(bs=8, label=118):\n",
    "    reads_kmer, (labels_kmer, positions_kmer) = string_input_batch_to_tensors(batch_reads, k=150)\n",
    "    break\n",
    "\n",
    "reads_kmer.shape, labels_kmer.shape, positions_kmer.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we apply `split_kmer_batch_into_50mers` to split the kmers into a series of 50-mer reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb 50-mer reads in a batch:  808\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "TensorShape([808, 50, 5])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads_50mer = split_kmer_batch_into_50mers(reads_kmer)\n",
    "nb_50mer_reads = (reads_kmer.shape[1]-49) * reads_kmer.shape[0]\n",
    "print('Nb 50-mer reads in a batch: ',nb_50mer_reads)\n",
    "assert reads_50mer.shape == (nb_50mer_reads, 50,5)\n",
    "\n",
    "reads_50mer.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run the Loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before running the next cell, define:\n",
    "- size of the batch `b`\n",
    "- the number of bp in a k-mer\n",
    "- the number n for top-n predictions to save\n",
    "- whether to run the full loop or only a few batches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'single_1seq_150bp'"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aln = AlnFileReader(p2aln)\n",
    "p2aln.stem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['aln_start_pos', 'readid', 'readnb', 'refseq_strand', 'refseqid', 'refseqnb', 'refsource', 'reftaxonomyid', 'read_pos'])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aln.re_keys\n",
    "o = next(aln)\n",
    "aln.parse_definition_line_with_position(o['definition line']).keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_db(p2db: Path, k: int, top_n: int=5):\n",
    "    conn = sqlite3.connect(p2db)\n",
    "    cursor = conn.cursor()\n",
    "    query = \"SELECT name FROM sqlite_master WHERE type='table' AND name='predictions'\"\n",
    "    res = cursor.execute(query).fetchone()\n",
    "    if res is None or 'predictions' not in res:\n",
    "        print('Creating tables in database...')\n",
    "        create_tables(cursor, k)\n",
    "    return conn\n",
    "\n",
    "def table_columns(cursor, table):\n",
    "    query = f\"PRAGMA table_info({table})\"\n",
    "    cursor.execute(query)\n",
    "    cols = [row[1] for row in cursor.fetchall()]\n",
    "    return cols\n",
    "\n",
    "def create_tables(cursor, k, top_n=5):\n",
    "    # Create table for predictions\n",
    "    pred_cols_str = 'readid refseqid refsource refseq_strand taxonomyid'.split(' ')\n",
    "    pred_cols_int = 'lbl_true lbl_pred pos_true pos_pred'.split(' ')\n",
    "    top_pred_cols = [f\"top_{top_n}_lbl_pred_{i}\" for i in range(top_n)]\n",
    "\n",
    "    query = \"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS predictions (\n",
    "        id INTEGER PRIMARY KEY,\n",
    "    \"\"\"\n",
    "    for col in pred_cols_str:\n",
    "        query += f\"{col} TEXT, \"\n",
    "    for col in pred_cols_int:\n",
    "        query += f\"{col} INTEGER, \"\n",
    "    for col in top_pred_cols:\n",
    "        query += f\"{col} INTEGER, \"\n",
    "    query = query[:-2]+')'\n",
    "    # print(query)\n",
    "    cursor.execute(query)\n",
    "\n",
    "    query = \"CREATE INDEX idx_preds ON predictions (readid, refseqid, pos_true);\"\n",
    "    cursor.execute(query)\n",
    "    print('Table `predictions` created.')\n",
    "\n",
    "    # Create table for probabilities (one per 50-mer in order to keep small nb or columns in table)\n",
    "\n",
    "    query = f\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS label_probabilities (\n",
    "        id INTEGER PRIMARY KEY,\n",
    "        read_kmer_id TEXT,\n",
    "        read_50mer_nb INTEGER,\n",
    "    \"\"\"\n",
    "    query += ','.join([f\"prob_{i:03d}\" for i in range(187)]) + \" REAL, \"\n",
    "    query += \"FOREIGN KEY (read_kmer_id) REFERENCES predictions(readid)\"\n",
    "    query += ')'\n",
    "    cursor.execute(query)\n",
    "\n",
    "    query = \"CREATE INDEX idx_probs ON label_probabilities (read_kmer_id, read_50mer_nb);\"\n",
    "    cursor.execute(query)\n",
    "    print(f'Table `label_probabilities` created.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def top_predictions(probs, n=3):\n",
    "\n",
    "    def top_n_most_frequent(preds, n=3):\n",
    "        \"\"\"Returns the top n most frequent predictions for each read\"\"\"\n",
    "        # print(preds.shape)\n",
    "        uniques, counts = np.unique(preds, return_counts=True)\n",
    "        top_idx = np.argsort(counts)[-n:]\n",
    "        return uniques.take(top_idx)\n",
    "\n",
    "    top_preds_in_50mers = np.argsort(probs, axis=-1)[:, :, -n:]\n",
    "    nb_seq, nb_50mer, nb_lbls = top_preds_in_50mers.shape\n",
    "    # print(top_preds_in_50mers.shape)\n",
    "    top_preds_in_kmer = top_preds_in_50mers.reshape(nb_seq,nb_50mer * nb_lbls)\n",
    "    # print(top_preds_in_kmer.shape)\n",
    "\n",
    "    return np.apply_along_axis(top_n_most_frequent, axis=1, arr=top_preds_in_kmer, n=n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating tables in database...\n",
      "Table `predictions` created.\n",
      "Table `label_probabilities` created.\n"
     ]
    }
   ],
   "source": [
    "p2db = pfs.data / 'ncbi/infer_results/yf-ncbi' / f'{p2aln.stem}.db'\n",
    "conn = open_db(p2db, k=150)\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run prediction loop with the following parameters:\n",
      "   512 k-mer per batch; 150 bp per sequence; keep top-5 predictions\n",
      "Creating CNN Model (Original)\n",
      "Loading parameters from pretrained_model.h5\n",
      "Created pretrained model\n",
      "Model loaded and ready to run ...\n",
      "Creating tables in database...\n",
      "Table `predictions` created.\n",
      "Table `label_probabilities` created.\n",
      "Starting prediction loop ...\n",
      "20:04:19    Batch   1 ...\n",
      "20:04:19      Starting prediction for 512 kmer reads ...\n",
      "20:04:49      Reshaping predictions ...\n",
      "20:04:49      Combining predictions ...\n",
      "20:05:01      Preparing prediction report ...\n",
      "20:05:01      Saving batch prediction report to db...\n",
      "20:05:01      Preparing label probabilities report ...\n",
      "20:05:18      Saving batch label probabilities report to db...\n",
      "20:05:32      Batch processing time: 73.55 sec\n",
      "20:05:32    Batch   2 ...\n",
      "20:05:33      Starting prediction for 512 kmer reads ...\n",
      "20:06:02      Reshaping predictions ...\n",
      "20:06:02      Combining predictions ...\n",
      "20:06:13      Preparing prediction report ...\n",
      "20:06:13      Saving batch prediction report to db...\n",
      "20:06:13      Preparing label probabilities report ...\n",
      "20:06:31      Saving batch label probabilities report to db...\n",
      "20:06:45      Batch processing time: 72.32 sec\n",
      "20:06:45    Batch   3 ...\n",
      "20:06:45      Starting prediction for 512 kmer reads ...\n",
      "20:07:14      Reshaping predictions ...\n",
      "20:07:15      Combining predictions ...\n",
      "20:07:26      Preparing prediction report ...\n",
      "20:07:26      Saving batch prediction report to db...\n",
      "20:07:27      Preparing label probabilities report ...\n",
      "20:07:44      Saving batch label probabilities report to db...\n",
      "20:07:57      Batch processing time: 72.55 sec\n",
      "20:07:57    Batch   4 ...\n",
      "20:07:58      Starting prediction for 512 kmer reads ...\n",
      "20:08:27      Reshaping predictions ...\n",
      "20:08:27      Combining predictions ...\n",
      "20:08:38      Preparing prediction report ...\n",
      "20:08:38      Saving batch prediction report to db...\n",
      "20:08:38      Preparing label probabilities report ...\n",
      "20:08:57      Saving batch label probabilities report to db...\n",
      "20:09:10      Batch processing time: 72.74 sec\n",
      "20:09:10    Batch   5 ...\n",
      "20:09:10      Starting prediction for 512 kmer reads ...\n",
      "20:09:40      Reshaping predictions ...\n",
      "20:09:40      Combining predictions ...\n",
      "20:09:50      Preparing prediction report ...\n",
      "20:09:50      Saving batch prediction report to db...\n",
      "20:09:51      Preparing label probabilities report ...\n",
      "20:10:09      Saving batch label probabilities report to db...\n",
      "20:10:23      Batch processing time: 73.12 sec\n",
      "20:10:23    Batch   6 ...\n",
      "20:10:24      Starting prediction for 512 kmer reads ...\n",
      "20:10:53      Reshaping predictions ...\n",
      "20:10:53      Combining predictions ...\n",
      "20:11:04      Preparing prediction report ...\n",
      "20:11:04      Saving batch prediction report to db...\n",
      "20:11:04      Preparing label probabilities report ...\n",
      "20:11:22      Saving batch label probabilities report to db...\n",
      "20:11:36      Batch processing time: 72.80 sec\n",
      "20:11:36    Batch   7 ...\n",
      "20:11:36      Starting prediction for 512 kmer reads ...\n",
      "20:12:06      Reshaping predictions ...\n",
      "20:12:06      Combining predictions ...\n",
      "20:12:17      Preparing prediction report ...\n",
      "20:12:17      Saving batch prediction report to db...\n",
      "20:12:17      Preparing label probabilities report ...\n",
      "20:12:36      Saving batch label probabilities report to db...\n",
      "20:12:51      Batch processing time: 74.90 sec\n",
      "20:12:51    Batch   8 ...\n",
      "20:12:51      Starting prediction for 512 kmer reads ...\n",
      "20:13:21      Reshaping predictions ...\n",
      "20:13:21      Combining predictions ...\n",
      "20:13:32      Preparing prediction report ...\n",
      "20:13:32      Saving batch prediction report to db...\n",
      "20:13:33      Preparing label probabilities report ...\n",
      "20:13:51      Saving batch label probabilities report to db...\n",
      "20:14:05      Batch processing time: 74.65 sec\n",
      "20:14:05    Batch   9 ...\n",
      "20:14:06      Starting prediction for 512 kmer reads ...\n",
      "20:14:35      Reshaping predictions ...\n",
      "20:14:35      Combining predictions ...\n",
      "20:14:47      Preparing prediction report ...\n",
      "20:14:47      Saving batch prediction report to db...\n",
      "20:14:47      Preparing label probabilities report ...\n",
      "20:15:05      Saving batch label probabilities report to db...\n",
      "20:15:19      Batch processing time: 73.84 sec\n",
      "20:15:19    Batch  10 ...\n",
      "20:15:20      Starting prediction for 512 kmer reads ...\n",
      "20:15:49      Reshaping predictions ...\n",
      "20:15:49      Combining predictions ...\n",
      "20:16:01      Preparing prediction report ...\n",
      "20:16:01      Saving batch prediction report to db...\n",
      "20:16:02      Preparing label probabilities report ...\n",
      "20:16:20      Saving batch label probabilities report to db...\n",
      "20:16:34      Batch processing time: 74.82 sec\n",
      "20:16:34    Batch  11 ...\n",
      "20:16:35      Starting prediction for 512 kmer reads ...\n",
      "20:17:04      Reshaping predictions ...\n",
      "20:17:04      Combining predictions ...\n",
      "20:17:15      Preparing prediction report ...\n",
      "20:17:15      Saving batch prediction report to db...\n",
      "20:17:16      Preparing label probabilities report ...\n",
      "20:17:33      Saving batch label probabilities report to db...\n",
      "20:17:48      Batch processing time: 73.62 sec\n",
      "20:17:48    Batch  12 ...\n",
      "20:17:48      Starting prediction for 512 kmer reads ...\n",
      "20:18:18      Reshaping predictions ...\n",
      "20:18:18      Combining predictions ...\n",
      "20:18:26      Preparing prediction report ...\n",
      "20:18:26      Saving batch prediction report to db...\n",
      "20:18:27      Preparing label probabilities report ...\n",
      "20:18:42      Saving batch label probabilities report to db...\n",
      "20:18:55      Batch processing time: 67.59 sec\n",
      "20:18:55    Batch  13 ...\n",
      "20:18:56      Starting prediction for 512 kmer reads ...\n",
      "20:19:25      Reshaping predictions ...\n",
      "20:19:25      Combining predictions ...\n",
      "20:19:36      Preparing prediction report ...\n",
      "20:19:36      Saving batch prediction report to db...\n",
      "20:19:36      Preparing label probabilities report ...\n",
      "20:19:52      Saving batch label probabilities report to db...\n",
      "20:20:06      Batch processing time: 70.70 sec\n",
      "20:20:06    Batch  14 ...\n",
      "20:20:07      Starting prediction for 512 kmer reads ...\n",
      "20:20:36      Reshaping predictions ...\n",
      "20:20:36      Combining predictions ...\n",
      "20:20:45      Preparing prediction report ...\n",
      "20:20:45      Saving batch prediction report to db...\n",
      "20:20:45      Preparing label probabilities report ...\n",
      "20:21:00      Saving batch label probabilities report to db...\n",
      "20:21:15      Batch processing time: 68.64 sec\n",
      "20:21:15    Batch  15 ...\n",
      "20:21:15      Starting prediction for 512 kmer reads ...\n",
      "20:21:44      Reshaping predictions ...\n",
      "20:21:44      Combining predictions ...\n",
      "20:21:53      Preparing prediction report ...\n",
      "20:21:53      Saving batch prediction report to db...\n",
      "20:21:54      Preparing label probabilities report ...\n",
      "20:22:08      Saving batch label probabilities report to db...\n",
      "20:22:24      Batch processing time: 69.03 sec\n",
      "20:22:24    Batch  16 ...\n",
      "20:22:24      Starting prediction for 512 kmer reads ...\n",
      "20:22:54      Reshaping predictions ...\n",
      "20:22:54      Combining predictions ...\n",
      "20:23:02      Preparing prediction report ...\n",
      "20:23:02      Saving batch prediction report to db...\n",
      "20:23:04      Preparing label probabilities report ...\n",
      "20:23:19      Saving batch label probabilities report to db...\n",
      "20:23:35      Batch processing time: 71.12 sec\n",
      "20:23:35    Batch  17 ...\n",
      "20:23:35      Starting prediction for 512 kmer reads ...\n",
      "20:24:05      Reshaping predictions ...\n",
      "20:24:05      Combining predictions ...\n",
      "20:24:14      Preparing prediction report ...\n",
      "20:24:14      Saving batch prediction report to db...\n",
      "20:24:14      Preparing label probabilities report ...\n",
      "20:24:29      Saving batch label probabilities report to db...\n",
      "20:24:45      Batch processing time: 69.87 sec\n",
      "20:24:45    Batch  18 ...\n",
      "20:24:45      Starting prediction for 512 kmer reads ...\n",
      "20:25:15      Reshaping predictions ...\n",
      "20:25:15      Combining predictions ...\n",
      "20:25:23      Preparing prediction report ...\n",
      "20:25:23      Saving batch prediction report to db...\n",
      "20:25:24      Preparing label probabilities report ...\n",
      "20:25:39      Saving batch label probabilities report to db...\n",
      "20:25:54      Batch processing time: 69.70 sec\n",
      "20:25:54    Batch  19 ...\n",
      "20:25:55      Starting prediction for 512 kmer reads ...\n",
      "20:26:24      Reshaping predictions ...\n",
      "20:26:24      Combining predictions ...\n",
      "20:26:35      Preparing prediction report ...\n",
      "20:26:35      Saving batch prediction report to db...\n",
      "20:26:36      Preparing label probabilities report ...\n",
      "20:26:52      Saving batch label probabilities report to db...\n",
      "20:27:09      Batch processing time: 74.10 sec\n",
      "20:27:09    Batch  20 ...\n",
      "20:27:09      Starting prediction for 512 kmer reads ...\n",
      "20:27:38      Reshaping predictions ...\n",
      "20:27:38      Combining predictions ...\n",
      "20:27:47      Preparing prediction report ...\n",
      "20:27:47      Saving batch prediction report to db...\n",
      "20:27:48      Preparing label probabilities report ...\n",
      "20:28:05      Saving batch label probabilities report to db...\n",
      "20:28:19      Batch processing time: 70.17 sec\n",
      "20:28:19    Batch  21 ...\n",
      "20:28:19      Starting prediction for 512 kmer reads ...\n",
      "20:28:49      Reshaping predictions ...\n",
      "20:28:49      Combining predictions ...\n",
      "20:28:59      Preparing prediction report ...\n",
      "20:28:59      Saving batch prediction report to db...\n",
      "20:29:00      Preparing label probabilities report ...\n",
      "20:29:16      Saving batch label probabilities report to db...\n",
      "20:29:31      Batch processing time: 72.14 sec\n",
      "20:29:31    Batch  22 ...\n",
      "20:29:31      Starting prediction for 512 kmer reads ...\n",
      "20:30:01      Reshaping predictions ...\n",
      "20:30:01      Combining predictions ...\n",
      "20:30:11      Preparing prediction report ...\n",
      "20:30:11      Saving batch prediction report to db...\n",
      "20:30:13      Preparing label probabilities report ...\n",
      "20:30:28      Saving batch label probabilities report to db...\n",
      "20:30:44      Batch processing time: 72.93 sec\n",
      "20:30:44    Batch  23 ...\n",
      "20:30:44      Starting prediction for 512 kmer reads ...\n",
      "20:31:14      Reshaping predictions ...\n",
      "20:31:14      Combining predictions ...\n",
      "20:31:23      Preparing prediction report ...\n",
      "20:31:23      Saving batch prediction report to db...\n",
      "20:31:24      Preparing label probabilities report ...\n",
      "20:31:40      Saving batch label probabilities report to db...\n",
      "20:31:55      Batch processing time: 71.50 sec\n",
      "20:31:55    Batch  24 ...\n",
      "20:31:56      Starting prediction for 512 kmer reads ...\n",
      "20:32:25      Reshaping predictions ...\n",
      "20:32:25      Combining predictions ...\n",
      "20:32:36      Preparing prediction report ...\n",
      "20:32:36      Saving batch prediction report to db...\n",
      "20:32:37      Preparing label probabilities report ...\n",
      "20:32:54      Saving batch label probabilities report to db...\n",
      "20:33:08      Batch processing time: 72.46 sec\n",
      "20:33:08    Batch  25 ...\n",
      "20:33:08      Starting prediction for 512 kmer reads ...\n",
      "20:33:38      Reshaping predictions ...\n",
      "20:33:38      Combining predictions ...\n",
      "20:33:46      Preparing prediction report ...\n",
      "20:33:46      Saving batch prediction report to db...\n",
      "20:33:47      Preparing label probabilities report ...\n",
      "20:34:02      Saving batch label probabilities report to db...\n",
      "20:34:17      Batch processing time: 69.41 sec\n",
      "20:34:17    Batch  26 ...\n",
      "20:34:18      Starting prediction for 512 kmer reads ...\n",
      "20:34:47      Reshaping predictions ...\n",
      "20:34:47      Combining predictions ...\n",
      "20:34:57      Preparing prediction report ...\n",
      "20:34:57      Saving batch prediction report to db...\n",
      "20:34:57      Preparing label probabilities report ...\n",
      "20:35:13      Saving batch label probabilities report to db...\n",
      "20:35:29      Batch processing time: 71.55 sec\n",
      "20:35:29    Batch  27 ...\n",
      "20:35:29      Starting prediction for 512 kmer reads ...\n",
      "20:35:59      Reshaping predictions ...\n",
      "20:35:59      Combining predictions ...\n",
      "20:36:07      Preparing prediction report ...\n",
      "20:36:07      Saving batch prediction report to db...\n",
      "20:36:07      Preparing label probabilities report ...\n",
      "20:36:23      Saving batch label probabilities report to db...\n",
      "20:36:37      Batch processing time: 67.81 sec\n",
      "20:36:37    Batch  28 ...\n",
      "20:36:37      Starting prediction for 512 kmer reads ...\n",
      "20:37:07      Reshaping predictions ...\n",
      "20:37:07      Combining predictions ...\n",
      "20:37:16      Preparing prediction report ...\n",
      "20:37:16      Saving batch prediction report to db...\n",
      "20:37:16      Preparing label probabilities report ...\n",
      "20:37:32      Saving batch label probabilities report to db...\n",
      "20:37:46      Batch processing time: 69.50 sec\n",
      "20:37:46    Batch  29 ...\n",
      "20:37:47      Starting prediction for 512 kmer reads ...\n",
      "20:38:16      Reshaping predictions ...\n",
      "20:38:16      Combining predictions ...\n",
      "20:38:25      Preparing prediction report ...\n",
      "20:38:25      Saving batch prediction report to db...\n",
      "20:38:25      Preparing label probabilities report ...\n",
      "20:38:41      Saving batch label probabilities report to db...\n",
      "20:38:57      Batch processing time: 70.39 sec\n",
      "20:38:57    Batch  30 ...\n",
      "20:38:57      Starting prediction for 512 kmer reads ...\n",
      "20:39:26      Reshaping predictions ...\n",
      "20:39:26      Combining predictions ...\n",
      "20:39:35      Preparing prediction report ...\n",
      "20:39:35      Saving batch prediction report to db...\n",
      "20:39:36      Preparing label probabilities report ...\n",
      "20:39:52      Saving batch label probabilities report to db...\n",
      "20:40:07      Batch processing time: 70.64 sec\n",
      "20:40:07    Batch  31 ...\n",
      "20:40:08      Starting prediction for 512 kmer reads ...\n",
      "20:40:37      Reshaping predictions ...\n",
      "20:40:37      Combining predictions ...\n",
      "20:40:48      Preparing prediction report ...\n",
      "20:40:48      Saving batch prediction report to db...\n",
      "20:40:48      Preparing label probabilities report ...\n",
      "20:41:03      Saving batch label probabilities report to db...\n",
      "20:41:17      Batch processing time: 69.90 sec\n",
      "20:41:17    Batch  32 ...\n",
      "20:41:18      Starting prediction for 512 kmer reads ...\n",
      "20:41:47      Reshaping predictions ...\n",
      "20:41:47      Combining predictions ...\n",
      "20:41:57      Preparing prediction report ...\n",
      "20:41:57      Saving batch prediction report to db...\n",
      "20:41:57      Preparing label probabilities report ...\n",
      "20:42:16      Saving batch label probabilities report to db...\n",
      "20:42:30      Batch processing time: 72.92 sec\n",
      "20:42:30    Batch  33 ...\n",
      "20:42:31      Starting prediction for 512 kmer reads ...\n",
      "20:43:00      Reshaping predictions ...\n",
      "20:43:00      Combining predictions ...\n",
      "20:43:09      Preparing prediction report ...\n",
      "20:43:09      Saving batch prediction report to db...\n",
      "20:43:09      Preparing label probabilities report ...\n",
      "20:43:24      Saving batch label probabilities report to db...\n",
      "20:43:40      Batch processing time: 69.60 sec\n"
     ]
    }
   ],
   "source": [
    "b = 512   # number of k-mer in a batch\n",
    "k = 150\n",
    "true_label = 118\n",
    "top_n = 5   # n for top-n prediction to keep\n",
    "run_all_batches = True\n",
    "nb_batches_to_run = 2\n",
    "\n",
    "uid = datetime.today().strftime('%Y-%m-%d_%H_%M_%S')\n",
    "p2db = pfs.data / 'ncbi/infer_results/yf-ncbi' / f'{p2aln.stem}.db'\n",
    "nb_50mer = k - 49\n",
    "\n",
    "print(f\"Run prediction loop with the following parameters:\")\n",
    "print(f\"   {b} k-mer per batch; {k} bp per sequence; keep top-{top_n} predictions\")\n",
    "\n",
    "aln.reset_iterator()\n",
    "model = create_model_original(path2parameters=p2model)\n",
    "print(f\"Model loaded and ready to run ...\")\n",
    "\n",
    "# open connection to sqlite db\n",
    "conn = open_db(p2db, k=k)\n",
    "cursor = conn.cursor()\n",
    "\n",
    "# create a dataframe to store results\n",
    "pred_cols_str = 'readid refseqid refsource refseq_strand taxonomyid'.split(' ')\n",
    "pred_cols_int = 'lbl_true lbl_pred pos_true pos_pred'.split(' ')\n",
    "top_pred_cols = [f\"top_{top_n}_lbl_pred_{i}\" for i in range(top_n)]\n",
    "prob_cols = [f\"prob_{i:03d}\" for i in range(187)]\n",
    "\n",
    "\n",
    "def tprint(string):\n",
    "    print(f\"{datetime.now().strftime('%H:%M:%S')}    {string}\")\n",
    "\n",
    "print(f\"Starting prediction loop ...\")\n",
    "for i,(metadata_batch, reads_batch) in enumerate(aln.cnn_virus_input_generator(bs=b, label=true_label)):\n",
    "    loop_start = datetime.now()\n",
    "    tprint(f\"Batch {i+1:3d} ...\")\n",
    "\n",
    "    reads_kmer, (labels_true, position_true) = string_input_batch_to_tensors(reads_batch, k=k)\n",
    "    reads_50mer = split_kmer_batch_into_50mers(reads_kmer)\n",
    "    assert reads_50mer.shape == ((reads_kmer.shape[1]-49) * b, 50, 5), f\"Problem with shape in batch {i+1}: {reads_50mer.shape}\"\n",
    "\n",
    "    tprint(f'  Starting prediction for {b:,} kmer reads ...')\n",
    "    label_probs, position_probs = model.predict(reads_50mer)\n",
    "\n",
    "    tprint('  Reshaping predictions ...')\n",
    "    label_probs_kmer = tf.reshape(label_probs, shape=(b,nb_50mer,-1))\n",
    "    position_probs_kmer = tf.reshape(position_probs, shape=(b,nb_50mer,-1))\n",
    "    # successive_preds = count_successive_label_preds(label_probs_kmer)\n",
    "\n",
    "    tprint('  Combining predictions ...')\n",
    "    combined_predictions = tf.map_fn(\n",
    "        fn=combine_prediction_batch,\n",
    "        elems=[label_probs_kmer, position_probs_kmer], \n",
    "        fn_output_signature=tf.int64\n",
    "        )\n",
    "\n",
    "    label_predictions = combined_predictions[:,0]\n",
    "    position_predictions = combined_predictions[:,1]\n",
    "    top_preds = top_predictions(label_probs_kmer, n=top_n)\n",
    "\n",
    "    # Add results for batch\n",
    "    tprint('  Preparing prediction report ...')\n",
    "    # 'readid', 'refseqid', 'refsource', 'refseq_strand', 'taxonomyid',\n",
    "    # 'lbl_true', 'lbl_pred', 'pos_true', 'pos_pred'\n",
    "    # 'top_5_lbl_pred_0', 'top_5_lbl_pred_1', 'top_5_lbl_pred_2',\n",
    "    # 'top_5_lbl_pred_3', 'top_5_lbl_pred_4'\n",
    "    preds_report = np.concatenate(\n",
    "        [\n",
    "            np.expand_dims(np.array(metadata_batch['readid']), axis=1),         # readid \n",
    "            np.expand_dims(np.array(metadata_batch['refseqid']), axis=1),       # refseqid\n",
    "            np.expand_dims(np.array(metadata_batch['refsource']), axis=1),      # refsource\n",
    "            np.expand_dims(np.array(metadata_batch['refseq_strand']), axis=1),  # refseq_strand\n",
    "            np.expand_dims(np.array(metadata_batch['reftaxonomyid']), axis=1),  # taxonomyid\n",
    "            np.expand_dims(np.array([true_label]*b), axis=1),                   # lbl_true\n",
    "            np.expand_dims(label_predictions, axis=1),                          # lbl_pred\n",
    "            np.expand_dims(np.array(metadata_batch['aln_start_pos']), axis=1),  # pos_true\n",
    "            np.expand_dims(position_predictions, axis=1),                       # pos_pred\n",
    "            top_preds[:, ::-1],                                                 # top_5_lbl_pred_0, top_5_lbl_pred_1, top_5_lbl_pred_2, top_5_lbl_pred_3, top_5_lbl_pred_4\n",
    "        ],\n",
    "        axis=1\n",
    "    )\n",
    "\n",
    "    df_preds = pd.DataFrame(\n",
    "        data=preds_report, \n",
    "        columns=pred_cols_str + pred_cols_int + top_pred_cols\n",
    "        )\n",
    "    tprint('  Saving batch prediction report to db...')\n",
    "    df_preds.to_sql('predictions', conn, if_exists='append', index=False)\n",
    "\n",
    "    tprint('  Preparing label probabilities report ...')\n",
    "    df_probs = None\n",
    "    for read_50mer_nb in range(nb_50mer):\n",
    "        probs_report = np.concatenate(\n",
    "            [\n",
    "                np.expand_dims(np.array(metadata_batch['readid']), axis=1),        # readid \n",
    "                np.expand_dims(np.array([read_50mer_nb]*b), axis=1),               # read_50mer_nb\n",
    "                label_probs_kmer[:, read_50mer_nb, :]\n",
    "            ],\n",
    "            axis=1\n",
    "        )\n",
    "\n",
    "        df = pd.DataFrame(\n",
    "            data=probs_report, \n",
    "            columns=['read_kmer_id', 'read_50mer_nb'] + prob_cols\n",
    "            )\n",
    "        df_probs = df if df_probs is None else pd.concat([df_probs, df], axis=0)\n",
    "\n",
    "    tprint('  Saving batch label probabilities report to db...')\n",
    "    df_probs.to_sql('label_probabilities', conn, if_exists='append', index=False)\n",
    "\n",
    "    tprint(f\"  Batch processing time: {(datetime.now() - loop_start).total_seconds():.2f} sec\")\n",
    "    if not run_all_batches and i+1 >= nb_batches_to_run: \n",
    "        print('Stopping')\n",
    "        break\n",
    "\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Estimated size of DB:\n",
    "- 17,000 kmer reads -> 6,844,380 kb\n",
    "- kb per kmer reads = 405 kb\n",
    "\n",
    "Total nbr of kmer reads for 69 sequences: 1,161,034\n",
    "- 1,161,034 * 405 kb = 470,000,000 kb = **470 Gb**\n",
    "\n",
    "Inference time: one batch of 512 kmer read takes 30 s\n",
    "- 17,000 / 512 * 30 s = 995 s = 16.6 min\n",
    "- 1,161,034/512 * 30 s = 57,000 s = 15.8 h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>readid</th>\n",
       "      <th>refseqid</th>\n",
       "      <th>refsource</th>\n",
       "      <th>refseq_strand</th>\n",
       "      <th>taxonomyid</th>\n",
       "      <th>lbl_true</th>\n",
       "      <th>lbl_pred</th>\n",
       "      <th>pos_true</th>\n",
       "      <th>pos_pred</th>\n",
       "      <th>read_pos</th>\n",
       "      <th>top_5_lbl_pred_0</th>\n",
       "      <th>top_5_lbl_pred_1</th>\n",
       "      <th>top_5_lbl_pred_2</th>\n",
       "      <th>top_5_lbl_pred_3</th>\n",
       "      <th>top_5_lbl_pred_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>11089:ncbi:1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>-</td>\n",
       "      <td>11089</td>\n",
       "      <td>118</td>\n",
       "      <td>32</td>\n",
       "      <td>3164</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>117</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id              readid      refseqid refsource refseq_strand taxonomyid  \\\n",
       "0   1  11089:ncbi:1-17000  11089:ncbi:1      ncbi             -      11089   \n",
       "\n",
       "   lbl_true  lbl_pred  pos_true  pos_pred read_pos  top_5_lbl_pred_0  \\\n",
       "0       118        32      3164         7     None                18   \n",
       "\n",
       "   top_5_lbl_pred_1  top_5_lbl_pred_2  top_5_lbl_pred_3  top_5_lbl_pred_4  \n",
       "0                 0                32               117               118  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>read_kmer_id</th>\n",
       "      <th>read_50mer_nb</th>\n",
       "      <th>prob_000</th>\n",
       "      <th>prob_001</th>\n",
       "      <th>prob_002</th>\n",
       "      <th>prob_003</th>\n",
       "      <th>prob_004</th>\n",
       "      <th>prob_005</th>\n",
       "      <th>prob_006</th>\n",
       "      <th>...</th>\n",
       "      <th>prob_177</th>\n",
       "      <th>prob_178</th>\n",
       "      <th>prob_179</th>\n",
       "      <th>prob_180</th>\n",
       "      <th>prob_181</th>\n",
       "      <th>prob_182</th>\n",
       "      <th>prob_183</th>\n",
       "      <th>prob_184</th>\n",
       "      <th>prob_185</th>\n",
       "      <th>prob_186</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0779117e-09</td>\n",
       "      <td>4.6550543e-16</td>\n",
       "      <td>4.0059014e-09</td>\n",
       "      <td>5.709354e-14</td>\n",
       "      <td>3.985621e-13</td>\n",
       "      <td>1.4781354e-07</td>\n",
       "      <td>3.6368833e-13</td>\n",
       "      <td>...</td>\n",
       "      <td>6.536237e-14</td>\n",
       "      <td>1.3692335e-18</td>\n",
       "      <td>4.8284808e-14</td>\n",
       "      <td>4.28663e-27</td>\n",
       "      <td>1.7636006e-08</td>\n",
       "      <td>2.0901565e-20</td>\n",
       "      <td>1.2649406e-15</td>\n",
       "      <td>3.2826548e-25</td>\n",
       "      <td>1.3461108e-31</td>\n",
       "      <td>1.695962e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>513</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>1</td>\n",
       "      <td>7.798936e-12</td>\n",
       "      <td>3.0194006e-16</td>\n",
       "      <td>9.897284e-13</td>\n",
       "      <td>4.3037067e-11</td>\n",
       "      <td>8.870653e-16</td>\n",
       "      <td>5.2778373e-09</td>\n",
       "      <td>5.065814e-15</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4544616e-15</td>\n",
       "      <td>1.7843624e-16</td>\n",
       "      <td>9.904177e-13</td>\n",
       "      <td>3.457488e-21</td>\n",
       "      <td>7.375788e-07</td>\n",
       "      <td>1.7063393e-21</td>\n",
       "      <td>6.5437836e-11</td>\n",
       "      <td>5.6732623e-25</td>\n",
       "      <td>4.585938e-27</td>\n",
       "      <td>1.069439e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1025</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0980089e-12</td>\n",
       "      <td>1.9275111e-15</td>\n",
       "      <td>1.2620867e-10</td>\n",
       "      <td>5.787828e-11</td>\n",
       "      <td>5.338331e-13</td>\n",
       "      <td>2.7942612e-11</td>\n",
       "      <td>5.7700195e-10</td>\n",
       "      <td>...</td>\n",
       "      <td>5.753128e-14</td>\n",
       "      <td>1.6327466e-15</td>\n",
       "      <td>4.4719354e-06</td>\n",
       "      <td>6.113753e-17</td>\n",
       "      <td>4.3007142e-07</td>\n",
       "      <td>1.6321036e-17</td>\n",
       "      <td>1.3829523e-09</td>\n",
       "      <td>7.2506775e-19</td>\n",
       "      <td>4.256189e-23</td>\n",
       "      <td>8.687584e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1537</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>3</td>\n",
       "      <td>4.4603436e-09</td>\n",
       "      <td>4.5842836e-14</td>\n",
       "      <td>3.9782825e-12</td>\n",
       "      <td>1.9334909e-08</td>\n",
       "      <td>1.5190636e-10</td>\n",
       "      <td>2.5700672e-08</td>\n",
       "      <td>1.488595e-13</td>\n",
       "      <td>...</td>\n",
       "      <td>2.6302808e-07</td>\n",
       "      <td>7.203852e-11</td>\n",
       "      <td>4.3689414e-07</td>\n",
       "      <td>2.1555071e-17</td>\n",
       "      <td>0.000116948606</td>\n",
       "      <td>5.1792988e-18</td>\n",
       "      <td>1.0885436e-09</td>\n",
       "      <td>3.184696e-18</td>\n",
       "      <td>1.5609967e-19</td>\n",
       "      <td>3.963395e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2049</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>4</td>\n",
       "      <td>8.554256e-06</td>\n",
       "      <td>2.554389e-13</td>\n",
       "      <td>0.0006343455</td>\n",
       "      <td>3.5571188e-07</td>\n",
       "      <td>2.3673221e-09</td>\n",
       "      <td>0.097307</td>\n",
       "      <td>7.587622e-09</td>\n",
       "      <td>...</td>\n",
       "      <td>3.750018e-07</td>\n",
       "      <td>8.665608e-11</td>\n",
       "      <td>6.433393e-09</td>\n",
       "      <td>5.5440495e-20</td>\n",
       "      <td>0.00476838</td>\n",
       "      <td>1.4658719e-17</td>\n",
       "      <td>5.7866656e-13</td>\n",
       "      <td>1.6711752e-17</td>\n",
       "      <td>5.3012568e-21</td>\n",
       "      <td>1.876316e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>49153</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>96</td>\n",
       "      <td>0.9953505</td>\n",
       "      <td>5.3798974e-20</td>\n",
       "      <td>1.0473234e-18</td>\n",
       "      <td>6.393374e-11</td>\n",
       "      <td>2.8581829e-10</td>\n",
       "      <td>3.2485372e-15</td>\n",
       "      <td>2.7151905e-12</td>\n",
       "      <td>...</td>\n",
       "      <td>5.3093893e-15</td>\n",
       "      <td>1.3183516e-18</td>\n",
       "      <td>3.1293095e-09</td>\n",
       "      <td>2.2244292e-16</td>\n",
       "      <td>1.6725651e-10</td>\n",
       "      <td>3.994397e-15</td>\n",
       "      <td>2.2199666e-19</td>\n",
       "      <td>1.2594476e-13</td>\n",
       "      <td>3.5062112e-13</td>\n",
       "      <td>1.599368e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>49665</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>97</td>\n",
       "      <td>0.7476607</td>\n",
       "      <td>8.0384427e-16</td>\n",
       "      <td>3.6662344e-12</td>\n",
       "      <td>1.3936913e-13</td>\n",
       "      <td>3.1677278e-12</td>\n",
       "      <td>3.3102418e-15</td>\n",
       "      <td>1.0410654e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8784293e-12</td>\n",
       "      <td>3.711314e-15</td>\n",
       "      <td>1.8920882e-05</td>\n",
       "      <td>3.1332e-16</td>\n",
       "      <td>4.1800877e-07</td>\n",
       "      <td>5.0522433e-12</td>\n",
       "      <td>1.0039908e-14</td>\n",
       "      <td>6.3892506e-13</td>\n",
       "      <td>5.0943116e-14</td>\n",
       "      <td>1.197173e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>50177</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>98</td>\n",
       "      <td>0.9361404</td>\n",
       "      <td>2.664984e-16</td>\n",
       "      <td>7.8721495e-17</td>\n",
       "      <td>3.706427e-12</td>\n",
       "      <td>1.3358595e-14</td>\n",
       "      <td>1.254372e-16</td>\n",
       "      <td>1.8388244e-10</td>\n",
       "      <td>...</td>\n",
       "      <td>1.06356905e-16</td>\n",
       "      <td>4.1285527e-16</td>\n",
       "      <td>1.4968358e-11</td>\n",
       "      <td>2.585157e-18</td>\n",
       "      <td>8.469594e-08</td>\n",
       "      <td>3.6794658e-11</td>\n",
       "      <td>2.3980224e-18</td>\n",
       "      <td>5.1243676e-15</td>\n",
       "      <td>1.01574735e-17</td>\n",
       "      <td>2.688872e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>50689</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>99</td>\n",
       "      <td>0.998154</td>\n",
       "      <td>1.2311882e-15</td>\n",
       "      <td>1.7796784e-17</td>\n",
       "      <td>5.404334e-11</td>\n",
       "      <td>1.2882388e-10</td>\n",
       "      <td>1.5932526e-10</td>\n",
       "      <td>8.845724e-11</td>\n",
       "      <td>...</td>\n",
       "      <td>1.8041767e-10</td>\n",
       "      <td>3.959757e-12</td>\n",
       "      <td>3.682707e-12</td>\n",
       "      <td>2.1554382e-12</td>\n",
       "      <td>4.0533246e-06</td>\n",
       "      <td>5.84147e-08</td>\n",
       "      <td>6.353569e-15</td>\n",
       "      <td>1.6207057e-10</td>\n",
       "      <td>1.1721044e-15</td>\n",
       "      <td>3.890514e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>51201</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>100</td>\n",
       "      <td>0.22425258</td>\n",
       "      <td>4.8676975e-13</td>\n",
       "      <td>3.5484584e-15</td>\n",
       "      <td>6.286311e-08</td>\n",
       "      <td>1.3584297e-11</td>\n",
       "      <td>9.7854135e-11</td>\n",
       "      <td>2.0057087e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>2.981896e-11</td>\n",
       "      <td>1.4885541e-12</td>\n",
       "      <td>0.601138</td>\n",
       "      <td>5.832164e-17</td>\n",
       "      <td>3.1379425e-08</td>\n",
       "      <td>1.38640115e-11</td>\n",
       "      <td>7.054615e-14</td>\n",
       "      <td>4.1462553e-09</td>\n",
       "      <td>5.9365905e-15</td>\n",
       "      <td>3.292315e-18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows × 190 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        id        read_kmer_id  read_50mer_nb       prob_000       prob_001  \\\n",
       "0        1  11089:ncbi:1-17000              0  3.0779117e-09  4.6550543e-16   \n",
       "1      513  11089:ncbi:1-17000              1   7.798936e-12  3.0194006e-16   \n",
       "2     1025  11089:ncbi:1-17000              2  1.0980089e-12  1.9275111e-15   \n",
       "3     1537  11089:ncbi:1-17000              3  4.4603436e-09  4.5842836e-14   \n",
       "4     2049  11089:ncbi:1-17000              4   8.554256e-06   2.554389e-13   \n",
       "..     ...                 ...            ...            ...            ...   \n",
       "96   49153  11089:ncbi:1-17000             96      0.9953505  5.3798974e-20   \n",
       "97   49665  11089:ncbi:1-17000             97      0.7476607  8.0384427e-16   \n",
       "98   50177  11089:ncbi:1-17000             98      0.9361404   2.664984e-16   \n",
       "99   50689  11089:ncbi:1-17000             99       0.998154  1.2311882e-15   \n",
       "100  51201  11089:ncbi:1-17000            100     0.22425258  4.8676975e-13   \n",
       "\n",
       "          prob_002       prob_003       prob_004       prob_005  \\\n",
       "0    4.0059014e-09   5.709354e-14   3.985621e-13  1.4781354e-07   \n",
       "1     9.897284e-13  4.3037067e-11   8.870653e-16  5.2778373e-09   \n",
       "2    1.2620867e-10   5.787828e-11   5.338331e-13  2.7942612e-11   \n",
       "3    3.9782825e-12  1.9334909e-08  1.5190636e-10  2.5700672e-08   \n",
       "4     0.0006343455  3.5571188e-07  2.3673221e-09       0.097307   \n",
       "..             ...            ...            ...            ...   \n",
       "96   1.0473234e-18   6.393374e-11  2.8581829e-10  3.2485372e-15   \n",
       "97   3.6662344e-12  1.3936913e-13  3.1677278e-12  3.3102418e-15   \n",
       "98   7.8721495e-17   3.706427e-12  1.3358595e-14   1.254372e-16   \n",
       "99   1.7796784e-17   5.404334e-11  1.2882388e-10  1.5932526e-10   \n",
       "100  3.5484584e-15   6.286311e-08  1.3584297e-11  9.7854135e-11   \n",
       "\n",
       "          prob_006  ...        prob_177       prob_178       prob_179  \\\n",
       "0    3.6368833e-13  ...    6.536237e-14  1.3692335e-18  4.8284808e-14   \n",
       "1     5.065814e-15  ...   1.4544616e-15  1.7843624e-16   9.904177e-13   \n",
       "2    5.7700195e-10  ...    5.753128e-14  1.6327466e-15  4.4719354e-06   \n",
       "3     1.488595e-13  ...   2.6302808e-07   7.203852e-11  4.3689414e-07   \n",
       "4     7.587622e-09  ...    3.750018e-07   8.665608e-11   6.433393e-09   \n",
       "..             ...  ...             ...            ...            ...   \n",
       "96   2.7151905e-12  ...   5.3093893e-15  1.3183516e-18  3.1293095e-09   \n",
       "97   1.0410654e-08  ...   2.8784293e-12   3.711314e-15  1.8920882e-05   \n",
       "98   1.8388244e-10  ...  1.06356905e-16  4.1285527e-16  1.4968358e-11   \n",
       "99    8.845724e-11  ...   1.8041767e-10   3.959757e-12   3.682707e-12   \n",
       "100  2.0057087e-06  ...    2.981896e-11  1.4885541e-12       0.601138   \n",
       "\n",
       "          prob_180        prob_181        prob_182       prob_183  \\\n",
       "0      4.28663e-27   1.7636006e-08   2.0901565e-20  1.2649406e-15   \n",
       "1     3.457488e-21    7.375788e-07   1.7063393e-21  6.5437836e-11   \n",
       "2     6.113753e-17   4.3007142e-07   1.6321036e-17  1.3829523e-09   \n",
       "3    2.1555071e-17  0.000116948606   5.1792988e-18  1.0885436e-09   \n",
       "4    5.5440495e-20      0.00476838   1.4658719e-17  5.7866656e-13   \n",
       "..             ...             ...             ...            ...   \n",
       "96   2.2244292e-16   1.6725651e-10    3.994397e-15  2.2199666e-19   \n",
       "97      3.1332e-16   4.1800877e-07   5.0522433e-12  1.0039908e-14   \n",
       "98    2.585157e-18    8.469594e-08   3.6794658e-11  2.3980224e-18   \n",
       "99   2.1554382e-12   4.0533246e-06     5.84147e-08   6.353569e-15   \n",
       "100   5.832164e-17   3.1379425e-08  1.38640115e-11   7.054615e-14   \n",
       "\n",
       "          prob_184        prob_185      prob_186  \n",
       "0    3.2826548e-25   1.3461108e-31  1.695962e-21  \n",
       "1    5.6732623e-25    4.585938e-27  1.069439e-21  \n",
       "2    7.2506775e-19    4.256189e-23  8.687584e-22  \n",
       "3     3.184696e-18   1.5609967e-19  3.963395e-22  \n",
       "4    1.6711752e-17   5.3012568e-21  1.876316e-18  \n",
       "..             ...             ...           ...  \n",
       "96   1.2594476e-13   3.5062112e-13  1.599368e-17  \n",
       "97   6.3892506e-13   5.0943116e-14  1.197173e-14  \n",
       "98   5.1243676e-15  1.01574735e-17  2.688872e-14  \n",
       "99   1.6207057e-10   1.1721044e-15  3.890514e-13  \n",
       "100  4.1462553e-09   5.9365905e-15  3.292315e-18  \n",
       "\n",
       "[101 rows x 190 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with open_db(p2db=p2db, k=150) as conn:\n",
    "    display(pd.read_sql_query(\"SELECT * FROM predictions WHERE readid = '11089:ncbi:1-17000';\", conn))\n",
    "    display(pd.read_sql_query(\"SELECT * FROM label_probabilities WHERE read_kmer_id = '11089:ncbi:1-17000';\", conn))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open_db(p2db, 150) as conn:\n",
    "#     cursor = conn.cursor()\n",
    "#     cursor.execute(\"DELETE FROM predictions;\")\n",
    "#     cursor.execute(\"DELETE FROM label_probabilities;\")\n",
    "#     conn.commit()\n",
    "\n",
    "# with open_db(p2db, 150) as conn:\n",
    "#     cursor = conn.cursor()\n",
    "#     cursor.execute(\"DROP TABLE predictions\")\n",
    "#     cursor.execute(\"DROP TABLE label_probabilities\")\n",
    "#     conn.commit()\n",
    "\n",
    "# conn.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test saving the database on NAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/mnt/k/metagentools/ncbi/infer_results/yf-ncbi/single_1seq_150bp.db')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p2db = Path('/mnt/k/metagentools') / 'ncbi/infer_results/yf-ncbi' / f'{p2aln.stem}.db'\n",
    "assert p2db.parent.exists()\n",
    "p2db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run prediction loop with the following parameters:\n",
      "   32 k-mer per batch; 150 bp per sequence; keep top-5 predictions\n",
      "Creating CNN Model (Original)\n",
      "Loading parameters from pretrained_model.h5\n",
      "Created pretrained model\n",
      "Model loaded and ready to run ...\n",
      "Starting prediction loop ...\n",
      "21:26:08    Batch   1 ...\n",
      "21:26:08      Starting prediction for 32 kmer reads ...\n",
      "21:26:12      Reshaping predictions ...\n",
      "21:26:12      Combining predictions ...\n",
      "21:26:13      Preparing prediction report ...\n",
      "21:26:13      Saving batch prediction report to db...\n",
      "21:26:13      Preparing label probabilities report ...\n",
      "21:26:14      Saving batch label probabilities report to db...\n",
      "21:26:17      Batch processing time: 9.23 sec\n",
      "21:26:17    Batch   2 ...\n",
      "21:26:18      Starting prediction for 32 kmer reads ...\n",
      "21:26:19      Reshaping predictions ...\n",
      "21:26:19      Combining predictions ...\n",
      "21:26:20      Preparing prediction report ...\n",
      "21:26:20      Saving batch prediction report to db...\n",
      "21:26:20      Preparing label probabilities report ...\n",
      "21:26:21      Saving batch label probabilities report to db...\n",
      "21:26:24      Batch processing time: 7.02 sec\n",
      "21:26:24    Batch   3 ...\n",
      "21:26:25      Starting prediction for 32 kmer reads ...\n",
      "21:26:27      Reshaping predictions ...\n",
      "21:26:27      Combining predictions ...\n",
      "21:26:27      Preparing prediction report ...\n",
      "21:26:27      Saving batch prediction report to db...\n",
      "21:26:27      Preparing label probabilities report ...\n",
      "21:26:28      Saving batch label probabilities report to db...\n",
      "21:26:32      Batch processing time: 7.08 sec\n",
      "21:26:32    Batch   4 ...\n",
      "21:26:32      Starting prediction for 32 kmer reads ...\n",
      "21:26:33      Reshaping predictions ...\n",
      "21:26:33      Combining predictions ...\n",
      "21:26:34      Preparing prediction report ...\n",
      "21:26:34      Saving batch prediction report to db...\n",
      "21:26:34      Preparing label probabilities report ...\n",
      "21:26:35      Saving batch label probabilities report to db...\n",
      "21:26:39      Batch processing time: 7.25 sec\n",
      "Stopping\n"
     ]
    }
   ],
   "source": [
    "b = 32   # number of k-mer in a batch\n",
    "k = 150\n",
    "true_label = 118\n",
    "top_n = 5   # n for top-n prediction to keep\n",
    "run_all_batches = False\n",
    "nb_batches_to_run = 4\n",
    "\n",
    "uid = datetime.today().strftime('%Y-%m-%d_%H_%M_%S')\n",
    "# p2db = Path('/mnt/k') / 'ncbi/infer_results/yf-ncbi' / f'{p2aln.stem}.db'\n",
    "nb_50mer = k - 49\n",
    "\n",
    "print(f\"Run prediction loop with the following parameters:\")\n",
    "print(f\"   {b} k-mer per batch; {k} bp per sequence; keep top-{top_n} predictions\")\n",
    "\n",
    "aln.reset_iterator()\n",
    "model = create_model_original(path2parameters=p2model)\n",
    "print(f\"Model loaded and ready to run ...\")\n",
    "\n",
    "# open connection to sqlite db\n",
    "conn = open_db(p2db, k=k)\n",
    "cursor = conn.cursor()\n",
    "\n",
    "# create a dataframe to store results\n",
    "pred_cols_str = 'readid refseqid refsource refseq_strand taxonomyid'.split(' ')\n",
    "pred_cols_int = 'lbl_true lbl_pred pos_true pos_pred'.split(' ')\n",
    "top_pred_cols = [f\"top_{top_n}_lbl_pred_{i}\" for i in range(top_n)]\n",
    "prob_cols = [f\"prob_{i:03d}\" for i in range(187)]\n",
    "\n",
    "\n",
    "def tprint(string):\n",
    "    print(f\"{datetime.now().strftime('%H:%M:%S')}    {string}\")\n",
    "\n",
    "print(f\"Starting prediction loop ...\")\n",
    "for i,(metadata_batch, reads_batch) in enumerate(aln.cnn_virus_input_generator(bs=b, label=true_label)):\n",
    "    loop_start = datetime.now()\n",
    "    tprint(f\"Batch {i+1:3d} ...\")\n",
    "\n",
    "    reads_kmer, (labels_true, position_true) = string_input_batch_to_tensors(reads_batch, k=k)\n",
    "    reads_50mer = split_kmer_batch_into_50mers(reads_kmer)\n",
    "    assert reads_50mer.shape == ((reads_kmer.shape[1]-49) * b, 50, 5), f\"Problem with shape in batch {i+1}: {reads_50mer.shape}\"\n",
    "\n",
    "    tprint(f'  Starting prediction for {b:,} kmer reads ...')\n",
    "    label_probs, position_probs = model.predict(reads_50mer)\n",
    "\n",
    "    tprint('  Reshaping predictions ...')\n",
    "    label_probs_kmer = tf.reshape(label_probs, shape=(b,nb_50mer,-1))\n",
    "    position_probs_kmer = tf.reshape(position_probs, shape=(b,nb_50mer,-1))\n",
    "    # successive_preds = count_successive_label_preds(label_probs_kmer)\n",
    "\n",
    "    tprint('  Combining predictions ...')\n",
    "    combined_predictions = tf.map_fn(\n",
    "        fn=combine_prediction_batch,\n",
    "        elems=[label_probs_kmer, position_probs_kmer], \n",
    "        fn_output_signature=tf.int64\n",
    "        )\n",
    "\n",
    "    label_predictions = combined_predictions[:,0]\n",
    "    position_predictions = combined_predictions[:,1]\n",
    "    top_preds = top_predictions(label_probs_kmer, n=top_n)\n",
    "\n",
    "    # Add results for batch\n",
    "    tprint('  Preparing prediction report ...')\n",
    "    # 'readid', 'refseqid', 'refsource', 'refseq_strand', 'taxonomyid',\n",
    "    # 'lbl_true', 'lbl_pred', 'pos_true', 'pos_pred'\n",
    "    # 'top_5_lbl_pred_0', 'top_5_lbl_pred_1', 'top_5_lbl_pred_2',\n",
    "    # 'top_5_lbl_pred_3', 'top_5_lbl_pred_4'\n",
    "    preds_report = np.concatenate(\n",
    "        [\n",
    "            np.expand_dims(np.array(metadata_batch['readid']), axis=1),         # readid \n",
    "            np.expand_dims(np.array(metadata_batch['refseqid']), axis=1),       # refseqid\n",
    "            np.expand_dims(np.array(metadata_batch['refsource']), axis=1),      # refsource\n",
    "            np.expand_dims(np.array(metadata_batch['refseq_strand']), axis=1),  # refseq_strand\n",
    "            np.expand_dims(np.array(metadata_batch['reftaxonomyid']), axis=1),  # taxonomyid\n",
    "            np.expand_dims(np.array([true_label]*b), axis=1),                   # lbl_true\n",
    "            np.expand_dims(label_predictions, axis=1),                          # lbl_pred\n",
    "            np.expand_dims(np.array(metadata_batch['aln_start_pos']), axis=1),  # pos_true\n",
    "            np.expand_dims(position_predictions, axis=1),                       # pos_pred\n",
    "            top_preds[:, ::-1],                                                 # top_5_lbl_pred_0, top_5_lbl_pred_1, top_5_lbl_pred_2, top_5_lbl_pred_3, top_5_lbl_pred_4\n",
    "        ],\n",
    "        axis=1\n",
    "    )\n",
    "\n",
    "    df_preds = pd.DataFrame(\n",
    "        data=preds_report, \n",
    "        columns=pred_cols_str + pred_cols_int + top_pred_cols\n",
    "        )\n",
    "    tprint('  Saving batch prediction report to db...')\n",
    "    df_preds.to_sql('predictions', conn, if_exists='append', index=False)\n",
    "\n",
    "    tprint('  Preparing label probabilities report ...')\n",
    "    df_probs = None\n",
    "    for read_50mer_nb in range(nb_50mer):\n",
    "        probs_report = np.concatenate(\n",
    "            [\n",
    "                np.expand_dims(np.array(metadata_batch['readid']), axis=1),        # readid \n",
    "                np.expand_dims(np.array([read_50mer_nb]*b), axis=1),               # read_50mer_nb\n",
    "                label_probs_kmer[:, read_50mer_nb, :]\n",
    "            ],\n",
    "            axis=1\n",
    "        )\n",
    "\n",
    "        df = pd.DataFrame(\n",
    "            data=probs_report, \n",
    "            columns=['read_kmer_id', 'read_50mer_nb'] + prob_cols\n",
    "            )\n",
    "        df_probs = df if df_probs is None else pd.concat([df_probs, df], axis=0)\n",
    "\n",
    "    tprint('  Saving batch label probabilities report to db...')\n",
    "    df_probs.to_sql('label_probabilities', conn, if_exists='append', index=False)\n",
    "\n",
    "    tprint(f\"  Batch processing time: {(datetime.now() - loop_start).total_seconds():.2f} sec\")\n",
    "    if not run_all_batches and i+1 >= nb_batches_to_run: \n",
    "        print('Stopping')\n",
    "        break\n",
    "\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>readid</th>\n",
       "      <th>refseqid</th>\n",
       "      <th>refsource</th>\n",
       "      <th>refseq_strand</th>\n",
       "      <th>taxonomyid</th>\n",
       "      <th>lbl_true</th>\n",
       "      <th>lbl_pred</th>\n",
       "      <th>pos_true</th>\n",
       "      <th>pos_pred</th>\n",
       "      <th>top_5_lbl_pred_0</th>\n",
       "      <th>top_5_lbl_pred_1</th>\n",
       "      <th>top_5_lbl_pred_2</th>\n",
       "      <th>top_5_lbl_pred_3</th>\n",
       "      <th>top_5_lbl_pred_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>11089:ncbi:1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>-</td>\n",
       "      <td>11089</td>\n",
       "      <td>118</td>\n",
       "      <td>32</td>\n",
       "      <td>3164</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>117</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id              readid      refseqid refsource refseq_strand taxonomyid  \\\n",
       "0   1  11089:ncbi:1-17000  11089:ncbi:1      ncbi             -      11089   \n",
       "\n",
       "   lbl_true  lbl_pred  pos_true  pos_pred  top_5_lbl_pred_0  top_5_lbl_pred_1  \\\n",
       "0       118        32      3164         7                18                 0   \n",
       "\n",
       "   top_5_lbl_pred_2  top_5_lbl_pred_3  top_5_lbl_pred_4  \n",
       "0                32               117               118  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>read_kmer_id</th>\n",
       "      <th>read_50mer_nb</th>\n",
       "      <th>prob_000</th>\n",
       "      <th>prob_001</th>\n",
       "      <th>prob_002</th>\n",
       "      <th>prob_003</th>\n",
       "      <th>prob_004</th>\n",
       "      <th>prob_005</th>\n",
       "      <th>prob_006</th>\n",
       "      <th>...</th>\n",
       "      <th>prob_177</th>\n",
       "      <th>prob_178</th>\n",
       "      <th>prob_179</th>\n",
       "      <th>prob_180</th>\n",
       "      <th>prob_181</th>\n",
       "      <th>prob_182</th>\n",
       "      <th>prob_183</th>\n",
       "      <th>prob_184</th>\n",
       "      <th>prob_185</th>\n",
       "      <th>prob_186</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0779117e-09</td>\n",
       "      <td>4.6550543e-16</td>\n",
       "      <td>4.0059014e-09</td>\n",
       "      <td>5.709354e-14</td>\n",
       "      <td>3.985621e-13</td>\n",
       "      <td>1.4781354e-07</td>\n",
       "      <td>3.6368833e-13</td>\n",
       "      <td>...</td>\n",
       "      <td>6.536237e-14</td>\n",
       "      <td>1.3692335e-18</td>\n",
       "      <td>4.8284808e-14</td>\n",
       "      <td>4.28663e-27</td>\n",
       "      <td>1.7636006e-08</td>\n",
       "      <td>2.0901565e-20</td>\n",
       "      <td>1.2649406e-15</td>\n",
       "      <td>3.2826548e-25</td>\n",
       "      <td>1.3461108e-31</td>\n",
       "      <td>1.695962e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>33</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>1</td>\n",
       "      <td>7.798936e-12</td>\n",
       "      <td>3.0194006e-16</td>\n",
       "      <td>9.897284e-13</td>\n",
       "      <td>4.3037067e-11</td>\n",
       "      <td>8.870653e-16</td>\n",
       "      <td>5.2778373e-09</td>\n",
       "      <td>5.065814e-15</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4544616e-15</td>\n",
       "      <td>1.7843624e-16</td>\n",
       "      <td>9.904177e-13</td>\n",
       "      <td>3.457488e-21</td>\n",
       "      <td>7.375788e-07</td>\n",
       "      <td>1.7063393e-21</td>\n",
       "      <td>6.5437836e-11</td>\n",
       "      <td>5.6732623e-25</td>\n",
       "      <td>4.585938e-27</td>\n",
       "      <td>1.069439e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>65</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0980089e-12</td>\n",
       "      <td>1.9275111e-15</td>\n",
       "      <td>1.2620867e-10</td>\n",
       "      <td>5.787828e-11</td>\n",
       "      <td>5.338331e-13</td>\n",
       "      <td>2.7942612e-11</td>\n",
       "      <td>5.7700195e-10</td>\n",
       "      <td>...</td>\n",
       "      <td>5.753128e-14</td>\n",
       "      <td>1.6327466e-15</td>\n",
       "      <td>4.4719354e-06</td>\n",
       "      <td>6.113753e-17</td>\n",
       "      <td>4.3007142e-07</td>\n",
       "      <td>1.6321036e-17</td>\n",
       "      <td>1.3829523e-09</td>\n",
       "      <td>7.2506775e-19</td>\n",
       "      <td>4.256189e-23</td>\n",
       "      <td>8.687584e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>97</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>3</td>\n",
       "      <td>4.4603436e-09</td>\n",
       "      <td>4.5842836e-14</td>\n",
       "      <td>3.9782825e-12</td>\n",
       "      <td>1.9334909e-08</td>\n",
       "      <td>1.5190636e-10</td>\n",
       "      <td>2.5700672e-08</td>\n",
       "      <td>1.488595e-13</td>\n",
       "      <td>...</td>\n",
       "      <td>2.6302808e-07</td>\n",
       "      <td>7.203852e-11</td>\n",
       "      <td>4.3689414e-07</td>\n",
       "      <td>2.1555071e-17</td>\n",
       "      <td>0.000116948606</td>\n",
       "      <td>5.1792988e-18</td>\n",
       "      <td>1.0885436e-09</td>\n",
       "      <td>3.184696e-18</td>\n",
       "      <td>1.5609967e-19</td>\n",
       "      <td>3.963395e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>129</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>4</td>\n",
       "      <td>8.554256e-06</td>\n",
       "      <td>2.554389e-13</td>\n",
       "      <td>0.0006343455</td>\n",
       "      <td>3.5571188e-07</td>\n",
       "      <td>2.3673221e-09</td>\n",
       "      <td>0.097307</td>\n",
       "      <td>7.587622e-09</td>\n",
       "      <td>...</td>\n",
       "      <td>3.750018e-07</td>\n",
       "      <td>8.665608e-11</td>\n",
       "      <td>6.433393e-09</td>\n",
       "      <td>5.5440495e-20</td>\n",
       "      <td>0.00476838</td>\n",
       "      <td>1.4658719e-17</td>\n",
       "      <td>5.7866656e-13</td>\n",
       "      <td>1.6711752e-17</td>\n",
       "      <td>5.3012568e-21</td>\n",
       "      <td>1.876316e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>3073</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>96</td>\n",
       "      <td>0.9953505</td>\n",
       "      <td>5.3798974e-20</td>\n",
       "      <td>1.0473234e-18</td>\n",
       "      <td>6.393374e-11</td>\n",
       "      <td>2.8581829e-10</td>\n",
       "      <td>3.2485372e-15</td>\n",
       "      <td>2.7151905e-12</td>\n",
       "      <td>...</td>\n",
       "      <td>5.3093893e-15</td>\n",
       "      <td>1.3183516e-18</td>\n",
       "      <td>3.1293095e-09</td>\n",
       "      <td>2.2244292e-16</td>\n",
       "      <td>1.6725651e-10</td>\n",
       "      <td>3.994397e-15</td>\n",
       "      <td>2.2199666e-19</td>\n",
       "      <td>1.2594476e-13</td>\n",
       "      <td>3.5062112e-13</td>\n",
       "      <td>1.599368e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>3105</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>97</td>\n",
       "      <td>0.7476607</td>\n",
       "      <td>8.0384427e-16</td>\n",
       "      <td>3.6662344e-12</td>\n",
       "      <td>1.3936913e-13</td>\n",
       "      <td>3.1677278e-12</td>\n",
       "      <td>3.3102418e-15</td>\n",
       "      <td>1.0410654e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8784293e-12</td>\n",
       "      <td>3.711314e-15</td>\n",
       "      <td>1.8920882e-05</td>\n",
       "      <td>3.1332e-16</td>\n",
       "      <td>4.1800877e-07</td>\n",
       "      <td>5.0522433e-12</td>\n",
       "      <td>1.0039908e-14</td>\n",
       "      <td>6.3892506e-13</td>\n",
       "      <td>5.0943116e-14</td>\n",
       "      <td>1.197173e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>3137</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>98</td>\n",
       "      <td>0.9361404</td>\n",
       "      <td>2.664984e-16</td>\n",
       "      <td>7.8721495e-17</td>\n",
       "      <td>3.706427e-12</td>\n",
       "      <td>1.3358595e-14</td>\n",
       "      <td>1.254372e-16</td>\n",
       "      <td>1.8388244e-10</td>\n",
       "      <td>...</td>\n",
       "      <td>1.06356905e-16</td>\n",
       "      <td>4.1285527e-16</td>\n",
       "      <td>1.4968358e-11</td>\n",
       "      <td>2.585157e-18</td>\n",
       "      <td>8.469594e-08</td>\n",
       "      <td>3.6794658e-11</td>\n",
       "      <td>2.3980224e-18</td>\n",
       "      <td>5.1243676e-15</td>\n",
       "      <td>1.01574735e-17</td>\n",
       "      <td>2.688872e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>3169</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>99</td>\n",
       "      <td>0.998154</td>\n",
       "      <td>1.2311882e-15</td>\n",
       "      <td>1.7796784e-17</td>\n",
       "      <td>5.404334e-11</td>\n",
       "      <td>1.2882388e-10</td>\n",
       "      <td>1.5932526e-10</td>\n",
       "      <td>8.845724e-11</td>\n",
       "      <td>...</td>\n",
       "      <td>1.8041767e-10</td>\n",
       "      <td>3.959757e-12</td>\n",
       "      <td>3.682707e-12</td>\n",
       "      <td>2.1554382e-12</td>\n",
       "      <td>4.0533246e-06</td>\n",
       "      <td>5.84147e-08</td>\n",
       "      <td>6.353569e-15</td>\n",
       "      <td>1.6207057e-10</td>\n",
       "      <td>1.1721044e-15</td>\n",
       "      <td>3.890514e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>3201</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>100</td>\n",
       "      <td>0.22425258</td>\n",
       "      <td>4.8676975e-13</td>\n",
       "      <td>3.5484584e-15</td>\n",
       "      <td>6.286311e-08</td>\n",
       "      <td>1.3584297e-11</td>\n",
       "      <td>9.7854135e-11</td>\n",
       "      <td>2.0057087e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>2.981896e-11</td>\n",
       "      <td>1.4885541e-12</td>\n",
       "      <td>0.601138</td>\n",
       "      <td>5.832164e-17</td>\n",
       "      <td>3.1379425e-08</td>\n",
       "      <td>1.38640115e-11</td>\n",
       "      <td>7.054615e-14</td>\n",
       "      <td>4.1462553e-09</td>\n",
       "      <td>5.9365905e-15</td>\n",
       "      <td>3.292315e-18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows × 190 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       id        read_kmer_id  read_50mer_nb       prob_000       prob_001  \\\n",
       "0       1  11089:ncbi:1-17000              0  3.0779117e-09  4.6550543e-16   \n",
       "1      33  11089:ncbi:1-17000              1   7.798936e-12  3.0194006e-16   \n",
       "2      65  11089:ncbi:1-17000              2  1.0980089e-12  1.9275111e-15   \n",
       "3      97  11089:ncbi:1-17000              3  4.4603436e-09  4.5842836e-14   \n",
       "4     129  11089:ncbi:1-17000              4   8.554256e-06   2.554389e-13   \n",
       "..    ...                 ...            ...            ...            ...   \n",
       "96   3073  11089:ncbi:1-17000             96      0.9953505  5.3798974e-20   \n",
       "97   3105  11089:ncbi:1-17000             97      0.7476607  8.0384427e-16   \n",
       "98   3137  11089:ncbi:1-17000             98      0.9361404   2.664984e-16   \n",
       "99   3169  11089:ncbi:1-17000             99       0.998154  1.2311882e-15   \n",
       "100  3201  11089:ncbi:1-17000            100     0.22425258  4.8676975e-13   \n",
       "\n",
       "          prob_002       prob_003       prob_004       prob_005  \\\n",
       "0    4.0059014e-09   5.709354e-14   3.985621e-13  1.4781354e-07   \n",
       "1     9.897284e-13  4.3037067e-11   8.870653e-16  5.2778373e-09   \n",
       "2    1.2620867e-10   5.787828e-11   5.338331e-13  2.7942612e-11   \n",
       "3    3.9782825e-12  1.9334909e-08  1.5190636e-10  2.5700672e-08   \n",
       "4     0.0006343455  3.5571188e-07  2.3673221e-09       0.097307   \n",
       "..             ...            ...            ...            ...   \n",
       "96   1.0473234e-18   6.393374e-11  2.8581829e-10  3.2485372e-15   \n",
       "97   3.6662344e-12  1.3936913e-13  3.1677278e-12  3.3102418e-15   \n",
       "98   7.8721495e-17   3.706427e-12  1.3358595e-14   1.254372e-16   \n",
       "99   1.7796784e-17   5.404334e-11  1.2882388e-10  1.5932526e-10   \n",
       "100  3.5484584e-15   6.286311e-08  1.3584297e-11  9.7854135e-11   \n",
       "\n",
       "          prob_006  ...        prob_177       prob_178       prob_179  \\\n",
       "0    3.6368833e-13  ...    6.536237e-14  1.3692335e-18  4.8284808e-14   \n",
       "1     5.065814e-15  ...   1.4544616e-15  1.7843624e-16   9.904177e-13   \n",
       "2    5.7700195e-10  ...    5.753128e-14  1.6327466e-15  4.4719354e-06   \n",
       "3     1.488595e-13  ...   2.6302808e-07   7.203852e-11  4.3689414e-07   \n",
       "4     7.587622e-09  ...    3.750018e-07   8.665608e-11   6.433393e-09   \n",
       "..             ...  ...             ...            ...            ...   \n",
       "96   2.7151905e-12  ...   5.3093893e-15  1.3183516e-18  3.1293095e-09   \n",
       "97   1.0410654e-08  ...   2.8784293e-12   3.711314e-15  1.8920882e-05   \n",
       "98   1.8388244e-10  ...  1.06356905e-16  4.1285527e-16  1.4968358e-11   \n",
       "99    8.845724e-11  ...   1.8041767e-10   3.959757e-12   3.682707e-12   \n",
       "100  2.0057087e-06  ...    2.981896e-11  1.4885541e-12       0.601138   \n",
       "\n",
       "          prob_180        prob_181        prob_182       prob_183  \\\n",
       "0      4.28663e-27   1.7636006e-08   2.0901565e-20  1.2649406e-15   \n",
       "1     3.457488e-21    7.375788e-07   1.7063393e-21  6.5437836e-11   \n",
       "2     6.113753e-17   4.3007142e-07   1.6321036e-17  1.3829523e-09   \n",
       "3    2.1555071e-17  0.000116948606   5.1792988e-18  1.0885436e-09   \n",
       "4    5.5440495e-20      0.00476838   1.4658719e-17  5.7866656e-13   \n",
       "..             ...             ...             ...            ...   \n",
       "96   2.2244292e-16   1.6725651e-10    3.994397e-15  2.2199666e-19   \n",
       "97      3.1332e-16   4.1800877e-07   5.0522433e-12  1.0039908e-14   \n",
       "98    2.585157e-18    8.469594e-08   3.6794658e-11  2.3980224e-18   \n",
       "99   2.1554382e-12   4.0533246e-06     5.84147e-08   6.353569e-15   \n",
       "100   5.832164e-17   3.1379425e-08  1.38640115e-11   7.054615e-14   \n",
       "\n",
       "          prob_184        prob_185      prob_186  \n",
       "0    3.2826548e-25   1.3461108e-31  1.695962e-21  \n",
       "1    5.6732623e-25    4.585938e-27  1.069439e-21  \n",
       "2    7.2506775e-19    4.256189e-23  8.687584e-22  \n",
       "3     3.184696e-18   1.5609967e-19  3.963395e-22  \n",
       "4    1.6711752e-17   5.3012568e-21  1.876316e-18  \n",
       "..             ...             ...           ...  \n",
       "96   1.2594476e-13   3.5062112e-13  1.599368e-17  \n",
       "97   6.3892506e-13   5.0943116e-14  1.197173e-14  \n",
       "98   5.1243676e-15  1.01574735e-17  2.688872e-14  \n",
       "99   1.6207057e-10   1.1721044e-15  3.890514e-13  \n",
       "100  4.1462553e-09   5.9365905e-15  3.292315e-18  \n",
       "\n",
       "[101 rows x 190 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with open_db(p2db=p2db, k=150) as conn:\n",
    "    display(pd.read_sql_query(\"SELECT * FROM predictions WHERE readid = '11089:ncbi:1-17000';\", conn))\n",
    "    display(pd.read_sql_query(\"SELECT * FROM label_probabilities WHERE read_kmer_id = '11089:ncbi:1-17000';\", conn))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(0, 1, 'readid'), (1, 2, 'refseqid'), (2, 8, 'pos_true')]\n",
      "[(0, 1, 'read_kmer_id'), (1, 2, 'read_50mer_nb')]\n"
     ]
    }
   ],
   "source": [
    "with open_db(p2db, 150) as conn:\n",
    "    cursor = conn.cursor()\n",
    "    # query=\"PRAGMA index_list('predictions');\"\n",
    "    # res = cursor.execute(query)\n",
    "    # print(res.fetchall())\n",
    "    query=\"PRAGMA index_info('idx_preds')\"\n",
    "    print(cursor.execute(query).fetchall())\n",
    "    query=\"PRAGMA index_info('idx_probs')\"\n",
    "    print(cursor.execute(query).fetchall())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inference for 25% of the simreads on 69 sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Estimate required space for full inference with probabilities is *470 Gb*. This is too large to save even on my NAS. Will first build a table with 25% of the total reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "infer_results\n",
      "  |--yf-ncbi\n",
      "  |    |--single_1seq_150bp.db (0)\n",
      "  |    |--single_69seq_150bp-bck.db (1)\n",
      "  |    |--single_69seq_150bp.db (2)\n"
     ]
    }
   ],
   "source": [
    "p2aln = pfs.data / 'ncbi/simreads/yf/single_69seq_150bp/single_69seq_150bp.aln'\n",
    "assert p2aln.exists()\n",
    "p2db = Path('/mnt/k/metagentools') / 'ncbi/infer_results/yf-ncbi' / f'{p2aln.stem}.db'\n",
    "assert p2db.parent.exists()\n",
    "files_in_tree(p2db.parent);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of reads: 1,161,033\n"
     ]
    }
   ],
   "source": [
    "aln = AlnFileReader(p2aln)\n",
    "for nb_kmer_reads, o in enumerate(aln):\n",
    "    pass\n",
    "print(f\"Total number of reads: {nb_kmer_reads:,d}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def skip_existing_predictions(gen, p2db, bs):\n",
    "    # Identify the readnb for the last saved prediction\n",
    "    print('Checking predictions already in database...')\n",
    "    with open_db(p2db=p2db, k=150) as conn:\n",
    "        last_predictions_id = conn.execute(\"SELECT MAX(id) FROM predictions\").fetchone()[0]\n",
    "        nb_predictions = conn.execute(\"SELECT COUNT(*) FROM predictions\").fetchone()[0]\n",
    "        last_readid = conn.execute(f\"SELECT readid FROM predictions WHERE id = {last_predictions_id};\").fetchone()[0]\n",
    "        # regex = re.compile(r'\\d*:ncbi:(?P<kmer_read>\\d-\\d*)')\n",
    "        # match = regex.search(last_readid)\n",
    "        # kmer_read = match.group('read_nb')\n",
    "\n",
    "    print(f\"Database includes {nb_predictions:,d} predictions, corresponding to {nb_predictions//bs:,d} batches\")\n",
    "    print(f\"Last prediction id: {last_predictions_id:,d} for kmer read '{last_readid}'\")\n",
    "    print(f\"Skipping already processed predictions ...\")\n",
    "\n",
    "    for i, (batch_meta, batch_reads) in enumerate(gen):\n",
    "        if i%100 == 0: \n",
    "            print(f\"   Skipped first {(i+1)*bs:,d} kmer reads ({i+1:,d} batches)\")\n",
    " \n",
    "        if last_readid in batch_meta['readid']:\n",
    "            print(f\"   Reached last batch of saved prediction (batch {i+1:,d})\")\n",
    "            print('Can procees with normal prediction inference')\n",
    "            nb_kmer_reads_skipped = nb_predictions\n",
    "            nb_batches_skipped = i+1\n",
    "            break\n",
    "    return nb_batches_skipped, nb_kmer_reads_skipped\n",
    "\n",
    "# aln.reset_iterator()\n",
    "# gen2 = aln.cnn_virus_input_generator(bs=512, label=118)\n",
    "# nbs, nrs = skip_existing_predictions(gen=gen2, p2db=p2db, bs=512)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b = 512   # number of k-mer in a batch\n",
    "k = 150\n",
    "true_label = 118\n",
    "top_n = 5   # n for top-n prediction to keep\n",
    "run_all_batches = False\n",
    "nb_batches_to_run = int(nb_kmer_reads / b * 0.25)\n",
    "\n",
    "uid = datetime.today().strftime('%Y-%m-%d_%H_%M_%S')\n",
    "nb_50mer = k - 49\n",
    "\n",
    "print(f\"Run prediction loop with the following parameters:\")\n",
    "print(f\"   {b} k-mer per batch; {k} bp per sequence; keep top-{top_n} predictions\")\n",
    "\n",
    "aln.reset_iterator()\n",
    "\n",
    "model = create_model_original(path2parameters=p2model)\n",
    "print(f\"Model loaded and ready to run ...\")\n",
    "\n",
    "# open connection to sqlite db\n",
    "conn = open_db(p2db, k=k)\n",
    "cursor = conn.cursor()\n",
    "\n",
    "# create a dataframe to store results\n",
    "pred_cols_str = 'readid refseqid refsource refseq_strand taxonomyid'.split(' ')\n",
    "pred_cols_int = 'lbl_true lbl_pred pos_true pos_pred'.split(' ')\n",
    "top_pred_cols = [f\"top_{top_n}_lbl_pred_{i}\" for i in range(top_n)]\n",
    "prob_cols = [f\"prob_{i:03d}\" for i in range(187)]\n",
    "\n",
    "\n",
    "def tprint(string):\n",
    "    print(f\"{datetime.now().strftime('%H:%M:%S')}    {string}\")\n",
    "\n",
    "print(f\"Starting prediction loop ...\")\n",
    "gen = aln.cnn_virus_input_generator(bs=b, label=true_label)\n",
    "\n",
    "# Skip kmer reads that are already processed\n",
    "nb_batches_skipped, nb_kmer_reads_skipped = skip_existing_predictions(gen=gen, p2db=p2db, bs=b)\n",
    "\n",
    "# Proceed with prediction inference \n",
    "for i,(metadata_batch, reads_batch) in enumerate(gen):\n",
    "    loop_start = datetime.now()\n",
    "    tprint(f\"Batch {i+1:3,d} (aln batch {nb_batches_skipped+i+1:3,d}) ...\")\n",
    "\n",
    "    reads_kmer, (labels_true, position_true) = string_input_batch_to_tensors(reads_batch, k=k)\n",
    "    reads_50mer = split_kmer_batch_into_50mers(reads_kmer)\n",
    "    assert reads_50mer.shape == ((reads_kmer.shape[1]-49) * b, 50, 5), f\"Problem with shape in batch {i+1}: {reads_50mer.shape}\"\n",
    "\n",
    "    tprint(f'  Starting prediction for {b:,} kmer reads ...')\n",
    "    label_probs, position_probs = model.predict(reads_50mer)\n",
    "\n",
    "    tprint('  Reshaping predictions ...')\n",
    "    label_probs_kmer = tf.reshape(label_probs, shape=(b,nb_50mer,-1))\n",
    "    position_probs_kmer = tf.reshape(position_probs, shape=(b,nb_50mer,-1))\n",
    "\n",
    "    tprint('  Combining predictions ...')\n",
    "    combined_predictions = tf.map_fn(\n",
    "        fn=combine_prediction_batch,\n",
    "        elems=[label_probs_kmer, position_probs_kmer], \n",
    "        fn_output_signature=tf.int64\n",
    "        )\n",
    "\n",
    "    label_predictions = combined_predictions[:,0]\n",
    "    position_predictions = combined_predictions[:,1]\n",
    "    top_preds = top_predictions(label_probs_kmer, n=top_n)\n",
    "\n",
    "    # Add results for current batch\n",
    "    tprint('  Preparing prediction report ...')\n",
    "    preds_report = np.concatenate(\n",
    "        [\n",
    "            np.expand_dims(np.array(metadata_batch['readid']), axis=1),         # readid \n",
    "            np.expand_dims(np.array(metadata_batch['refseqid']), axis=1),       # refseqid\n",
    "            np.expand_dims(np.array(metadata_batch['refsource']), axis=1),      # refsource\n",
    "            np.expand_dims(np.array(metadata_batch['refseq_strand']), axis=1),  # refseq_strand\n",
    "            np.expand_dims(np.array(metadata_batch['reftaxonomyid']), axis=1),  # taxonomyid\n",
    "            np.expand_dims(np.array([true_label]*b), axis=1),                   # lbl_true\n",
    "            np.expand_dims(label_predictions, axis=1),                          # lbl_pred\n",
    "            np.expand_dims(np.array(metadata_batch['aln_start_pos']), axis=1),  # pos_true\n",
    "            np.expand_dims(position_predictions, axis=1),                       # pos_pred\n",
    "            top_preds[:, ::-1],                                                 # top_5_lbl_pred_0, top_5_lbl_pred_1, top_5_lbl_pred_2, top_5_lbl_pred_3, top_5_lbl_pred_4\n",
    "        ],\n",
    "        axis=1\n",
    "    )\n",
    "\n",
    "    df_preds = pd.DataFrame(\n",
    "        data=preds_report, \n",
    "        columns=pred_cols_str + pred_cols_int + top_pred_cols\n",
    "        )\n",
    "    tprint('  Saving batch prediction report to db...')\n",
    "    df_preds.to_sql('predictions', conn, if_exists='append', index=False)\n",
    "\n",
    "    tprint('  Preparing label probabilities report ...')\n",
    "    df_probs = None\n",
    "    for read_50mer_nb in range(nb_50mer):\n",
    "        probs_report = np.concatenate(\n",
    "            [\n",
    "                np.expand_dims(np.array(metadata_batch['readid']), axis=1),        # readid \n",
    "                np.expand_dims(np.array([read_50mer_nb]*b), axis=1),               # read_50mer_nb\n",
    "                label_probs_kmer[:, read_50mer_nb, :]\n",
    "            ],\n",
    "            axis=1\n",
    "        )\n",
    "\n",
    "        df = pd.DataFrame(\n",
    "            data=probs_report, \n",
    "            columns=['read_kmer_id', 'read_50mer_nb'] + prob_cols\n",
    "            )\n",
    "        df_probs = df if df_probs is None else pd.concat([df_probs, df], axis=0)\n",
    "\n",
    "    tprint('  Saving batch label probabilities report to db...')\n",
    "    df_probs.to_sql('label_probabilities', conn, if_exists='append', index=False)\n",
    "\n",
    "    tprint(f\"  Batch processing time: {(datetime.now() - loop_start).total_seconds():.2f} sec\")\n",
    "    if not run_all_batches and i+1 >= nb_batches_to_run: \n",
    "        print('Stopping')\n",
    "        break\n",
    "\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open_db(p2db=p2db, k=150) as conn:\n",
    "#     cursor=conn.cursor()\n",
    "#     display(table_columns(cursor=cursor, table='predictions'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>readid</th>\n",
       "      <th>refseqid</th>\n",
       "      <th>refsource</th>\n",
       "      <th>refseq_strand</th>\n",
       "      <th>taxonomyid</th>\n",
       "      <th>lbl_true</th>\n",
       "      <th>lbl_pred</th>\n",
       "      <th>pos_true</th>\n",
       "      <th>pos_pred</th>\n",
       "      <th>top_5_lbl_pred_0</th>\n",
       "      <th>top_5_lbl_pred_1</th>\n",
       "      <th>top_5_lbl_pred_2</th>\n",
       "      <th>top_5_lbl_pred_3</th>\n",
       "      <th>top_5_lbl_pred_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>11089:ncbi:1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>-</td>\n",
       "      <td>11089</td>\n",
       "      <td>118</td>\n",
       "      <td>10</td>\n",
       "      <td>7804</td>\n",
       "      <td>0</td>\n",
       "      <td>118</td>\n",
       "      <td>32</td>\n",
       "      <td>10</td>\n",
       "      <td>62</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id              readid      refseqid refsource refseq_strand taxonomyid  \\\n",
       "0   1  11089:ncbi:1-17000  11089:ncbi:1      ncbi             -      11089   \n",
       "\n",
       "   lbl_true  lbl_pred  pos_true  pos_pred  top_5_lbl_pred_0  top_5_lbl_pred_1  \\\n",
       "0       118        10      7804         0               118                32   \n",
       "\n",
       "   top_5_lbl_pred_2  top_5_lbl_pred_3  top_5_lbl_pred_4  \n",
       "0                10                62                94  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>read_kmer_id</th>\n",
       "      <th>read_50mer_nb</th>\n",
       "      <th>prob_000</th>\n",
       "      <th>prob_001</th>\n",
       "      <th>prob_002</th>\n",
       "      <th>prob_003</th>\n",
       "      <th>prob_004</th>\n",
       "      <th>prob_005</th>\n",
       "      <th>prob_006</th>\n",
       "      <th>...</th>\n",
       "      <th>prob_177</th>\n",
       "      <th>prob_178</th>\n",
       "      <th>prob_179</th>\n",
       "      <th>prob_180</th>\n",
       "      <th>prob_181</th>\n",
       "      <th>prob_182</th>\n",
       "      <th>prob_183</th>\n",
       "      <th>prob_184</th>\n",
       "      <th>prob_185</th>\n",
       "      <th>prob_186</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>0</td>\n",
       "      <td>9.452837e-16</td>\n",
       "      <td>1.8033755e-24</td>\n",
       "      <td>3.709243e-18</td>\n",
       "      <td>3.2934015e-16</td>\n",
       "      <td>1.11393635e-13</td>\n",
       "      <td>4.909559e-09</td>\n",
       "      <td>3.2027174e-19</td>\n",
       "      <td>...</td>\n",
       "      <td>4.620669e-10</td>\n",
       "      <td>2.1956457e-23</td>\n",
       "      <td>4.973716e-17</td>\n",
       "      <td>1.2870734e-19</td>\n",
       "      <td>6.685301e-13</td>\n",
       "      <td>2.7962992e-13</td>\n",
       "      <td>5.0206733e-22</td>\n",
       "      <td>2.3932547e-15</td>\n",
       "      <td>1.2987661e-21</td>\n",
       "      <td>3.557021e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>513</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>1</td>\n",
       "      <td>9.987287e-13</td>\n",
       "      <td>9.640493e-23</td>\n",
       "      <td>6.864665e-13</td>\n",
       "      <td>5.297059e-17</td>\n",
       "      <td>5.2123617e-09</td>\n",
       "      <td>1.9526384e-07</td>\n",
       "      <td>6.811955e-14</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00029164975</td>\n",
       "      <td>9.457506e-20</td>\n",
       "      <td>4.247617e-15</td>\n",
       "      <td>2.2502056e-20</td>\n",
       "      <td>1.3939608e-09</td>\n",
       "      <td>4.1289537e-12</td>\n",
       "      <td>1.0989764e-16</td>\n",
       "      <td>5.3598934e-11</td>\n",
       "      <td>1.09030056e-14</td>\n",
       "      <td>1.850025e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1025</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>2</td>\n",
       "      <td>4.1150392e-14</td>\n",
       "      <td>1.0409154e-21</td>\n",
       "      <td>3.1731182e-18</td>\n",
       "      <td>4.837065e-13</td>\n",
       "      <td>1.8273633e-12</td>\n",
       "      <td>5.6225136e-13</td>\n",
       "      <td>1.5127502e-14</td>\n",
       "      <td>...</td>\n",
       "      <td>2.7924674e-08</td>\n",
       "      <td>9.881187e-25</td>\n",
       "      <td>1.2789581e-19</td>\n",
       "      <td>5.2265872e-23</td>\n",
       "      <td>3.6289507e-16</td>\n",
       "      <td>5.854743e-18</td>\n",
       "      <td>4.597412e-22</td>\n",
       "      <td>2.7675505e-16</td>\n",
       "      <td>3.6334886e-19</td>\n",
       "      <td>7.191011e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1537</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>3</td>\n",
       "      <td>2.4102365e-13</td>\n",
       "      <td>3.9195824e-19</td>\n",
       "      <td>4.2152547e-16</td>\n",
       "      <td>5.363886e-15</td>\n",
       "      <td>7.9354416e-13</td>\n",
       "      <td>1.9598592e-10</td>\n",
       "      <td>2.0204985e-14</td>\n",
       "      <td>...</td>\n",
       "      <td>2.3507164e-12</td>\n",
       "      <td>3.4494873e-22</td>\n",
       "      <td>1.4030163e-15</td>\n",
       "      <td>5.3813346e-20</td>\n",
       "      <td>2.1669545e-18</td>\n",
       "      <td>3.9168753e-20</td>\n",
       "      <td>4.6393943e-18</td>\n",
       "      <td>2.162604e-17</td>\n",
       "      <td>5.868014e-17</td>\n",
       "      <td>3.671587e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2049</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>4</td>\n",
       "      <td>5.402446e-15</td>\n",
       "      <td>1.3982798e-24</td>\n",
       "      <td>5.1237304e-19</td>\n",
       "      <td>2.6393469e-12</td>\n",
       "      <td>7.070969e-12</td>\n",
       "      <td>4.5762413e-08</td>\n",
       "      <td>3.941572e-13</td>\n",
       "      <td>...</td>\n",
       "      <td>2.2205168e-10</td>\n",
       "      <td>2.4770586e-20</td>\n",
       "      <td>2.2223746e-16</td>\n",
       "      <td>1.3230725e-16</td>\n",
       "      <td>1.473535e-13</td>\n",
       "      <td>1.0368611e-15</td>\n",
       "      <td>7.554584e-18</td>\n",
       "      <td>6.905381e-15</td>\n",
       "      <td>6.823121e-19</td>\n",
       "      <td>5.281248e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>49153</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>96</td>\n",
       "      <td>1.266811e-12</td>\n",
       "      <td>1.7694516e-13</td>\n",
       "      <td>6.9926794e-11</td>\n",
       "      <td>7.442386e-16</td>\n",
       "      <td>2.1827995e-12</td>\n",
       "      <td>2.2899127e-11</td>\n",
       "      <td>6.6601347e-13</td>\n",
       "      <td>...</td>\n",
       "      <td>2.1055688e-09</td>\n",
       "      <td>2.6956715e-11</td>\n",
       "      <td>4.02884e-10</td>\n",
       "      <td>3.6730546e-06</td>\n",
       "      <td>7.332998e-13</td>\n",
       "      <td>9.214267e-17</td>\n",
       "      <td>4.41918e-18</td>\n",
       "      <td>1.0378567e-09</td>\n",
       "      <td>3.9226554e-16</td>\n",
       "      <td>1.512387e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>49665</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>97</td>\n",
       "      <td>3.6751678e-11</td>\n",
       "      <td>5.2148962e-11</td>\n",
       "      <td>1.1142021e-07</td>\n",
       "      <td>1.3979555e-13</td>\n",
       "      <td>0.00023110698</td>\n",
       "      <td>1.8854205e-11</td>\n",
       "      <td>1.128952e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>3.426913e-08</td>\n",
       "      <td>3.5629816e-17</td>\n",
       "      <td>2.7815642e-11</td>\n",
       "      <td>7.9123413e-07</td>\n",
       "      <td>9.112657e-12</td>\n",
       "      <td>1.8065088e-14</td>\n",
       "      <td>2.2273976e-15</td>\n",
       "      <td>2.6934228e-07</td>\n",
       "      <td>3.2327324e-15</td>\n",
       "      <td>1.599009e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>50177</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>98</td>\n",
       "      <td>3.987377e-10</td>\n",
       "      <td>3.995657e-08</td>\n",
       "      <td>4.0260434e-06</td>\n",
       "      <td>4.5705413e-09</td>\n",
       "      <td>0.00026610115</td>\n",
       "      <td>3.9758719e-10</td>\n",
       "      <td>1.0729461e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>6.970071e-08</td>\n",
       "      <td>1.210523e-17</td>\n",
       "      <td>2.2871915e-10</td>\n",
       "      <td>3.9424535e-06</td>\n",
       "      <td>1.6976076e-09</td>\n",
       "      <td>4.1861083e-09</td>\n",
       "      <td>3.708444e-10</td>\n",
       "      <td>6.0206723e-05</td>\n",
       "      <td>4.0571067e-09</td>\n",
       "      <td>5.209376e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>50689</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>99</td>\n",
       "      <td>3.773077e-12</td>\n",
       "      <td>6.944977e-13</td>\n",
       "      <td>6.4031043e-09</td>\n",
       "      <td>3.3788874e-09</td>\n",
       "      <td>7.6050675e-08</td>\n",
       "      <td>4.657467e-08</td>\n",
       "      <td>3.358606e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0940438e-08</td>\n",
       "      <td>1.2781211e-18</td>\n",
       "      <td>2.6286571e-08</td>\n",
       "      <td>1.7933372e-08</td>\n",
       "      <td>4.844457e-09</td>\n",
       "      <td>1.0032285e-13</td>\n",
       "      <td>2.9824456e-15</td>\n",
       "      <td>3.3517054e-08</td>\n",
       "      <td>4.4160476e-16</td>\n",
       "      <td>2.173458e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>51201</td>\n",
       "      <td>11089:ncbi:1-17000</td>\n",
       "      <td>100</td>\n",
       "      <td>1.2701454e-07</td>\n",
       "      <td>2.2293822e-15</td>\n",
       "      <td>2.2755298e-12</td>\n",
       "      <td>4.4974868e-14</td>\n",
       "      <td>1.7995131e-11</td>\n",
       "      <td>7.290591e-08</td>\n",
       "      <td>8.492367e-10</td>\n",
       "      <td>...</td>\n",
       "      <td>2.5019067e-10</td>\n",
       "      <td>2.4869705e-15</td>\n",
       "      <td>9.0042565e-12</td>\n",
       "      <td>5.4628693e-08</td>\n",
       "      <td>2.1697435e-09</td>\n",
       "      <td>5.919073e-16</td>\n",
       "      <td>1.073037e-15</td>\n",
       "      <td>1.08923145e-13</td>\n",
       "      <td>1.4590386e-16</td>\n",
       "      <td>1.094292e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows × 190 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        id        read_kmer_id  read_50mer_nb       prob_000       prob_001  \\\n",
       "0        1  11089:ncbi:1-17000              0   9.452837e-16  1.8033755e-24   \n",
       "1      513  11089:ncbi:1-17000              1   9.987287e-13   9.640493e-23   \n",
       "2     1025  11089:ncbi:1-17000              2  4.1150392e-14  1.0409154e-21   \n",
       "3     1537  11089:ncbi:1-17000              3  2.4102365e-13  3.9195824e-19   \n",
       "4     2049  11089:ncbi:1-17000              4   5.402446e-15  1.3982798e-24   \n",
       "..     ...                 ...            ...            ...            ...   \n",
       "96   49153  11089:ncbi:1-17000             96   1.266811e-12  1.7694516e-13   \n",
       "97   49665  11089:ncbi:1-17000             97  3.6751678e-11  5.2148962e-11   \n",
       "98   50177  11089:ncbi:1-17000             98   3.987377e-10   3.995657e-08   \n",
       "99   50689  11089:ncbi:1-17000             99   3.773077e-12   6.944977e-13   \n",
       "100  51201  11089:ncbi:1-17000            100  1.2701454e-07  2.2293822e-15   \n",
       "\n",
       "          prob_002       prob_003        prob_004       prob_005  \\\n",
       "0     3.709243e-18  3.2934015e-16  1.11393635e-13   4.909559e-09   \n",
       "1     6.864665e-13   5.297059e-17   5.2123617e-09  1.9526384e-07   \n",
       "2    3.1731182e-18   4.837065e-13   1.8273633e-12  5.6225136e-13   \n",
       "3    4.2152547e-16   5.363886e-15   7.9354416e-13  1.9598592e-10   \n",
       "4    5.1237304e-19  2.6393469e-12    7.070969e-12  4.5762413e-08   \n",
       "..             ...            ...             ...            ...   \n",
       "96   6.9926794e-11   7.442386e-16   2.1827995e-12  2.2899127e-11   \n",
       "97   1.1142021e-07  1.3979555e-13   0.00023110698  1.8854205e-11   \n",
       "98   4.0260434e-06  4.5705413e-09   0.00026610115  3.9758719e-10   \n",
       "99   6.4031043e-09  3.3788874e-09   7.6050675e-08   4.657467e-08   \n",
       "100  2.2755298e-12  4.4974868e-14   1.7995131e-11   7.290591e-08   \n",
       "\n",
       "          prob_006  ...       prob_177       prob_178       prob_179  \\\n",
       "0    3.2027174e-19  ...   4.620669e-10  2.1956457e-23   4.973716e-17   \n",
       "1     6.811955e-14  ...  0.00029164975   9.457506e-20   4.247617e-15   \n",
       "2    1.5127502e-14  ...  2.7924674e-08   9.881187e-25  1.2789581e-19   \n",
       "3    2.0204985e-14  ...  2.3507164e-12  3.4494873e-22  1.4030163e-15   \n",
       "4     3.941572e-13  ...  2.2205168e-10  2.4770586e-20  2.2223746e-16   \n",
       "..             ...  ...            ...            ...            ...   \n",
       "96   6.6601347e-13  ...  2.1055688e-09  2.6956715e-11    4.02884e-10   \n",
       "97    1.128952e-07  ...   3.426913e-08  3.5629816e-17  2.7815642e-11   \n",
       "98   1.0729461e-07  ...   6.970071e-08   1.210523e-17  2.2871915e-10   \n",
       "99    3.358606e-07  ...  3.0940438e-08  1.2781211e-18  2.6286571e-08   \n",
       "100   8.492367e-10  ...  2.5019067e-10  2.4869705e-15  9.0042565e-12   \n",
       "\n",
       "          prob_180       prob_181       prob_182       prob_183  \\\n",
       "0    1.2870734e-19   6.685301e-13  2.7962992e-13  5.0206733e-22   \n",
       "1    2.2502056e-20  1.3939608e-09  4.1289537e-12  1.0989764e-16   \n",
       "2    5.2265872e-23  3.6289507e-16   5.854743e-18   4.597412e-22   \n",
       "3    5.3813346e-20  2.1669545e-18  3.9168753e-20  4.6393943e-18   \n",
       "4    1.3230725e-16   1.473535e-13  1.0368611e-15   7.554584e-18   \n",
       "..             ...            ...            ...            ...   \n",
       "96   3.6730546e-06   7.332998e-13   9.214267e-17    4.41918e-18   \n",
       "97   7.9123413e-07   9.112657e-12  1.8065088e-14  2.2273976e-15   \n",
       "98   3.9424535e-06  1.6976076e-09  4.1861083e-09   3.708444e-10   \n",
       "99   1.7933372e-08   4.844457e-09  1.0032285e-13  2.9824456e-15   \n",
       "100  5.4628693e-08  2.1697435e-09   5.919073e-16   1.073037e-15   \n",
       "\n",
       "           prob_184        prob_185      prob_186  \n",
       "0     2.3932547e-15   1.2987661e-21  3.557021e-18  \n",
       "1     5.3598934e-11  1.09030056e-14  1.850025e-18  \n",
       "2     2.7675505e-16   3.6334886e-19  7.191011e-22  \n",
       "3      2.162604e-17    5.868014e-17  3.671587e-12  \n",
       "4      6.905381e-15    6.823121e-19  5.281248e-14  \n",
       "..              ...             ...           ...  \n",
       "96    1.0378567e-09   3.9226554e-16  1.512387e-09  \n",
       "97    2.6934228e-07   3.2327324e-15  1.599009e-12  \n",
       "98    6.0206723e-05   4.0571067e-09  5.209376e-10  \n",
       "99    3.3517054e-08   4.4160476e-16  2.173458e-09  \n",
       "100  1.08923145e-13   1.4590386e-16  1.094292e-09  \n",
       "\n",
       "[101 rows x 190 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with open_db(p2db=p2db, k=150) as conn:\n",
    "    display(pd.read_sql_query(\"SELECT * FROM predictions WHERE readid = '11089:ncbi:1-17000';\", conn))\n",
    "    display(pd.read_sql_query(\"SELECT * FROM label_probabilities WHERE read_kmer_id = '11089:ncbi:1-17000';\", conn))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13:21:22    Retrieve last prediction id and total nb predictions...\n",
      "13:21:22    388,096 (0.03 sec)\n",
      "13:21:47    388,096 (24.83 sec)\n",
      "13:21:48    388,096 (0.68 sec)\n"
     ]
    }
   ],
   "source": [
    "with open_db(p2db=p2db, k=150) as conn:\n",
    "    tprint('Retrieve last prediction id and total nb predictions...')\n",
    "    time_starts = datetime.now()\n",
    "    results = conn.execute(\"SELECT MAX(id) FROM predictions\").fetchone()\n",
    "    tprint(f\"{results[0]:,d} ({(datetime.now()-time_starts).total_seconds():.2f} sec)\")\n",
    "    time_starts = datetime.now()\n",
    "    results = conn.execute(\"SELECT COUNT(readid) FROM predictions WHERE readid IS NOT NULL;\").fetchone()\n",
    "    tprint(f\"{results[0]:,d} ({(datetime.now()-time_starts).total_seconds():.2f} sec)\")\n",
    "    time_starts = datetime.now()\n",
    "    results = conn.execute(\"SELECT COUNT(*) FROM predictions\").fetchone()\n",
    "    tprint(f\"{results[0]:,d} ({(datetime.now()-time_starts).total_seconds():.2f} sec)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(388096, '11089:ncbi:23-2905', '23-2905')"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open_db(p2db=p2db, k=150) as conn:\n",
    "    last_readid = conn.execute(\"SELECT MAX(id) FROM predictions\").fetchone()[0]\n",
    "    readid = conn.execute(f\"SELECT readid FROM predictions WHERE id = {last_readid};\").fetchone()[0]\n",
    "    regex = re.compile(r'\\d*:ncbi:(?P<read_nb>\\d*-\\d*)')\n",
    "    m = regex.search(readid)\n",
    "    read_nb = '???' if m is None else m.group('read_nb')\n",
    "\n",
    "last_readid, readid, read_nb\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Technical note to accelerate `skip_existing_predictions`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The slow step in `skip_existing_predictions` is the query to get the total number of rows in the table. We can accelerate this by maintaining an accurate row count in a separate column when inserting new rows into a SQLite table named \"predictions\" that has a primary key \"id\" and an indexed column \"readid\". \n",
    "\n",
    "This can be done with the following steps:\n",
    "\n",
    "1. **Create a trigger** that fires after an INSERT operation on the \"predictions\" table. The trigger will update the row count in a separate table or column.\n",
    "\n",
    "2. **Create a table** to store the row count, for example:\n",
    "```sql\n",
    "    CREATE TABLE table_stats (\n",
    "    id INTEGER PRIMARY KEY,\n",
    "    row_count INTEGER NOT NULL\n",
    "    );\n",
    "```\n",
    "\n",
    "3. **Create the trigger** to update the row count after an INSERT:\n",
    "```sql\n",
    "    CREATE TRIGGER update_prediction_count\n",
    "    AFTER INSERT ON predictions\n",
    "    BEGIN\n",
    "    UPDATE table_stats \n",
    "    SET row_count = row_count + 1\n",
    "    WHERE id = 1;\n",
    "    END;\n",
    "```\n",
    "\n",
    "This trigger assumes there is only one row in the \"prediction_stats\" table with an id of 1. If you want to store the count per \"readid\", you can modify the trigger to:\n",
    "\n",
    "```sql\n",
    "    CREATE TRIGGER update_prediction_count\n",
    "    AFTER INSERT ON predictions  \n",
    "    BEGIN\n",
    "    INSERT INTO table_stats (readid, row_count)\n",
    "    VALUES (NEW.readid, 1)\n",
    "    ON CONFLICT(readid) DO UPDATE SET row_count = row_count + 1;\n",
    "    END;\n",
    "```\n",
    "\n",
    "This will insert a new row for each unique \"readid\" with an initial count of 1, and update the row_count if the \"readid\" already exists.\n",
    "\n",
    "4. **Insert a row** into the \"prediction_stats\" table with an initial count:\n",
    "```sql\n",
    "    INSERT INTO table_stats (id, row_count) VALUES (1, 0);\n",
    "```\n",
    "\n",
    "Now, whenever a new row is inserted into the \"predictions\" table, the trigger will automatically update the row count in the \"prediction_stats\" table. This maintains an accurate count without needing to perform a full table scan with COUNT(*).\n",
    "\n",
    "Remember to handle deletions as well by creating a BEFORE DELETE trigger that decrements the row count accordingly.\n",
    "\n",
    "Citations:\n",
    "- 1. https://www.sqlitetutorial.net/sqlite-insert/\n",
    "- 2. https://www.sqlitetutorial.net/sqlite-count-function/\n",
    "- 3. https://stackoverflow.com/questions/55007800/dynamic-way-to-insert-data-into-sqlite-table-when-column-counts-change\n",
    "- 4. https://docs.python.org/es/3/library/sqlite3.html\n",
    "- 5. https://www.sql-easy.com/learn/sqlite-count/\n",
    "- 6. https://stackoverflow.com/questions/4474873/what-is-the-most-efficient-way-to-count-rows-in-a-table-in-sqlite\n",
    "- 7. https://sqlite.org/forum/info/57c04743e1b6aa10\n",
    "- 8. https://sqlite.org/forum/forumpost/f832398c19"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# End of Section"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.6 ('metagentools')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "d79c725d3d254ae089d5edf9eb2dce3237f80d64dd85a8bedc17bd8054b8b312"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
