<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.56">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Data preprocessing and transform functions, data reader classes, datasets for CNN Virus data.">

<title>data – metagentools</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@*/dist/embed-amd.js" crossorigin="anonymous"></script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="data – metagentools">
<meta property="og:description" content="Data preprocessing and transform functions, data reader classes, datasets for CNN Virus data.">
<meta property="og:site_name" content="metagentools">
<meta name="twitter:title" content="data – metagentools">
<meta name="twitter:description" content="Data preprocessing and transform functions, data reader classes, datasets for CNN Virus data.">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">metagentools</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./cnn_virus_architecture.html">CNN Virus</a></li><li class="breadcrumb-item"><a href="./cnn_virus_data.html">data</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">metagentools</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">General Code</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">core</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./wandb.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">wandb</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./art.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">art</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">bio</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">CNN Virus</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cnn_virus_architecture.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">architecture</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cnn_virus_data.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cnn_virus_utils.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">utils</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#data-structure-for-cnn-virus-project" id="toc-data-structure-for-cnn-virus-project" class="nav-link active" data-scroll-target="#data-structure-for-cnn-virus-project">Data structure for CNN Virus project</a>
  <ul class="collapse">
  <li><a href="#data-directory" id="toc-data-directory" class="nav-link" data-scroll-target="#data-directory">Data directory</a></li>
  <li><a href="#original-datasets" id="toc-original-datasets" class="nav-link" data-scroll-target="#original-datasets">Original datasets</a>
  <ul class="collapse">
  <li><a href="#original-data-virus-labels" id="toc-original-data-virus-labels" class="nav-link" data-scroll-target="#original-data-virus-labels">Original Data Virus Labels</a></li>
  <li><a href="#originallabels" id="toc-originallabels" class="nav-link" data-scroll-target="#originallabels">OriginalLabels</a></li>
  </ul></li>
  <li><a href="#ncbi-reference-sequences-for-simulated-reads" id="toc-ncbi-reference-sequences-for-simulated-reads" class="nav-link" data-scroll-target="#ncbi-reference-sequences-for-simulated-reads">NCBI reference sequences for simulated reads</a></li>
  <li><a href="#model-related-data" id="toc-model-related-data" class="nav-link" data-scroll-target="#model-related-data">Model related data</a></li>
  </ul></li>
  <li><a href="#parsing-sequence-files" id="toc-parsing-sequence-files" class="nav-link" data-scroll-target="#parsing-sequence-files">Parsing sequence files</a>
  <ul class="collapse">
  <li><a href="#fasta-file" id="toc-fasta-file" class="nav-link" data-scroll-target="#fasta-file">FASTA file</a>
  <ul class="collapse">
  <li><a href="#fastafilereader" id="toc-fastafilereader" class="nav-link" data-scroll-target="#fastafilereader">FastaFileReader</a></li>
  <li><a href="#fastafilereader.print_first_chunks" id="toc-fastafilereader.print_first_chunks" class="nav-link" data-scroll-target="#fastafilereader.print_first_chunks">FastaFileReader.print_first_chunks</a></li>
  <li><a href="#parsing-metadata" id="toc-parsing-metadata" class="nav-link" data-scroll-target="#parsing-metadata">Parsing metadata</a></li>
  <li><a href="#textfilebasereader.parse_text" id="toc-textfilebasereader.parse_text" class="nav-link" data-scroll-target="#textfilebasereader.parse_text">TextFileBaseReader.parse_text</a></li>
  <li><a href="#textfilebasereader.set_parsing_rules" id="toc-textfilebasereader.set_parsing_rules" class="nav-link" data-scroll-target="#textfilebasereader.set_parsing_rules">TextFileBaseReader.set_parsing_rules</a></li>
  <li><a href="#fastafilereader.parse_file" id="toc-fastafilereader.parse_file" class="nav-link" data-scroll-target="#fastafilereader.parse_file">FastaFileReader.parse_file</a></li>
  </ul></li>
  <li><a href="#fastq-file" id="toc-fastq-file" class="nav-link" data-scroll-target="#fastq-file">FASTQ file</a>
  <ul class="collapse">
  <li><a href="#fastqfilereader" id="toc-fastqfilereader" class="nav-link" data-scroll-target="#fastqfilereader">FastqFileReader</a></li>
  <li><a href="#fastqfilereader.parse_file" id="toc-fastqfilereader.parse_file" class="nav-link" data-scroll-target="#fastqfilereader.parse_file">FastqFileReader.parse_file</a></li>
  </ul></li>
  <li><a href="#aln-alignment-files" id="toc-aln-alignment-files" class="nav-link" data-scroll-target="#aln-alignment-files">ALN Alignment Files</a>
  <ul class="collapse">
  <li><a href="#alnfilereader" id="toc-alnfilereader" class="nav-link" data-scroll-target="#alnfilereader">AlnFileReader</a></li>
  <li><a href="#textfilebasereader.parse_text-1" id="toc-textfilebasereader.parse_text-1" class="nav-link" data-scroll-target="#textfilebasereader.parse_text-1">TextFileBaseReader.parse_text</a></li>
  <li><a href="#alnfilereader.parse_definition_line_with_position" id="toc-alnfilereader.parse_definition_line_with_position" class="nav-link" data-scroll-target="#alnfilereader.parse_definition_line_with_position">AlnFileReader.parse_definition_line_with_position</a></li>
  <li><a href="#alnfilereader.parse_file" id="toc-alnfilereader.parse_file" class="nav-link" data-scroll-target="#alnfilereader.parse_file">AlnFileReader.parse_file</a></li>
  <li><a href="#alnfilereader.parse_header_reference_sequences" id="toc-alnfilereader.parse_header_reference_sequences" class="nav-link" data-scroll-target="#alnfilereader.parse_header_reference_sequences">AlnFileReader.parse_header_reference_sequences</a></li>
  <li><a href="#alnfilereader.set_header_parsing_rules" id="toc-alnfilereader.set_header_parsing_rules" class="nav-link" data-scroll-target="#alnfilereader.set_header_parsing_rules">AlnFileReader.set_header_parsing_rules</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#build-datasets" id="toc-build-datasets" class="nav-link" data-scroll-target="#build-datasets">Build datasets</a>
  <ul class="collapse">
  <li><a href="#text-based-inference-datasets" id="toc-text-based-inference-datasets" class="nav-link" data-scroll-target="#text-based-inference-datasets">Text based inference datasets</a>
  <ul class="collapse">
  <li><a href="#create_infer_ds_from_fastq" id="toc-create_infer_ds_from_fastq" class="nav-link" data-scroll-target="#create_infer_ds_from_fastq">create_infer_ds_from_fastq</a></li>
  <li><a href="#strings_to_tensors" id="toc-strings_to_tensors" class="nav-link" data-scroll-target="#strings_to_tensors">strings_to_tensors</a></li>
  </ul></li>
  <li><a href="#tfrecord-based-inference-datasets" id="toc-tfrecord-based-inference-datasets" class="nav-link" data-scroll-target="#tfrecord-based-inference-datasets">TFRecord based inference datasets</a>
  <ul class="collapse">
  <li><a href="#split_kmer_into_50mers" id="toc-split_kmer_into_50mers" class="nav-link" data-scroll-target="#split_kmer_into_50mers">split_kmer_into_50mers</a></li>
  <li><a href="#base_string_kmers_to_tensors" id="toc-base_string_kmers_to_tensors" class="nav-link" data-scroll-target="#base_string_kmers_to_tensors">base_string_kmers_to_tensors</a></li>
  <li><a href="#split_kmer_batch_into_50mers" id="toc-split_kmer_batch_into_50mers" class="nav-link" data-scroll-target="#split_kmer_batch_into_50mers">split_kmer_batch_into_50mers</a></li>
  <li><a href="#tfrecord_from_fastq" id="toc-tfrecord_from_fastq" class="nav-link" data-scroll-target="#tfrecord_from_fastq">tfrecord_from_fastq</a></li>
  <li><a href="#tfrecord_from_text" id="toc-tfrecord_from_text" class="nav-link" data-scroll-target="#tfrecord_from_text">tfrecord_from_text</a></li>
  <li><a href="#get_dataset_from_tfr" id="toc-get_dataset_from_tfr" class="nav-link" data-scroll-target="#get_dataset_from_tfr">get_dataset_from_tfr</a></li>
  </ul></li>
  <li><a href="#postprocessing" id="toc-postprocessing" class="nav-link" data-scroll-target="#postprocessing">Postprocessing</a>
  <ul class="collapse">
  <li><a href="#combine_predictions" id="toc-combine_predictions" class="nav-link" data-scroll-target="#combine_predictions">combine_predictions</a></li>
  <li><a href="#combine_prediction_batch" id="toc-combine_prediction_batch" class="nav-link" data-scroll-target="#combine_prediction_batch">combine_prediction_batch</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#original-code-refactored" id="toc-original-code-refactored" class="nav-link" data-scroll-target="#original-code-refactored">Original Code (Refactored)</a>
  <ul class="collapse">
  <li><a href="#datagenerator_from_50mer" id="toc-datagenerator_from_50mer" class="nav-link" data-scroll-target="#datagenerator_from_50mer">DataGenerator_from_50mer</a></li>
  <li><a href="#get_params_150mer" id="toc-get_params_150mer" class="nav-link" data-scroll-target="#get_params_150mer">get_params_150mer</a></li>
  <li><a href="#get_params_50mer" id="toc-get_params_50mer" class="nav-link" data-scroll-target="#get_params_50mer">get_params_50mer</a></li>
  <li><a href="#get_learning_weights" id="toc-get_learning_weights" class="nav-link" data-scroll-target="#get_learning_weights">get_learning_weights</a></li>
  <li><a href="#get_kmer_from_50mer" id="toc-get_kmer_from_50mer" class="nav-link" data-scroll-target="#get_kmer_from_50mer">get_kmer_from_50mer</a></li>
  <li><a href="#get_kmer_from_150mer" id="toc-get_kmer_from_150mer" class="nav-link" data-scroll-target="#get_kmer_from_150mer">get_kmer_from_150mer</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/vtecftwy/metagentools/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./cnn_virus_architecture.html">CNN Virus</a></li><li class="breadcrumb-item"><a href="./cnn_virus_data.html">data</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">data</h1>
</div>

<div>
  <div class="description">
    Data preprocessing and transform functions, data reader classes, datasets for CNN Virus data.
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="data-structure-for-cnn-virus-project" class="level1">
<h1>Data structure for CNN Virus project</h1>
<p>There are many different types of files and datasets for this project. All data are located in directory <code>data</code>, under the project root. The following is an overview of the main types of data and in which directory they sit in the tree.</p>
<p>A description of the content of each directory can be stored in a file <code>readme.md</code> or another <code>*.md</code> file.</p>
<p>These <code>readme.md</code> files can be conveniently accessed using the <code>.readme(path)</code> method on <a href="https://vtecftwy.github.io/metagentools/core.html#projectfilesystem"><code>ProjectFileSystem</code></a>.</p>
<section id="data-directory" class="level2">
<h2 class="anchored" data-anchor-id="data-directory">Data directory</h2>
<div id="cell-6" class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>pfs <span class="op">=</span> ProjectFileSystem()</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>pfs.info()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Running linux on local computer
Device's home directory: /home/vtec
Project file structure:
 - Root ........ /home/vtec/projects/bio/metagentools 
 - Data Dir .... /home/vtec/projects/bio/metagentools/nbs-dev/data_dev 
 - Notebooks ... /home/vtec/projects/bio/metagentools/nbs</code></pre>
</div>
</div>
<p>Note that we have setup this notebook to use the development data directory <code>metagentools/nbs-dev/data_dev</code> and not the standard <code>metagentools/data</code>.</p>
<div id="cell-8" class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>pfs.readme()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="data-directory-for-metagentools-development" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="data-directory-for-metagentools-development">Data directory for <code>metagentools</code> development</h3>
<p>This directory includes all the data required to test and validate <code>metagentools</code> code.</p>
<pre class="text"><code>data_dev
 |--- CNN_Virus_data
 |     |--- 50mer_ds_100_seq
 |     |--- 150mer_ds_100_seq
 |     |--- train_short
 |     |--- val_short
 |     |--- weight_of_classes
 |--- ncbi
 |     |--- infer_results
 |     |     |--- cnn_virus
 |     |     |--- csv
 |     |     |--- xlsx
 |     |--- refsequences
 |     |     |--- cov
 |     |     |     |--- another_sequence.fa
 |     |     |     |--- cov_virus_sequences_one.fa
 |     |     |     |--- cov_virus_sequences_two.fa
 |     |     |     |--- sequences_two_no_matching_rule.fa
 |     |--- simreads
 |     |     |--- cov
 |     |     |     |--- paired_1seq_50bp
 |     |     |     |      |--- paired_1seq_50bp_1.aln
 |     |     |     |      |--- paired_1seq_50bp_1.fq
 |     |     |     |--- single_1seq_50bp
 |     |     |     |      |--- single_1seq_50bp_1.aln
 |     |     |     |      |--- single_1seq_50bp_1.fq
 |--- ....           
 |--- readme.md               </code></pre>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
</section>
<section id="original-datasets" class="level2">
<h2 class="anchored" data-anchor-id="original-datasets">Original datasets</h2>
<div id="cell-10" class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data<span class="op">/</span><span class="st">'CNN_Virus_data'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/CNN_Virus_data</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="cnn-virus-data" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="cnn-virus-data">CNN Virus data</h3>
<p>This directory includes data used to train and validate the initial CNN Virus model, as well as a few smaller datasets for experimenting.</p>
<section id="file-list-and-description" class="level4">
<h4 class="anchored" data-anchor-id="file-list-and-description">File list and description:</h4>
<section id="mer" class="level5">
<h5 class="anchored" data-anchor-id="mer">50-mer</h5>
<p>50-mer reads and their labels, in <em>text format</em> with one line per sample. Each line consists of three components, separated by tabs: the 50-mer read or sequence, the virus species label and the position label:</p>
<pre class="text"><code>'TTACNAGCTCCAGTCTAAGATTGTAACTGGCCTTTTTAAAGATTGCTCTA    94    5\n'</code></pre>
<p>Files: - <code>50mer_training</code>: dataset with 50,903,296 reads for training - <code>50mer_validating</code>: dataset with 1,000,000 reads for validation - <code>50mer_ds_100_reads</code>: small subset of 100 reads from the validating dataset for experiments</p>
</section>
<section id="mer-1" class="level5">
<h5 class="anchored" data-anchor-id="mer-1">150-mer</h5>
<p>150-mer reads and their labels in <em>text format</em> in a similar format as above:</p>
<pre class="text"><code>'TTCTTTCACCACCACAACCAGTCGGCCGTGGAGAGGCGTCGCCGCGTCTCGTTCGTCGAGGCCGATCGACTGCCGCATGAGAGCGGGTGGTATTCTTCCGAAGACGACGGAGACCGGGACGGTGATGAGGAAACTGGAGAGAGCCACAAC    6    0\n'</code></pre>
<p>Files: - <code>ICTV_150mer_benchmarking</code>: dataset with 10,0000 read - <code>150mer_ds_100_reads</code>: small subset of 100 reads from <code>ICTV_150mer_benchmarking</code></p>
</section>
<section id="longer-reads" class="level5">
<h5 class="anchored" data-anchor-id="longer-reads">Longer reads</h5>
<p>Reads of various length with no labels, in simple <em>fasta format</em>. Each read sequence is preceded by a definition line: <code>&gt; Sequence n</code>, where <code>n</code> is the sequence number.</p>
<p>Files: - <code>training_sequences_300bp.fasta</code>: dataset with 9,000 300-mer reads - <code>training_sequences_500bp.fasta</code>: dataset with 9,000 500-mer reads - <code>validation_sequences.fasta</code>: dataset with 564 reads of mixed lengths ranging from 163-mer to 497-mer</p>
</section>
<section id="other-files" class="level5">
<h5 class="anchored" data-anchor-id="other-files">Other files:</h5>
<ul>
<li><code>virus_name_mapping</code>: mapping between virus species and their numerical label</li>
<li><code>weight_of_classes</code>: weights for each virus species class in the training dataset</li>
</ul>
</section>
</section>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
<section id="original-data-virus-labels" class="level3">
<h3 class="anchored" data-anchor-id="original-data-virus-labels">Original Data Virus Labels</h3>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L41" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="originallabels" class="level3">
<h3 class="anchored" data-anchor-id="originallabels">OriginalLabels</h3>
<blockquote class="blockquote">
<pre><code> OriginalLabels (p2mapping=None)</code></pre>
</blockquote>
<p><em>Converts labels to species name for original data</em></p>
<p>Original data include 187 viruses, with label from 0 to 186.</p>
<div id="cell-14" class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>lbls <span class="op">=</span> OriginalLabels()</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> n <span class="kw">in</span> [<span class="dv">0</span>, <span class="dv">94</span>, <span class="dv">117</span>, <span class="dv">118</span>]:</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>n<span class="sc">:3d}</span><span class="ss">: </span><span class="sc">{</span>lbls<span class="sc">.</span>label2species(n)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  0: Variola_virus
 94: Middle_East_respiratory_syndrome-related_coronavirus
117: Severe_acute_respiratory_syndrome-related_coronavirus
118: Yellow_fever_virus</code></pre>
</div>
</div>
</section>
</section>
<section id="ncbi-reference-sequences-for-simulated-reads" class="level2">
<h2 class="anchored" data-anchor-id="ncbi-reference-sequences-for-simulated-reads">NCBI reference sequences for simulated reads</h2>
<div id="cell-16" class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>pfs <span class="op">=</span> ProjectFileSystem()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-17" class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data <span class="op">/</span> <span class="st">'ncbi'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/ncbi</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="ncbi-data" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="ncbi-data">NCBI Data</h3>
<p>This directory includes all data related to the work done with reference sequences from NCBI.</p>
<p>The data is organized in the following subfolders:</p>
<ul>
<li><code>refsequences</code>: reference CoV sequences downloaded from NCBI, and related metadata</li>
<li><code>simreads</code>: all data from simulated reads, using ART Illumina simulator and the reference sequences</li>
<li><code>infer_results</code>: results from the inference using models with the simulated reads</li>
<li><code>ds</code>: datasets in proper format for training or inference/prediction using the CNN Virus model</li>
</ul>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
<div id="cell-18" class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/ncbi/refsequences</code>:</p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>No markdown file in this folder</code></pre>
</div>
</div>
<div id="cell-19" class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data <span class="op">/</span> <span class="st">'ncbi/simreads'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/ncbi/simreads</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="ncbi-simulated-reads" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="ncbi-simulated-reads">NCBI simulated reads</h3>
<p>This directory includes all sets of simulated read sequence files generated from NCBI viral sequences using ARC Illumina.</p>
<pre class="ascii"><code>this-directory
    |--cov
    |    |
    |    |--single_10seq_50bp
    |    |    |--single_10seq_50bp.fq
    |    |    |--single_10seq_50bp.alnEnd
    |    |-- ...
    |    |--single_100seq_150bp
    |    |    |--single_100seq_150bp.fq
    |    |    |--single_100seq_150bp.aln
    |    |--paired_100seq_50bp
    |    |    |--paired_100seq_50bp2.aln
    |    |    |--paired_100seq_50bp1.aln
    |    |    |--paired_100seq_50bp2.fq
    |    |    |--paired_100seq_50bp1.fq
    |    |-- ...
    |    |
    |---yf
    |    |
    |    |--yf_AY968064-single-150bp
    |    |    |--yf_AY968064-single-1seq-150bp.fq
    |    |    |--yf_AY968064-single-1seq-150bp.aln
    |    |
    |--mRhiFer1
    |    |--mRhiFer1_v1.p.dna_rm.primary_assembly.1
    |    |    |--mRhiFer1_v1.p.dna_rm.primary_assembly.1.fq
    |    |    |--mRhiFer1_v1.p.dna_rm.primary_assembly.1.aln
    |    |
</code></pre>
<p>This directory includes several subdirectories, each for one virus, e.g.&nbsp;<code>cov</code> for corona, <code>yf</code> for yellow fever.</p>
<p>In each virus subdirectory, several simreads directory includes simulated reads with various parameters, named as <code>&lt;method&gt;_&lt;nb-seq&gt;_&lt;nb-bp&gt;</code> where” - <code>&lt;method&gt;</code> is either <code>single</code> or <code>paired</code> depending on the simulation method - <code>&lt;nb-seq&gt;</code> is the number of reference sequences used for simulation, and refers to the <code>fa</code> file used - <code>&lt;nb-bp&gt;</code> is the number of base pairs used to simulate reads</p>
<p>Each sub-directory includes simreads files made using a simulation method and a specific number of reference sequences. - <code>xxx.fq</code> and <code>xxx.aln</code> files when method is <code>single</code> - <code>xxx1.fq</code>, <code>xxx2.fq</code>, <code>xxx1.aln</code> and <code>xxx2.aln</code> files when method is <code>paired</code>.</p>
<p>Example: - <code>paired_10seq_50bp</code> means that the simreads were generated by using the <code>paired</code> method to simulate 50-bp reads, and using the <code>fa</code> file <code>cov_virus_sequences_010-seqs.fa</code>. - <code>single_100seq_50bp</code> means that the simreads were generated by using the <code>single</code> method to simulate 50-bp reads, and using the <code>fa</code> file <code>cov_virus_sequences_100-seqs.fa</code>. Note that this generated 20,660,104 reads !</p>
<section id="simread-file-formats" class="level4">
<h4 class="anchored" data-anchor-id="simread-file-formats">Simread file formats</h4>
<p>Simulated reads information is split between two files: - <strong>FASTQ</strong> (<code>.fq</code>) files providing the read sequences and their ASCII quality scores - <strong>ALN</strong> (<code>.aln</code>) files with alignment information</p>
<section id="fastq-.fq" class="level5">
<h5 class="anchored" data-anchor-id="fastq-.fq">FASTQ (<code>.fq</code>)</h5>
<p>FASTQ files generated by ART Illumina have the following structure (showing 5 reads), with 4 lines for each read:</p>
<pre class="ascii"><code>@2591237:ncbi:1-60400
ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
+
CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG
@2591237:ncbi:1-60399
GATCAATGTGGCATCTACAATACAGACAGCATGAAGCACCACCAAAGGAC
+
BCBCCFGGGGGGGG1CGGGG&lt;GGBGGGGGFGCGGGGGGDGGG/GG1GGGG
@2591237:ncbi:1-60398
ATCTACCAGTGGTAGATGGGTTCTTAATAATGAACATTATAGAGCTCTAC
+
CCCCCGGGEGG1GGF1G/GGEGGGGGGGGGGGGFFGGGGGGGGGGDGGDG
@2591237:ncbi:1-60397
CGTAAAGTAGAGGCTGTATGGTAGCTAGCACAAATGCCAGCACCAATAGG
+
BCCCCGGGFGGGGGGFGGGGFGG1GGGGGGG&gt;GG1GGGGGGGGGGE&lt;GGG
@2591237:ncbi:1-60396
GGTATCGGGTATCTCCTGCATCAATGCAAGGTCTTACAAAGATAAATACT
+
CBCCCGGG@CGGGGGGGGGGGG=GFGGGGDGGGFG1GGGGGGGG@GGGGG</code></pre>
<p>The following information can be parsed from the each read sequence in the FASTQ file:</p>
<ul>
<li>Line 1: <code>readid</code>, a unique ID for the read, under for format <code>@readid</code></li>
<li>Line 2: <code>readseq</code>, the sequence of the read</li>
<li>Line 3: a separator <code>+</code></li>
<li>Line 4: <code>read_qscores</code>, the base quality scores encoded in ASCII</li>
</ul>
<p>Example:</p>
<pre><code>@2591237:ncbi:1-60400
ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
+
CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG</code></pre>
<ul>
<li><code>readid</code> = <code>2591237:ncbi:1-60400</code></li>
<li><code>readseq</code> = <code>ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG</code>, a 50 bp read</li>
<li><code>read_qscores</code> = <code>CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG</code></li>
</ul>
</section>
</section>
<section id="aln-.aln" class="level4">
<h4 class="anchored" data-anchor-id="aln-.aln">ALN (<code>.aln</code>)</h4>
<p>ALN files generated by ART Illumina consist of : - a header with the ART-Ilumina command used for the simulation (<code>@CM</code>) and info on each of the reference sequences used for the simulations (<code>@SQ</code>). Header always starts with <code>##ART_Illumina</code> and ends with <code>##Header End</code> : - the body with 3 lines for each read: 1. definition line with <code>readid</code>, - reference sequence identification number <code>refseqid</code>, - the position in the read in the reference sequence <code>aln_start_pos</code> - the strand the read was taken from <code>ref_seq_strand</code>. <code>+</code> for coding strand and <code>-</code> for template strand 2. aligned reference sequence, that is the sequence segment in the original reference corresponding to the read 3. aligned read sequence, that is the simmulated read sequence, where each bp corresponds to the reference sequence bp in the same position.</p>
<p>Example of a ALN file generated by ART Illumina (showing 5 reads):</p>
<pre class="ascii"><code>##ART_Illumina    read_length    50
@CM    /bin/art_illumina -i /home/vtec/projects/bio/metagentools/data/cov_data/cov_virus_sequences_ten.fa -ss HS25 -l 50 -f 100 -o /home/vtec/projects/bio/metagentools/data/cov_simreads/single_10seq_50bp/single_10seq_50bp -rs 1674660835
@SQ    2591237:ncbi:1 1   MK211378    2591237    ncbi    1     Coronavirus BtRs-BetaCoV/YN2018D    30213
@SQ    11128:ncbi:2   2   LC494191    11128    ncbi    2     Bovine coronavirus    30942
@SQ    31631:ncbi:3   3   KY967361    31631    ncbi    3     Human coronavirus OC43        30661
@SQ    277944:ncbi:4  4   LC654455    277944    ncbi    4     Human coronavirus NL63    27516
@SQ    11120:ncbi:5   5   MN987231    11120    ncbi    5     Infectious bronchitis virus    27617
@SQ    28295:ncbi:6   6   KU893866    28295    ncbi    6     Porcine epidemic diarrhea virus    28043
@SQ    28295:ncbi:7   7   KJ645638    28295    ncbi    7     Porcine epidemic diarrhea virus    27998
@SQ    28295:ncbi:8   8   KJ645678    28295    ncbi    8     Porcine epidemic diarrhea virus    27998
@SQ    28295:ncbi:9   9   KR873434    28295    ncbi    9     Porcine epidemic diarrhea virus    28038
@SQ    1699095:ncbi:10 10  KT368904    1699095    ncbi    10     Camel alphacoronavirus    27395
##Header End
&gt;2591237:ncbi:1    2591237:ncbi:1-60400    14770    +
ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
&gt;2591237:ncbi:1    2591237:ncbi:1-60399    17012    -
GATCAATGTGGCATCTACAATACAGACAGCATGAAGCACCACCAAAGGAC
GATCAATGTGGCATCTACAATACAGACAGCATGAAGCACCACCAAAGGAC
&gt;2591237:ncbi:1    2591237:ncbi:1-60398    9188    +
ATCTACCAGTGGTAGATGGGTTCTTAATAATGAACATTATAGAGCTCTAC
ATCTACCAGTGGTAGATGGGTTCTTAATAATGAACATTATAGAGCTCTAC
.....</code></pre>
</section>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
</section>
<section id="model-related-data" class="level2">
<h2 class="anchored" data-anchor-id="model-related-data">Model related data</h2>
<div id="cell-21" class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>pfs <span class="op">=</span> ProjectFileSystem()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-22" class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data <span class="op">/</span> <span class="st">'saved'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/saved</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="saved-data-related-to-models" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="saved-data-related-to-models">Saved data related to models</h3>
<p>This directory includes all data related to models and saved: - saved model parameters - saved datasets</p>
<p>For example: - <code>cnn_virus_original/pretrained_model.h5</code> is the saved model parameters for the CNN Virus model - <code>cnn_virus_datasets/*.tfrecords</code> are the preprocessed datasets used for inference or training, saved in TFRecord format for performance</p>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
</section>
</section>
<section id="parsing-sequence-files" class="level1">
<h1>Parsing sequence files</h1>
<p>The following classes make it easier to read and parse files of different formats into their underlying components to generated the training, validation, testing and inference datasets for the model.</p>
<p>Each class inherits from <a href="https://vtecftwy.github.io/metagentools/core.html#textfilebasereader"><code>TextFileBaseReader</code></a> and adds:</p>
<ul>
<li>One or several text parsing method(s) to parse metadata according to a specific format</li>
<li>A file parsing method to extract metadata from all elements in the file, returning it as a key:value dictionary and optionally save the metadata as a json file.</li>
</ul>
<section id="fasta-file" class="level2">
<h2 class="anchored" data-anchor-id="fasta-file">FASTA file</h2>
<p>Extension of <a href="https://vtecftwy.github.io/metagentools/core.html#textfilebasereader"><code>TextFileBaseReader</code></a> class for fasta sequence files.</p>
<p>Structure of a FASTA sequence file:</p>
<pre class="ascii"><code>&gt;definition line - format varies from dataset to dataset
sequence line: sequence of bases</code></pre>
<p>Example for the NCBI datasets:</p>
<pre class="ascii"><code>&gt;seqid accession taxonomyid source seqnb organism
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...
&gt;2591237:ncbi:1 MK211378    2591237 ncbi    1 Coronavirus BtRs-BetaCoV/YN2018D
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...</code></pre>
<div id="cell-27" class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>p2fasta <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/cov_virus_sequences_two.fa'</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> TextFileBaseReader(p2fasta, nlines<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, t <span class="kw">in</span> <span class="bu">enumerate</span>(fasta):</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    txt <span class="op">=</span> t.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>)[:<span class="dv">80</span>] <span class="op">+</span> <span class="st">' ...'</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>txt<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D ...
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...
&gt;11128:ncbi:2   2   LC494191    11128   ncbi    Bovine coronavirus ...
CATCCCGCTTCACTGATCTCTTGTTAGATCTTTTCATAATCTAAACTTTATAAAAACATCCACTCCCTGTAGTCTATGCC ...</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L60" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="fastafilereader" class="level3">
<h3 class="anchored" data-anchor-id="fastafilereader">FastaFileReader</h3>
<blockquote class="blockquote">
<pre><code> FastaFileReader (path:str|pathlib.Path)</code></pre>
</blockquote>
<p><em>Wrap a FASTA file and retrieve its content in raw format and parsed format</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>path</td>
<td>str | Path</td>
<td>path to the Fasta file</td>
</tr>
</tbody>
</table>
<p>As an iterator, <a href="https://vtecftwy.github.io/metagentools/cnn_virus_data.html#fastafilereader"><code>FastaFileReader</code></a> returns a <code>dict</code> at each step, as follows:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'definition line'</span>: <span class="st">'string in file as the definition line for the sequence'</span>,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'sequence'</span>: <span class="st">'the full sequence'</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Illustration:</p>
<div id="cell-30" class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>p2fasta <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/cov_virus_sequences_two.fa'</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>iteration_output <span class="op">=</span> <span class="bu">next</span>(fasta)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(iteration_output[<span class="st">'definition line'</span>][:<span class="dv">80</span>], <span class="st">'...'</span>)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(iteration_output[<span class="st">'sequence'</span>][:<span class="dv">80</span>], <span class="st">'...'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D ...
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...</code></pre>
</div>
</div>
<div id="cell-31" class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"output type :     </span><span class="sc">{</span><span class="bu">type</span>(iteration_output)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"keys :            </span><span class="sc">{</span>iteration_output<span class="sc">.</span>keys()<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"definition line : </span><span class="sc">{</span>iteration_output[<span class="st">'definition line'</span>][:<span class="dv">80</span>]<span class="sc">}</span><span class="ss"> ...'"</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"sequence :       '</span><span class="sc">{</span>iteration_output[<span class="st">'sequence'</span>][:<span class="dv">100</span>]<span class="sc">}</span><span class="ss"> ...'"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>output type :     &lt;class 'dict'&gt;
keys :            dict_keys(['definition line', 'sequence'])
definition line : &gt;2591237:ncbi:1   1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D ...'
sequence :       'TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTAGCTGTCGCTCGGC ...'</code></pre>
</div>
</div>
<p>The <code>definition line</code> is a string, with tab separated values.</p>
<div id="cell-33" class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>display(iteration_output[<span class="st">'definition line'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'&gt;2591237:ncbi:1\t1\tMK211378\t2591237\tncbi\tCoronavirus BtRs-BetaCoV/YN2018D'</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L84" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="fastafilereader.print_first_chunks" class="level3">
<h3 class="anchored" data-anchor-id="fastafilereader.print_first_chunks">FastaFileReader.print_first_chunks</h3>
<blockquote class="blockquote">
<pre><code> FastaFileReader.print_first_chunks (nchunks:int=3)</code></pre>
</blockquote>
<p><em>Print the first <code>nchunks</code> chunks of text from the file</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>nchunks</td>
<td>int</td>
<td>3</td>
<td>number of chunks to print out</td>
</tr>
</tbody>
</table>
<p>This is convenient to quickly discover and explore new fasta files in raw text format:</p>
<div id="cell-36" class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>fasta.print_first_chunks(nchunks<span class="op">=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Sequence 1:
&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...

Sequence 2:
&gt;11128:ncbi:2   2   LC494191    11128   ncbi    Bovine coronavirus
CATCCCGCTTCACTGATCTCTTGTTAGATCTTTTCATAATCTAAACTTTATAAAAACATCCACTCCCTGTAGTCTATGCC ...</code></pre>
</div>
</div>
</section>
<section id="parsing-metadata" class="level3">
<h3 class="anchored" data-anchor-id="parsing-metadata">Parsing metadata</h3>
<p>The class also provides methods to parse metadata from the file content (definition line, headers, …).</p>
<p>A regex pattern is used for parsing metadata fom the definition lines in the reference sequence fasta file.</p>
<p>Below, we parse the data from the definition line of our Corona virus NCBI dataset (rule <code>fasta_ncbi_std</code>):</p>
<p>Sequence 1:</p>
<ul>
<li>Definition Line:</li>
</ul>
<pre class="ascii"><code>&gt;2591237:ncbi:1 [MK211378]  2591237 ncbi    1 [MK211378] 2591237    Coronavirus YN2018D     scientific name</code></pre>
<ul>
<li>Metadata:
<ul>
<li><code>seqid</code> = <code>2591237:ncbi:1</code></li>
<li><code>taxonomyid</code> = <code>2591237</code></li>
<li><code>source</code> = <code>ncbi</code></li>
<li><code>seqnb</code> = <code>1</code></li>
<li><code>accession</code> = <code>MK211378</code></li>
<li><code>species</code> = <code>Coronavirus BtRs-BetaCoV/YN2018D</code></li>
</ul></li>
</ul>
<p>Sequence 2:</p>
<ul>
<li>Definition Line</li>
</ul>
<pre class="ascii"><code>    &gt;11128:ncbi:2 [LC494191]</code></pre>
<ul>
<li>Metadata:
<ul>
<li><code>seqid</code> = <code>11128:ncbi:2</code></li>
<li><code>taxonomyid</code> = <code>11128</code></li>
<li><code>source</code> = <code>ncbi</code></li>
<li><code>seqnb</code> = <code>2</code></li>
<li><code>accession</code> = <code>LC494191</code></li>
<li><code>species</code> = <code>''</code></li>
</ul></li>
</ul>
<p><a href="https://vtecftwy.github.io/metagentools/cnn_virus_data.html#fastafilereader"><code>FastaFileReader</code></a> offers: - <code>parse_text</code> a method to parse the metadata - an option to set a default “parsing rule” for one instance with <code>set_parsing_rules</code>. - <code>parse_file</code> a method to parse the metadata from all sequences in the file and save it as a json file.</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/core.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="textfilebasereader.parse_text" class="level3">
<h3 class="anchored" data-anchor-id="textfilebasereader.parse_text">TextFileBaseReader.parse_text</h3>
<blockquote class="blockquote">
<pre><code> TextFileBaseReader.parse_text (txt:str, pattern:str=None,
                                keys:list[str]=None)</code></pre>
</blockquote>
<p>*Parse text using regex pattern and keys. Return a metadata dictionary.</p>
<p>The passed text is parsed using the regex pattern. The method return a dictionary in the format:</p>
<pre><code>{
    'key_1': 'metadata 1',
    'key_2': 'metadata 2',
    ...
}*</code></pre>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>txt</td>
<td>str</td>
<td></td>
<td>text to parse</td>
</tr>
<tr class="even">
<td>pattern</td>
<td>str</td>
<td>None</td>
<td>If None, uses standard regex pattern to extract metadata, otherwise, uses passed regex</td>
</tr>
<tr class="odd">
<td>keys</td>
<td>list</td>
<td>None</td>
<td>If None, uses standard regex list of keys, otherwise, uses passed list of keys (str)</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td></td>
<td><strong>parsed metadata in key/value format</strong></td>
</tr>
</tbody>
</table>
<p>Running the parser function with specifically defined <code>pattern</code> and <code>keys</code>.</p>
<div id="cell-44" class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>dfn_line, sequence <span class="op">=</span> <span class="bu">next</span>(fasta).values()</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(dfn_line.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D</code></pre>
</div>
</div>
<div id="cell-45" class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pattern = r"^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*\[(?P&lt;accession&gt;[\w\d]*)\]([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t]*(?P=seqnb)[\s\t]*\[(?P=accession)\][\s\t]*(?P=taxonomyid)[\s\t]*(?P&lt;species&gt;[\w\s\-\_\/]*))?"</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>pattern <span class="op">=</span> <span class="vs">r"^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*(?P=seqnb)[\s\t](?P&lt;accession&gt;[\w\d]*)([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t][\s\t]*(?P&lt;organism&gt;[\w\s\-\_\/]*))?"</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>keys <span class="op">=</span> <span class="st">'seqid taxonomyid accession source seqnb organism'</span>.split(<span class="st">' '</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-46" class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>fasta.parse_text(dfn_line, pattern<span class="op">=</span>pattern, keys<span class="op">=</span>keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'accession': 'MK211378',
 'organism': 'Coronavirus BtRs-BetaCoV/YN2018D',
 'seqid': '2591237:ncbi:1',
 'seqnb': '1',
 'source': 'ncbi',
 'taxonomyid': '2591237'}</code></pre>
</div>
</div>
<p>When a <a href="https://vtecftwy.github.io/metagentools/cnn_virus_data.html#fastafilereader"><code>FastaFileReader</code></a> instance is created, all existing rules in the file <code>default_parsing_rules.json</code> are tested on the first definition line of the fasta file and the one rule that parses the most matches will be selected automatically and saved in instance attributes <code>re_rule_name</code>, <code>re_pattern</code> and <code>re_keys</code>.</p>
<p><code>parse_file</code> extract metadata from each definition line in the fasta file and return a dictionary with all metadata.</p>
<div id="cell-48" class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta.re_rule_name)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta.re_pattern)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta.re_keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fasta_ncbi_std
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*(?P=seqnb)[\s\t](?P&lt;accession&gt;[\w\d]*)([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t][\s\t]*(?P&lt;organism&gt;[\w\s\-\_/]*))?
['seqid', 'taxonomyid', 'source', 'accession', 'seqnb', 'organism']</code></pre>
</div>
</div>
<div id="cell-49" class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>fasta.parse_text(dfn_line)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'accession': 'MK211378',
 'organism': 'Coronavirus BtRs-BetaCoV/YN2018D',
 'seqid': '2591237:ncbi:1',
 'seqnb': '1',
 'source': 'ncbi',
 'taxonomyid': '2591237'}</code></pre>
</div>
</div>
<p>When another fasta file, which has another definition line structure, is used, another parsing rule is selected.</p>
<div id="cell-51" class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>p2other <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/another_sequence.fa'</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> p2other.is_file()</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>it2 <span class="op">=</span> FastaFileReader(path<span class="op">=</span>p2other)</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>dfn_line, sequence <span class="op">=</span> <span class="bu">next</span>(it2).values()</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(dfn_line.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;1 dna_rm:primary_assembly primary_assembly:mRhiFer1_v1.p:1:1:124933378:1 REF</code></pre>
</div>
</div>
<div id="cell-52" class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(it2.re_rule_name)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(it2.re_pattern)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(it2.re_keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fasta_rhinolophus_ferrumequinum
^&gt;\d[\s\t](?P&lt;seq_type&gt;dna_rm):(?P&lt;id_type&gt;[\w\_]*)[\s\w](?P=id_type):(?P&lt;assy&gt;[\w\d\_]*)\.(?P&lt;seq_level&gt;[\w]*):\d*:\d*:(?P&lt;taxonomy&gt;\d*):(?P&lt;id&gt;\d*)[\s    ]REF$
['seq_type', 'id_type', 'assy', 'seq_level', 'taxonomy', 'id']</code></pre>
</div>
</div>
<div id="cell-53" class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>pprint(it2.parse_text(dfn_line))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'assy': 'mRhiFer1_v1',
 'id': '1',
 'id_type': 'primary_assembly',
 'seq_level': 'p',
 'seq_type': 'dna_rm',
 'taxonomy': '124933378'}</code></pre>
</div>
</div>
<p>This rule selection is performed by the class method <code>set_parsing_rule</code>. The method can also be called with specific <code>pattern</code> and <code>keys</code> to force parsing rule not yet saved in the json file.</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/core.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="textfilebasereader.set_parsing_rules" class="level3">
<h3 class="anchored" data-anchor-id="textfilebasereader.set_parsing_rules">TextFileBaseReader.set_parsing_rules</h3>
<blockquote class="blockquote">
<pre><code> TextFileBaseReader.set_parsing_rules (pattern:str|bool=None,
                                       keys:list[str]=None,
                                       verbose:bool=False)</code></pre>
</blockquote>
<p>*Set the standard regex parsing rule for the file.</p>
<p>Rules can be set:</p>
<ol type="1">
<li>manually by passing specific custom values for <code>pattern</code> and <code>keys</code></li>
<li>automatically, by testing all parsing rules saved in <code>parsing_rule.json</code></li>
</ol>
<p>Automatic selection of parsing rules works by testing each rule saved in <code>parsing_rule.json</code> on the first definition line of the fasta file, and selecting the one rule that generates the most metadata matches.</p>
<p>Rules consists of two parameters:</p>
<ul>
<li>The regex pattern including one <code>group</code> for each metadata item, e.g <code>(?P&lt;group_name&gt;regex_code)</code></li>
<li>The list of keys, i.e.&nbsp;the list with the name of each regex groups, used as key in the metadata dictionary</li>
</ul>
<p>This method updates the three following class attributes: <code>re_rule_name</code>, <code>re_pattern</code>, <code>re_keys</code>*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pattern</td>
<td>str | bool</td>
<td>None</td>
<td>regex pattern to apply to parse the text, search in parsing rules json if None</td>
</tr>
<tr class="even">
<td>keys</td>
<td>list</td>
<td>None</td>
<td>list of keys/group for regex, search in parsing rules json if None</td>
</tr>
<tr class="odd">
<td>verbose</td>
<td>bool</td>
<td>False</td>
<td>when True, provides information on each rule</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-56" class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>dfn_line, sequence <span class="op">=</span> <span class="bu">next</span>(fasta).values()</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"definition line: '</span><span class="sc">{</span>dfn_line[:<span class="op">-</span><span class="dv">1</span>]<span class="sc">}</span><span class="ss">'"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>definition line: '&gt;2591237:ncbi:1   1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018'</code></pre>
</div>
</div>
<p>Automatic parsing works by testing each saved rule for the value of <code>definition line</code> in the first sequence in the fasta file.</p>
<div id="cell-58" class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"key for text to parse: </span><span class="sc">{</span>fasta<span class="sc">.</span>text_to_parse_key<span class="sc">}</span><span class="ch">\n</span><span class="ss">"</span>)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>fasta.reset_iterator()</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Text to parse for testing (extracted from first iteration):'</span>)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">next</span>(fasta)[fasta.text_to_parse_key])</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>()</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>fasta.set_parsing_rules(verbose<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>key for text to parse: definition line

Text to parse for testing (extracted from first iteration):
&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D

--------------------------------------------------------------------------------
Rule &lt;fasta_ncbi_std&gt; generated 6 matches
--------------------------------------------------------------------------------
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*(?P=seqnb)[\s\t](?P&lt;accession&gt;[\w\d]*)([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t][\s\t]*(?P&lt;organism&gt;[\w\s\-\_/]*))?
['seqid', 'taxonomyid', 'source', 'accession', 'seqnb', 'organism']
--------------------------------------------------------------------------------
Rule &lt;fastq_art_illumina_ncbi_std&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;aln_art_illumina_ncbi_std&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;aln_art_illumina-refseq-ncbi-std&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;fasta_ncbi_cov&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;fasta_rhinolophus_ferrumequinum&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Selected rule with most matches: fasta_ncbi_std</code></pre>
</div>
</div>
<p>If no saved rule generates a match, <code>re_rule_name</code>, <code>re_pattern</code> and <code>re_keys</code> remain <code>None</code> and a warning message is issued to ask user to add a parsing rule manually.</p>
<div id="cell-60" class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>p2nomatch <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/sequences_two_no_matching_rule.fa'</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>fasta2 <span class="op">=</span> FastaFileReader(p2nomatch)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/vtec/projects/bio/metagentools/metagentools/core.py:520: UserWarning: 
        None of the saved parsing rules were able to extract metadata from the first line in this file.
        You must set a custom rule (pattern + keys) before parsing text, by using:
            `self.set_parsing_rules(custom_pattern, custom_list_of_keys)`
                
  warnings.warn(msg, category=UserWarning)</code></pre>
</div>
</div>
<div id="cell-61" class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>fasta2.re_rule_name <span class="kw">is</span> <span class="va">None</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>True</code></pre>
</div>
</div>
<p>But we still can set a standard rule manually, by passing a re pattern and the corresponding list of keys.</p>
<div id="cell-63" class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>pat <span class="op">=</span> <span class="vs">r"^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))\s*(?P&lt;text&gt;[\w\s]*)$"</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>keys <span class="op">=</span> <span class="st">"seqid taxonomyid source seqnb text"</span>.split()</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>fasta2.set_parsing_rules(pattern<span class="op">=</span>pat, keys<span class="op">=</span>keys)</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta2.re_rule_name)</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta2.re_pattern)</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta2.re_keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Custom Rule
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))\s*(?P&lt;text&gt;[\w\s]*)$
['seqid', 'taxonomyid', 'source', 'seqnb', 'text']</code></pre>
</div>
</div>
<div id="cell-64" class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>fasta2.reset_iterator()</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>dfn_line, sequence <span class="op">=</span> <span class="bu">next</span>(fasta2).values()</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"definition line: '</span><span class="sc">{</span>dfn_line[:<span class="op">-</span><span class="dv">1</span>]<span class="sc">}</span><span class="ss">'"</span>)</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>fasta2.parse_text(dfn_line)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>definition line: '&gt;2591237:ncbi:1 this sequence does not match any saved parsing rul'</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>{'seqid': '2591237:ncbi:1',
 'seqnb': '1',
 'source': 'ncbi',
 'taxonomyid': '2591237',
 'text': 'this sequence does not match any saved parsing rule'}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L97" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="fastafilereader.parse_file" class="level3">
<h3 class="anchored" data-anchor-id="fastafilereader.parse_file">FastaFileReader.parse_file</h3>
<blockquote class="blockquote">
<pre><code> FastaFileReader.parse_file (add_seq:bool=False, save_json:bool=False)</code></pre>
</blockquote>
<p><em>Read fasta file and return a dictionary with definition line metadata and optionally sequences</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>add_seq</td>
<td>bool</td>
<td>False</td>
<td>When True, add the full sequence to the parsed metadata dictionary</td>
</tr>
<tr class="even">
<td>save_json</td>
<td>bool</td>
<td>False</td>
<td>When True, save the file metadata as a json file of same stem name</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict[str]</strong></td>
<td></td>
<td><strong>Metadata as Key/Values pairs</strong></td>
</tr>
</tbody>
</table>
<div id="cell-66" class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>pprint(fasta.parse_file())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'11128:ncbi:2': {'accession': 'LC494191',
                  'organism': 'Bovine coronavirus',
                  'seqid': '11128:ncbi:2',
                  'seqnb': '2',
                  'source': 'ncbi',
                  'taxonomyid': '11128'},
 '2591237:ncbi:1': {'accession': 'MK211378',
                    'organism': 'Coronavirus BtRs-BetaCoV/YN2018D',
                    'seqid': '2591237:ncbi:1',
                    'seqnb': '1',
                    'source': 'ncbi',
                    'taxonomyid': '2591237'}}</code></pre>
</div>
</div>
<div id="cell-67" class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>fasta.parse_file(save_json<span class="op">=</span><span class="va">True</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Metadata for 'cov_virus_sequences_two.fa'&gt; saved as &lt;cov_virus_sequences_two_metadata.json&gt; in  
/home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov
</code></pre>
</div>
</div>
<div id="cell-68" class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">'../default_parsing_rules.json'</span>, <span class="st">'r'</span>) <span class="im">as</span> fp:</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>    pprint(json.load(fp), width<span class="op">=</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'aln_art_illumina-refseq-ncbi-std': {'keys': 'refseqid '
                                              'reftaxonomyid '
                                              'refsource '
                                              'refseqnb '
                                              'refseq_accession '
                                              'organism '
                                              'refseq_length',
                                      'pattern': '^@SQ[\\t\\s]*(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\\d*):(?P&lt;refsource&gt;\\w*):(?P&lt;refseqnb&gt;\\d*))[\\t\\s]*(?P=refseqnb)[\\t\\s]*(?P&lt;refseq_accession&gt;[\\w\\d]*)[\\t\\s]*(?P=reftaxonomyid)[\\t\\s]*(?P=refsource)[\\t\\s](?P&lt;organism&gt;.*)[\\t\\s](?P&lt;refseq_length&gt;\\d*)$'},
 'aln_art_illumina_ncbi_std': {'keys': 'refseqid '
                                       'reftaxonomyid '
                                       'refsource '
                                       'refseqnb '
                                       'readid '
                                       'readnb '
                                       'aln_start_pos '
                                       'refseq_strand',
                               'pattern': '^&gt;(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\\d*):(?P&lt;refsource&gt;\\w*):(?P&lt;refseqnb&gt;\\d*))(\\s|\t'
                                          ')*(?P&lt;readid&gt;(?P=reftaxonomyid):(?P=refsource):(?P=refseqnb)-(?P&lt;readnb&gt;\\d*(\\/\\d(-\\d)?)?))(\\s|\\t)(?P&lt;aln_start_pos&gt;\\d*)(\\s|\\t)(?P&lt;refseq_strand&gt;(-|\\+))$'},
 'fasta_ncbi_cov': {'keys': 'seqid '
                            'taxonomyid '
                            'source '
                            'accession '
                            'seqnb '
                            'organism',
                    'pattern': '^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\\d*))[\\s\\t]*\\[(?P&lt;accession&gt;[\\w\\d]*)\\]([\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P=source)[\\s\\t]*(?P=seqnb)[\\s\\t]*\\[(?P=accession)\\][\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P&lt;organism&gt;[\\w\\s\\-\\_\\/]*))?'},
 'fasta_ncbi_std': {'keys': 'seqid '
                            'taxonomyid '
                            'source '
                            'accession '
                            'seqnb '
                            'organism',
                    'pattern': '^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\\d*))[\\s\\t]*(?P=seqnb)[\\s\\t](?P&lt;accession&gt;[\\w\\d]*)([\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P=source)[\\s\\t][\\s\\t]*(?P&lt;organism&gt;[\\w\\s\\-\\_/]*))?'},
 'fasta_rhinolophus_ferrumequinum': {'keys': 'seq_type '
                                             'id_type '
                                             'assy '
                                             'seq_level '
                                             'taxonomy '
                                             'id',
                                     'pattern': '^&gt;\\d[\\s\\t](?P&lt;seq_type&gt;dna_rm):(?P&lt;id_type&gt;[\\w\\_]*)[\\s\\w](?P=id_type):(?P&lt;assy&gt;[\\w\\d\\_]*)\\.(?P&lt;seq_level&gt;[\\w]*):\\d*:\\d*:(?P&lt;taxonomy&gt;\\d*):(?P&lt;id&gt;\\d*)[\\s\t'
                                                ']REF$'},
 'fastq_art_illumina_ncbi_std': {'keys': 'readid '
                                         'reftaxonomyid '
                                         'refsource '
                                         'refseqnb '
                                         'readnb',
                                 'pattern': '^@(?P&lt;readid&gt;(?P&lt;reftaxonomyid&gt;\\d*):(?P&lt;refsource&gt;\\w*):(?P&lt;refseqnb&gt;\\d*)-(?P&lt;readnb&gt;\\d*(\\/\\d)?))$'}}</code></pre>
</div>
</div>
<div id="cell-69" class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>p2fasta <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/cov_virus_sequence_one.fa'</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>fasta_meta <span class="op">=</span> fasta.parse_file(save_json<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>pprint(fasta_meta)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Metadata for 'cov_virus_sequence_one.fa'&gt; saved as &lt;cov_virus_sequence_one_metadata.json&gt; in  
/home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov

{'2591237:ncbi:1': {'accession': 'MK211378',
                    'organism': 'Coronavirus BtRs-BetaCoV/YN2018D',
                    'seqid': '2591237:ncbi:1',
                    'seqnb': '1',
                    'source': 'ncbi',
                    'taxonomyid': '2591237'}}</code></pre>
</div>
</div>
</section>
</section>
<section id="fastq-file" class="level2">
<h2 class="anchored" data-anchor-id="fastq-file">FASTQ file</h2>
<p>Extension of <a href="https://vtecftwy.github.io/metagentools/core.html#textfilebasereader"><code>TextFileBaseReader</code></a> class for fastq sequence files.</p>
<p>Structure of a FASTQ sequence file:</p>
<div id="cell-72" class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>p2fastq <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/simreads/cov/single_1seq_150bp/single_1seq_150bp.fq'</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>fastq <span class="op">=</span> TextFileBaseReader(p2fastq, nlines<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, t <span class="kw">in</span> <span class="bu">enumerate</span>(fastq):</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>    txt <span class="op">=</span> t.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>)[:<span class="dv">80</span>]</span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>txt<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">11</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>@2591237:ncbi:1-40200
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC
+
CCCGGGCGGGGGCJGJJJGJJGJJJGJGGJGJJJGJGGGGGGGGCJGJGGGGGJJJJGCCGGGGGJCGCGJGJCG=GGGG
@2591237:ncbi:1-40199
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGAGTCATTTGA
+
=CCGGGGCGGGGGJJGJJGJGJG=GJJGJCGJJJCJ=JJJJGGJJCJGJGG=JGC1JJGG8GCJCGGGCGG(GCGGCGC=
@2591237:ncbi:1-40198
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT
+
C=CGGGGGGGGGGCJJJJ=JJJJJJJJJJJGGJJJJ1GJJ8GJJGGGJGGJJC=JJGGGCCGG88GG=GGGGGGCJGGGG</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L122" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="fastqfilereader" class="level3">
<h3 class="anchored" data-anchor-id="fastqfilereader">FastqFileReader</h3>
<blockquote class="blockquote">
<pre><code> FastqFileReader (path:str|pathlib.Path)</code></pre>
</blockquote>
<p><em>Iterator going through a fastq file’s sequences and return each section + prob error as a dict</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>path</td>
<td>str | Path</td>
<td>path to the fastq file</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td><strong>key/value with keys: definition line; sequence; q score; prob error</strong></td>
</tr>
</tbody>
</table>
<div id="cell-74" class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>fastq <span class="op">=</span> FastqFileReader(p2fastq)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>iteration_output <span class="op">=</span> <span class="bu">next</span>(fastq)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">type</span>(iteration_output))</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(iteration_output.keys())</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Definition line:  </span><span class="sc">{</span>iteration_output[<span class="st">'definition line'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Read sequence:    </span><span class="sc">{</span>iteration_output[<span class="st">'sequence'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Q scores (ASCII): </span><span class="sc">{</span>iteration_output[<span class="st">'read_qscores'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Prob error:       </span><span class="sc">{</span><span class="st">','</span><span class="sc">.</span>join([<span class="ss">f'</span><span class="sc">{</span>p<span class="sc">:.4f}</span><span class="ss">'</span> <span class="cf">for</span> p <span class="kw">in</span> iteration_output[<span class="st">'probs error'</span>]])<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'dict'&gt;
dict_keys(['definition line', 'sequence', 'read_qscores', 'probs error'])
Definition line:  @2591237:ncbi:1-40200
Read sequence:    TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAGTTTCAAGGAACTTTTAGTGTATGCTGCTGATCCAGCCATGCATGCAGCTT
Q scores (ASCII): CCCGGGCGGGGGCJGJJJGJJGJJJGJGGJGJJJGJGGGGGGGGCJGJGGGGGJJJJGCCGGGGGJCGCGJGJCG=GGGG=CGGGGGG1GCGCGGGGCCGJC8GGGGGGGGGGGCGGGGGGGGGGGC8GGGGGGCGGC1GGGCGGGGGCC
Prob error:       0.0004,0.0004,0.0004,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0001,0.0002,0.0001,0.0001,0.0001,0.0002,0.0001,0.0001,0.0002,0.0001,0.0001,0.0001,0.0002,0.0001,0.0002,0.0002,0.0001,0.0002,0.0001,0.0001,0.0001,0.0002,0.0001,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0001,0.0002,0.0001,0.0002,0.0002,0.0002,0.0002,0.0002,0.0001,0.0001,0.0001,0.0001,0.0002,0.0004,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0001,0.0004,0.0002,0.0004,0.0002,0.0001,0.0002,0.0001,0.0004,0.0002,0.0016,0.0002,0.0002,0.0002,0.0002,0.0016,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0251,0.0002,0.0004,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0004,0.0004,0.0002,0.0001,0.0004,0.0050,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0050,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0004,0.0251,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0004</code></pre>
</div>
</div>
<p>Five largest probabilities of error:</p>
<div id="cell-76" class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>np.sort(iteration_output[<span class="st">'probs error'</span>])[<span class="op">-</span><span class="dv">5</span>:]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>array([0.00158489, 0.00501187, 0.00501187, 0.02511886, 0.02511886])</code></pre>
</div>
</div>
<div id="cell-77" class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>np.argsort(iteration_output[<span class="st">'probs error'</span>])[<span class="op">-</span><span class="dv">5</span>:]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>array([ 80, 127, 102, 138,  88])</code></pre>
</div>
</div>
<div id="cell-78" class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>dfn_line <span class="op">=</span> iteration_output[<span class="st">'definition line'</span>]</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>meta <span class="op">=</span> fastq.parse_text(dfn_line)</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>meta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'readid': '2591237:ncbi:1-40200',
 'readnb': '40200',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}</code></pre>
</div>
</div>
<div id="cell-79" class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>fastq <span class="op">=</span> FastqFileReader(p2fastq)</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="bu">next</span>(fastq).keys()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>dict_keys(['definition line', 'sequence', 'read_qscores', 'probs error'])</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L163" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="fastqfilereader.parse_file" class="level3">
<h3 class="anchored" data-anchor-id="fastqfilereader.parse_file">FastqFileReader.parse_file</h3>
<blockquote class="blockquote">
<pre><code> FastqFileReader.parse_file (add_readseq:bool=False,
                             add_qscores:bool=False,
                             add_probs_error:bool=False,
                             save_json:bool=False)</code></pre>
</blockquote>
<p><em>Read fastq file, return a dict with definition line metadata and optionally read sequence and q scores, …</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>add_readseq</td>
<td>bool</td>
<td>False</td>
<td>When True, add the full sequence to the parsed metadata dictionary</td>
</tr>
<tr class="even">
<td>add_qscores</td>
<td>bool</td>
<td>False</td>
<td>Add the read ASCII Q Scores to the parsed dictionary when True</td>
</tr>
<tr class="odd">
<td>add_probs_error</td>
<td>bool</td>
<td>False</td>
<td>Add the read probability of error to the parsed dictionary when True</td>
</tr>
<tr class="even">
<td>save_json</td>
<td>bool</td>
<td>False</td>
<td>When True, save the file metadata as a json file of same stem name</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict[str]</strong></td>
<td></td>
<td><strong>Metadata as Key/Values pairs</strong></td>
</tr>
</tbody>
</table>
<div id="cell-81" class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>parsed <span class="op">=</span> fastq.parse_file(add_readseq<span class="op">=</span><span class="va">False</span>, add_qscores<span class="op">=</span><span class="va">False</span>, add_probs_error<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (k, v) <span class="kw">in</span> <span class="bu">enumerate</span>(parsed.items()):</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(k)</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>    pprint(v)</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;=</span><span class="dv">3</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>2591237:ncbi:1-40200
{'readid': '2591237:ncbi:1-40200',
 'readnb': '40200',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40199
{'readid': '2591237:ncbi:1-40199',
 'readnb': '40199',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40198
{'readid': '2591237:ncbi:1-40198',
 'readnb': '40198',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40197
{'readid': '2591237:ncbi:1-40197',
 'readnb': '40197',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}</code></pre>
</div>
</div>
<div id="cell-82" class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="op">=</span> fastq.parse_file(add_readseq<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(metadata).T</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>df.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">readid</th>
<th data-quarto-table-cell-role="th">readnb</th>
<th data-quarto-table-cell-role="th">refseqnb</th>
<th data-quarto-table-cell-role="th">refsource</th>
<th data-quarto-table-cell-role="th">reftaxonomyid</th>
<th data-quarto-table-cell-role="th">readseq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40200</td>
<td>2591237:ncbi:1-40200</td>
<td>40200</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCG...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40199</td>
<td>2591237:ncbi:1-40199</td>
<td>40199</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATT...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40198</td>
<td>2591237:ncbi:1-40198</td>
<td>40198</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTC...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40197</td>
<td>2591237:ncbi:1-40197</td>
<td>40197</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>TAATCACTGATAGCAGCATTGCCATCCTGAGCAAAGAAGAAGTGTT...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40196</td>
<td>2591237:ncbi:1-40196</td>
<td>40196</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>CTAATGTCAGTACGCCTACAATGCCTGCATCACGCATAGCATCGCA...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40195</td>
<td>2591237:ncbi:1-40195</td>
<td>40195</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>AAGCTGAAGCATACATAACACAGTCCTTAAGCCGATAACCAGACAA...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40194</td>
<td>2591237:ncbi:1-40194</td>
<td>40194</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>AGTGGAAGAACTTCACCGTCAAGATGAAACTCGACGGGGCTCTCCA...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40193</td>
<td>2591237:ncbi:1-40193</td>
<td>40193</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>GCGTCTCGAGTGCTTCGAGTTCACCGTTCTTGAGAACAACCTCCTC...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40192</td>
<td>2591237:ncbi:1-40192</td>
<td>40192</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>CTGGTAGTATCTAAGGCTCCACTGAAATACTTGTACTTGTTATATA...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40191</td>
<td>2591237:ncbi:1-40191</td>
<td>40191</td>
<td>1</td>
<td>ncbi</td>
<td>2591237</td>
<td>GTCTCTATCTGTAGTACTATGACAAATAGACAGTTTCATCAGAAAT...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-83" class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>fastq.set_parsing_rules(verbose<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--------------------------------------------------------------------------------
Rule &lt;fasta_ncbi_std&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;fastq_art_illumina_ncbi_std&gt; generated 5 matches
--------------------------------------------------------------------------------
^@(?P&lt;readid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*)-(?P&lt;readnb&gt;\d*(\/\d)?))$
['readid', 'reftaxonomyid', 'refsource', 'refseqnb', 'readnb']
--------------------------------------------------------------------------------
Rule &lt;aln_art_illumina_ncbi_std&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;aln_art_illumina-refseq-ncbi-std&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;fasta_ncbi_cov&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;fasta_rhinolophus_ferrumequinum&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Selected rule with most matches: fastq_art_illumina_ncbi_std</code></pre>
</div>
</div>
</section>
</section>
<section id="aln-alignment-files" class="level2">
<h2 class="anchored" data-anchor-id="aln-alignment-files">ALN Alignment Files</h2>
<p>Extension of <a href="https://vtecftwy.github.io/metagentools/core.html#textfilebasereader"><code>TextFileBaseReader</code></a> class for ALN read/sequence alignment files.</p>
<p>Structure of a ALN sequence file:</p>
<div id="cell-86" class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>p2aln <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/simreads/cov/single_1seq_150bp/single_1seq_150bp.aln'</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> p2aln.is_file()</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>aln <span class="op">=</span> TextFileBaseReader(p2aln, nlines<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, t <span class="kw">in</span> <span class="bu">enumerate</span>(aln):</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>    txt <span class="op">=</span> t.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>)[:<span class="dv">80</span>]</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>txt<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">12</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>##ART_Illumina  read_length 150
@CM /usr/bin/art_illumina -i /home/vtec/projects/bio/metagentools/data/ncbi/refs
@SQ 2591237:ncbi:1  1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D    3021
##Header End
&gt;2591237:ncbi:1 2591237:ncbi:1-40200    14370   +
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC
&gt;2591237:ncbi:1 2591237:ncbi:1-40199    15144   -
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGATTCATTTGA
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGAGTCATTTGA
&gt;2591237:ncbi:1 2591237:ncbi:1-40198    2971    -
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L192" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="alnfilereader" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader">AlnFileReader</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader (path:str|pathlib.Path)</code></pre>
</blockquote>
<p><em>Iterator going through an ALN file</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>path</td>
<td>str | Path</td>
<td>path to the aln file</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td><strong>key/value with keys:</strong></td>
</tr>
</tbody>
</table>
<div id="cell-88" class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>aln <span class="op">=</span> AlnFileReader(p2aln)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><a href="https://vtecftwy.github.io/metagentools/cnn_virus_data.html#alnfilereader"><code>AlnFileReader</code></a> iterator returns elements one by one, as dictionaries with each data line related to the read, accessible through the following keys:</p>
<ul>
<li>key <code>'definition line'</code>: <strong>read definition line</strong>, including read metadata</li>
<li>key <code>'ref_seq_aligned'</code>: <strong>aligned reference sequence</strong>, that is the sequence segment in the original reference corresponding to the read</li>
<li>key <code>'read_seq_aligned'</code>: <strong>aligned read</strong>, that is the simmulated read sequence, where each bp corresponds to the reference sequence bp in the same position.</li>
</ul>
<div id="cell-90" class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>one_iteration <span class="op">=</span> <span class="bu">next</span>(aln)</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>one_iteration.keys()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>dict_keys(['definition line', 'ref_seq_aligned', 'read_seq_aligned'])</code></pre>
</div>
</div>
<div id="cell-91" class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>pprint(one_iteration)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'definition line': '&gt;2591237:ncbi:1\t2591237:ncbi:1-40200\t14370\t+',
 'read_seq_aligned': 'TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAGTTTCAAGGAACTTTTAGTGTATGCTGCTGATCCAGCCATGCATGCAGCTT',
 'ref_seq_aligned': 'TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAGTTTCAAGGAACTTTTAGTGTATGCTGCTGATCCAGCCATGCATGCAGCTT'}</code></pre>
</div>
</div>
<div id="cell-92" class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>dfn_line, ref_seq_aligned, read_seq_aligned <span class="op">=</span> one_iteration.values()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-93" class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>dfn_line</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'&gt;2591237:ncbi:1\t2591237:ncbi:1-40200\t14370\t+'</code></pre>
</div>
</div>
<div id="cell-94" class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>ref_seq_aligned[:<span class="dv">100</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAG'</code></pre>
</div>
</div>
<div id="cell-95" class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>read_seq_aligned[:<span class="dv">100</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAG'</code></pre>
</div>
</div>
<div id="cell-96" class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>another_iteration <span class="op">=</span> <span class="bu">next</span>(aln)</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>pprint(another_iteration)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'definition line': '&gt;2591237:ncbi:1\t2591237:ncbi:1-40199\t15144\t-',
 'read_seq_aligned': 'TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGAGTCATTTGAGTTATAGTAGGGATGACATTACGCTTAGTATACGCGAAAAGTGCATCTTGATCCTCATAACTCATTGAGT',
 'ref_seq_aligned': 'TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGATTCATTTGAGTTATAGTAGGGATGACATTACGCTTAGTATACGCGAAAAGTGCATCTTGATCCTCATAACTCATTGAGT'}</code></pre>
</div>
</div>
<div id="cell-97" class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>aln.reset_iterator()</span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, d <span class="kw">in</span> <span class="bu">enumerate</span>(aln):</span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(d[<span class="st">'definition line'</span>])</span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(d[<span class="st">'ref_seq_aligned'</span>][:<span class="dv">80</span>], <span class="st">'...'</span>)</span>
<span id="cb115-5"><a href="#cb115-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(d[<span class="st">'read_seq_aligned'</span>][:<span class="dv">80</span>], <span class="st">'...</span><span class="ch">\n</span><span class="st">'</span>)</span>
<span id="cb115-6"><a href="#cb115-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">3</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;2591237:ncbi:1 2591237:ncbi:1-40200    14370   +
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC ...
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC ...

&gt;2591237:ncbi:1 2591237:ncbi:1-40199    15144   -
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGATTCATTTGA ...
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGAGTCATTTGA ...

&gt;2591237:ncbi:1 2591237:ncbi:1-40198    2971    -
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT ...
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT ...

&gt;2591237:ncbi:1 2591237:ncbi:1-40197    15485   -
TAATCACTGATAGCAGCATTGCCATCCTGAGCAAAGAAGAAGTGTTTTAGTTCAACAGAACTTCCTTCCTTAAAGAAACC ...
TAATCACTGATAGCAGCATTGCCATCCTGAGCAAAGAAGAAGTGTTTTAGTTCAACAGAACTTCCTTCCTTAAAGAAACC ...
</code></pre>
</div>
</div>
<p>Once instantiated, the <a href="https://vtecftwy.github.io/metagentools/cnn_virus_data.html#alnfilereader"><code>AlnFileReader</code></a> iterator gives access to the file’s header information through <code>header</code> instance attribute. It is a dictionary with two keys: <code>'command'</code> and <code>'reference sequences'</code>:</p>
<pre><code>    {'command':             'art-illumina command used to create the reads',
     'reference sequences': ['@SQ metadata on reference sequence 1 used for the reads',
                             '@SQ metadata on reference sequence 2 used for the reads', 
                             ...
                            ]
    }</code></pre>
<div id="cell-99" class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(aln.header[<span class="st">'command'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>/usr/bin/art_illumina -i /home/vtec/projects/bio/metagentools/data/ncbi/refsequences/cov/cov_refseq_001-seq1.fa -ss HS25 -l 150 -f 200 -o /home/vtec/projects/bio/metagentools/data/ncbi/simreads/cov/single_1seq_150bp/single_1seq_150bp -rs 1723893089</code></pre>
</div>
</div>
<div id="cell-100" class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> seq_info <span class="kw">in</span> aln.header[<span class="st">'reference sequences'</span>]:</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(seq_info)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>@SQ 2591237:ncbi:1  1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D    30213</code></pre>
</div>
</div>
<p>The <strong>read definition line</strong> includes key metadata, which need to be parsed using the appropriate parsing rule.</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/core.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="textfilebasereader.parse_text-1" class="level3">
<h3 class="anchored" data-anchor-id="textfilebasereader.parse_text-1">TextFileBaseReader.parse_text</h3>
<blockquote class="blockquote">
<pre><code> TextFileBaseReader.parse_text (txt:str, pattern:str=None,
                                keys:list[str]=None)</code></pre>
</blockquote>
<p>*Parse text using regex pattern and keys. Return a metadata dictionary.</p>
<p>The passed text is parsed using the regex pattern. The method return a dictionary in the format:</p>
<pre><code>{
    'key_1': 'metadata 1',
    'key_2': 'metadata 2',
    ...
}*</code></pre>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>txt</td>
<td>str</td>
<td></td>
<td>text to parse</td>
</tr>
<tr class="even">
<td>pattern</td>
<td>str</td>
<td>None</td>
<td>If None, uses standard regex pattern to extract metadata, otherwise, uses passed regex</td>
</tr>
<tr class="odd">
<td>keys</td>
<td>list</td>
<td>None</td>
<td>If None, uses standard regex list of keys, otherwise, uses passed list of keys (str)</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td></td>
<td><strong>parsed metadata in key/value format</strong></td>
</tr>
</tbody>
</table>
<div id="cell-103" class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>aln.parse_text(dfn_line, pattern, keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'aln_start_pos': '14370',
 'readid': '2591237:ncbi:1-40200',
 'readnb': '40200',
 'refseq_strand': '+',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L260" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="alnfilereader.parse_definition_line_with_position" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader.parse_definition_line_with_position">AlnFileReader.parse_definition_line_with_position</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader.parse_definition_line_with_position (dfn_line:str)</code></pre>
</blockquote>
<p><em>Parse definition line and adds relative position</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>dfn_line</td>
<td>str</td>
<td>fefinition line string to be parsed</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td><strong>parsed metadata in key/value format + relative position of the read</strong></td>
</tr>
</tbody>
</table>
<p>Upon instance creation, <a href="https://vtecftwy.github.io/metagentools/cnn_virus_data.html#alnfilereader"><code>AlnFileReader</code></a> automatically checks the <code>default_parsing_rules.json</code> file for a workable rule among saved rules. Saved rules include the rule for ART Illumina ALN files.</p>
<div id="cell-106" class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>aln.re_rule_name</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'aln_art_illumina_ncbi_std'</code></pre>
</div>
</div>
<p>It is therefore not required to pass a specific <code>pattern</code> and <code>keys</code> parameter.</p>
<div id="cell-108" class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>aln.parse_text(dfn_line)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'aln_start_pos': '14370',
 'readid': '2591237:ncbi:1-40200',
 'readnb': '40200',
 'refseq_strand': '+',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}</code></pre>
</div>
</div>
<p>ART Ilumina ALN files definition lines consist of:</p>
<ul>
<li>The <strong>read</strong> ID: <code>readid</code>, e.g.&nbsp;<code>2591237:ncbi:1-20100</code></li>
<li>the <strong>read</strong> number (order in the file): <code>readnb</code>, e.g.&nbsp;<code>20100</code></li>
<li>The <strong>read</strong> start position in the reference sequence: <code>aln_start_pos</code>, e.g.&nbsp;<code>23878</code></li>
<li>The <strong>reference sequence</strong> ID: <code>readid</code>, e.g.&nbsp;<code>2591237:ncbi:1-20100</code></li>
<li>The <strong>reference sequence</strong> number: <code>refseqnb</code>, e.g.&nbsp;<code>1</code></li>
<li>The <strong>reference sequence</strong> source: <code>refsource</code>, e.g.&nbsp;<code>ncbi</code></li>
<li>The <strong>reference sequence</strong> taxonomy: <code>reftaxonomyid</code>, e.g.&nbsp;<code>2591237</code></li>
<li>The <strong>reference sequence</strong> strand: <code>refseq_strand</code> wich is either <code>+</code> or <code>-</code>,</li>
</ul>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L272" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="alnfilereader.parse_file" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader.parse_file">AlnFileReader.parse_file</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader.parse_file (add_ref_seq_aligned:bool=False,
                           add_read_seq_aligned:bool=False)</code></pre>
</blockquote>
<p><em>Read ALN file, return a dict w/ alignment info for each read and optionaly aligned reference sequence &amp; read</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>add_ref_seq_aligned</td>
<td>bool</td>
<td>False</td>
<td>Add the reference sequence aligned to the parsed dictionary when True</td>
</tr>
<tr class="even">
<td>add_read_seq_aligned</td>
<td>bool</td>
<td>False</td>
<td>Add the read sequence aligned to the parsed dictionary when True</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict[str]</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-111" class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>parsed <span class="op">=</span> aln.parse_file()</span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (k, v) <span class="kw">in</span> <span class="bu">enumerate</span>(parsed.items()):</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(k)</span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>    pprint(v)</span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;</span> <span class="dv">3</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>2591237:ncbi:1-40200
{'aln_start_pos': '14370',
 'readid': '2591237:ncbi:1-40200',
 'readnb': '40200',
 'refseq_strand': '+',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40199
{'aln_start_pos': '15144',
 'readid': '2591237:ncbi:1-40199',
 'readnb': '40199',
 'refseq_strand': '-',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40198
{'aln_start_pos': '2971',
 'readid': '2591237:ncbi:1-40198',
 'readnb': '40198',
 'refseq_strand': '-',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40197
{'aln_start_pos': '15485',
 'readid': '2591237:ncbi:1-40197',
 'readnb': '40197',
 'refseq_strand': '-',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40196
{'aln_start_pos': '16221',
 'readid': '2591237:ncbi:1-40196',
 'readnb': '40196',
 'refseq_strand': '-',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L292" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="alnfilereader.parse_header_reference_sequences" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader.parse_header_reference_sequences">AlnFileReader.parse_header_reference_sequences</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader.parse_header_reference_sequences (pattern:str|None=None,
                                                 keys:list[str]|None=None)</code></pre>
</blockquote>
<p><em>Extract metadata from all header reference sequences</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pattern</td>
<td>str | None</td>
<td>None</td>
<td>regex pattern to apply to parse the reference sequence info</td>
</tr>
<tr class="even">
<td>keys</td>
<td>list[str] | None</td>
<td>None</td>
<td>list of keys: keys are both regex match group names and corresponding output dict keys</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict[str]</strong></td>
<td></td>
<td><strong>parsed metadata in key/value format</strong></td>
</tr>
</tbody>
</table>
<div id="cell-113" class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>pprint(aln.parse_header_reference_sequences())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'2591237:ncbi:1': {'organism': 'Coronavirus BtRs-BetaCoV/YN2018D',
                    'refseq_accession': 'MK211378',
                    'refseq_length': '30213',
                    'refseqid': '2591237:ncbi:1',
                    'refseqnb': '1',
                    'refsource': 'ncbi',
                    'reftaxonomyid': '2591237'}}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L307" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="alnfilereader.set_header_parsing_rules" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader.set_header_parsing_rules">AlnFileReader.set_header_parsing_rules</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader.set_header_parsing_rules (pattern:str|bool=None,
                                         keys:list[str]=None,
                                         verbose:bool=False)</code></pre>
</blockquote>
<p>*Set the regex parsing rule for reference sequence in ALN header.</p>
<p>Updates 3 class attributes: <code>re_header_rule_name</code>, <code>re_header_pattern</code>, <code>re_header_keys</code></p>
<p>TODO: refactor this and the method in Core: to use a single function for the common part and a parameter for the text_to_parse*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pattern</td>
<td>str | bool</td>
<td>None</td>
<td>regex pattern to apply to parse the text, search in parsing rules json if None</td>
</tr>
<tr class="even">
<td>keys</td>
<td>list[str]</td>
<td>None</td>
<td>list of keys/group for regex, search in parsing rules json if None</td>
</tr>
<tr class="odd">
<td>verbose</td>
<td>bool</td>
<td>False</td>
<td>when True, provides information on each rule</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-115" class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>aln.set_header_parsing_rules(verbose<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--------------------------------------------------------------------------------
Rule &lt;fasta_ncbi_std&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;fastq_art_illumina_ncbi_std&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;aln_art_illumina_ncbi_std&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;aln_art_illumina-refseq-ncbi-std&gt; generated 7 matches
--------------------------------------------------------------------------------
^@SQ[\t\s]*(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*))[\t\s]*(?P=refseqnb)[\t\s]*(?P&lt;refseq_accession&gt;[\w\d]*)[\t\s]*(?P=reftaxonomyid)[\t\s]*(?P=refsource)[\t\s](?P&lt;organism&gt;.*)[\t\s](?P&lt;refseq_length&gt;\d*)$
['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'refseq_accession', 'organism', 'refseq_length']
--------------------------------------------------------------------------------
Rule &lt;fasta_ncbi_cov&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Rule &lt;fasta_rhinolophus_ferrumequinum&gt; generated an error
No match on this line
--------------------------------------------------------------------------------
Selected rule with most matches: aln_art_illumina-refseq-ncbi-std</code></pre>
</div>
</div>
<div id="cell-116" class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(aln.re_header_rule_name)</span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(aln.re_header_pattern)</span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(aln.re_header_keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>aln_art_illumina-refseq-ncbi-std
^@SQ[\t\s]*(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*))[\t\s]*(?P=refseqnb)[\t\s]*(?P&lt;refseq_accession&gt;[\w\d]*)[\t\s]*(?P=reftaxonomyid)[\t\s]*(?P=refsource)[\t\s](?P&lt;organism&gt;.*)[\t\s](?P&lt;refseq_length&gt;\d*)$
['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'refseq_accession', 'organism', 'refseq_length']</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="build-datasets" class="level1">
<h1>Build datasets</h1>
<p>Sequence and reads are provided in various formats (text for original data, fastq + aln for simulated reads) and the model expects a specific format for training, validation and testing datasets.</p>
<p>The following functions allow to build the datasets in the format expected by the model from the raw data available.</p>
<p>In addition, text based dataset are not efficient, especially for training. Additional functions allow to save and parse dataset in TFRecord format.</p>
<p>There are two pipelines for building inference datasets: - via a text inference dataset, in the same format as the original paper’s data - via a TFRecord inference dataset for faster operations.</p>
<section id="text-based-inference-datasets" class="level2">
<h2 class="anchored" data-anchor-id="text-based-inference-datasets">Text based inference datasets</h2>
<p>In this pipeline, the steps are:</p>
<ol type="1">
<li>Create a text inference file and a metadata file from FASTQ and ALN with <a href="https://vtecftwy.github.io/metagentools/cnn_virus_data.html#create_infer_ds_from_fastq"><code>create_infer_ds_from_fastq</code></a></li>
<li>Create a <code>tf.data.TextLineDataset</code> from the text inference dataset</li>
<li>Transform it into an inference/training dataset with <code>.map</code> and <code>strings_to_tensors)</code></li>
</ol>
<div id="cell-121" class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> _map_read_to_label(read, refseq_meta, p2mapping<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">""""""</span></span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> p2mapping <span class="kw">is</span> <span class="va">None</span>:</span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a>        p2mapping <span class="op">=</span> PACKAGE_ROOT <span class="op">/</span> <span class="st">'data/ncbi/refsequences/taxonomyid-label-mapping.json'</span></span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">assert</span> p2mapping.is_file()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L389" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="create_infer_ds_from_fastq" class="level3">
<h3 class="anchored" data-anchor-id="create_infer_ds_from_fastq">create_infer_ds_from_fastq</h3>
<blockquote class="blockquote">
<pre><code> create_infer_ds_from_fastq (p2fastq:str|pathlib.Path,
                             output_dir:str|pathlib.Path|None=None,
                             overwrite_ds:bool=False,
                             nsamples:int|None=None)</code></pre>
</blockquote>
<p>*Build an inference dataset file as required by the CNN Virus model from a simreads fastq (ART format).</p>
<p>Also extract the fastq read sequence metadata, saves it in a metadata file and returns them as a DataFrame*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p2fastq</td>
<td>str | Path</td>
<td></td>
<td>Path to the fastq file (aln file path is inferred)</td>
</tr>
<tr class="even">
<td>output_dir</td>
<td>str | Path | None</td>
<td>None</td>
<td>Path to directory where ds file will be saved</td>
</tr>
<tr class="odd">
<td>overwrite_ds</td>
<td>bool</td>
<td>False</td>
<td>If True, overwrite existing ds file. If False, error is raised if ds file exists</td>
</tr>
<tr class="even">
<td>nsamples</td>
<td>int | None</td>
<td>None</td>
<td>Used to limit the number of reads to use for inference, use all if None</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>(Path, Path, pd.DataFrame)</strong></td>
<td></td>
<td><strong>Paths to dataset file, path to metadata file, dataframe with metadata</strong></td>
</tr>
</tbody>
</table>
<div id="cell-123" class="cell">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>path2ds, path2meta, meta <span class="op">=</span> create_infer_ds_from_fastq(</span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a>    p2fastq<span class="op">=</span>p2fastq, </span>
<span id="cb144-3"><a href="#cb144-3" aria-hidden="true" tabindex="-1"></a>    output_dir<span class="op">=</span>pfs.data <span class="op">/</span> <span class="st">'ncbi/ds/cov'</span>,</span>
<span id="cb144-4"><a href="#cb144-4" aria-hidden="true" tabindex="-1"></a>    overwrite_ds<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb144-5"><a href="#cb144-5" aria-hidden="true" tabindex="-1"></a>    nsamples<span class="op">=</span><span class="dv">100</span></span>
<span id="cb144-6"><a href="#cb144-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"3b7ed5c185f44f1c82931ff5225572b7","version_major":2,"version_minor":0,"quarto_mimetype":"application/vnd.jupyter.widget-view+json"}
</script>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Dataset with 100 reads</code></pre>
</div>
</div>
<div id="cell-124" class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"FASTAQ file name: </span><span class="sc">{</span>p2fastq<span class="sc">.</span>absolute()<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Path to dataset:  </span><span class="sc">{</span>path2ds<span class="sc">.</span>absolute()<span class="sc">}</span><span class="ss"> </span><span class="ch">\n</span><span class="ss">Path to metadata: </span><span class="sc">{</span>path2meta<span class="sc">.</span>absolute()<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>FASTAQ file name: /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/simreads/cov/single_1seq_150bp/single_1seq_150bp.fq
Path to dataset:  /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/ds/cov/single_1seq_150bp_ds 
Path to metadata: /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/ds/cov/single_1seq_150bp_metadata.csv</code></pre>
</div>
</div>
<div id="cell-125" class="cell">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>TextFileBaseReader(path2ds, nlines<span class="op">=</span><span class="dv">5</span>).print_first_chunks(nchunks<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>5-line chunk 1
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAGTTTCAAGGAACTTTTAGTGTATGCTGCTGATCCAGCCATGCATGCAGCTT  0   0
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGAGTCATTTGAGTTATAGTAGGGATGACATTACGCTTAGTATACGCGAAAAGTGCATCTTGATCCTCATAACTCATTGAGT  0   0
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTTAGAGGCTTAAATAACTGTCTCACTATTGAACTTATTATCATGTCAAGATTCCAAATAGCAATCCTGAAAG  0   0
TAATCACTGATAGCAGCATTGCCATCCTGAGCAAAGAAGAAGTGTTTTAGTTCAACAGAACTTCCTTCCTTAAAGAAACCTTTAGACACAGCAAAGTCATAAAAGTCTTTGTTAAAATTACCGGGTTTGACAGTTTGAAAAGCAACATTG  0   0
CTAATGTCAGTACGCCTACAATGCCTGCATCACGCATAGCATCGCAGAATTGTACAGTCTTTAATAATGCTTGGCGTACACGTTCACCTAAGTTAGCATATACGCGCAAGATGTCAGGATTCTCTACGAAGTCATACCAATCCTTCTTAT  0   0
</code></pre>
</div>
</div>
<div id="cell-126" class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>meta.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">read_ids</th>
<th data-quarto-table-cell-role="th">read_refseqs</th>
<th data-quarto-table-cell-role="th">read_start_pos</th>
<th data-quarto-table-cell-role="th">read_strand</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>2591237:ncbi:1-40200</td>
<td>2591237:ncbi:1</td>
<td>14370</td>
<td>+</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>2591237:ncbi:1-40199</td>
<td>2591237:ncbi:1</td>
<td>15144</td>
<td>-</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>2591237:ncbi:1-40198</td>
<td>2591237:ncbi:1</td>
<td>2971</td>
<td>-</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>2591237:ncbi:1-40197</td>
<td>2591237:ncbi:1</td>
<td>15485</td>
<td>-</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>2591237:ncbi:1-40196</td>
<td>2591237:ncbi:1</td>
<td>16221</td>
<td>-</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L457" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="strings_to_tensors" class="level3">
<h3 class="anchored" data-anchor-id="strings_to_tensors">strings_to_tensors</h3>
<blockquote class="blockquote">
<pre><code> strings_to_tensors (b:tensorflow.python.framework.tensor.Tensor)</code></pre>
</blockquote>
<p><em>Function converting a batch of bp strings into three tensors: (x_seqs, (y_labels, y_pos))</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>b</td>
<td>tf.Tensor</td>
<td>batch of strings</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tfrecord-based-inference-datasets" class="level2">
<h2 class="anchored" data-anchor-id="tfrecord-based-inference-datasets">TFRecord based inference datasets</h2>
<p>In this pipeline, the steps are:</p>
<ol type="1">
<li>Go from FASTQ and ALN to a RFRecord file and a metadata file with <a href="https://vtecftwy.github.io/metagentools/cnn_virus_data.html#tfrecord_from_fastq"><code>tfrecord_from_fastq</code></a> or <a href="https://vtecftwy.github.io/metagentools/cnn_virus_data.html#tfrecord_from_text"><code>tfrecord_from_text</code></a></li>
<li>Create a <code>tf.data.TFRecordDataset</code> from the saved TFRecord file</li>
<li>Transform it into an inference/training dataset with <code>.map</code> and <code>tfr_to_tensors</code></li>
</ol>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L583" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="split_kmer_into_50mers" class="level3">
<h3 class="anchored" data-anchor-id="split_kmer_into_50mers">split_kmer_into_50mers</h3>
<blockquote class="blockquote">
<pre><code> split_kmer_into_50mers (kmer:tensorflow.python.framework.tensor.Tensor)</code></pre>
</blockquote>
<p><em>Converts a k-mer read into several 50-mer reads, by shifting the k-mer one base at a time.</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>kmer</td>
<td>tf.Tensor</td>
<td>tensor representing a single k-mer read, after base</td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L601" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="base_string_kmers_to_tensors" class="level3">
<h3 class="anchored" data-anchor-id="base_string_kmers_to_tensors">base_string_kmers_to_tensors</h3>
<blockquote class="blockquote">
<pre><code> base_string_kmers_to_tensors
                               (b:tensorflow.python.framework.tensor.Tenso
                               r, k:int=50)</code></pre>
</blockquote>
<p><em>Function converting a batch of bp strings into three tensors: (x_seqs, (y_labels, y_pos))</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>b</td>
<td>tf.Tensor</td>
<td></td>
<td>batch of strings</td>
</tr>
<tr class="even">
<td>k</td>
<td>int</td>
<td>50</td>
<td>maximum read length in the batch</td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L661" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="split_kmer_batch_into_50mers" class="level3">
<h3 class="anchored" data-anchor-id="split_kmer_batch_into_50mers">split_kmer_batch_into_50mers</h3>
<blockquote class="blockquote">
<pre><code> split_kmer_batch_into_50mers
                               (kmer:tensorflow.python.framework.tensor.Te
                               nsor)</code></pre>
</blockquote>
<p>*Converts a batch of k-mer reads into several 50-mer reads, by shifting the k-mer one base at a time.</p>
<p>for a batch of <code>b</code> k-mer reads, returns a batch of <code>b - 49</code> 50-mer reads*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>kmer</td>
<td>tf.Tensor</td>
<td>tensor representing a batch of k-mer reads, after base encoding</td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L691" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="tfrecord_from_fastq" class="level3">
<h3 class="anchored" data-anchor-id="tfrecord_from_fastq">tfrecord_from_fastq</h3>
<blockquote class="blockquote">
<pre><code> tfrecord_from_fastq (p2fastq:pathlib.Path,
                      p2tfrds:pathlib.Path|None=None,
                      overwrite:bool=False)</code></pre>
</blockquote>
<p>*Creates a TFRecord dataset for inference from fastq and aln files, as well as a csv metadata file</p>
<p>The TFRecord dataset can be used for training or prediction, using the original CNN Virus model. The metadata file is a Pandas DataFrame converted into csv*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p2fastq</td>
<td>Path</td>
<td></td>
<td>Path to the fastaq file (should be associated with a aln file)</td>
</tr>
<tr class="even">
<td>p2tfrds</td>
<td>Path | None</td>
<td>None</td>
<td>Path to the TFRecord file, default creates a file in saved directory</td>
</tr>
<tr class="odd">
<td>overwrite</td>
<td>bool</td>
<td>False</td>
<td>When True, overides any existing file, When False, raises an error</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>(Path, Path)</strong></td>
<td></td>
<td><strong>Paths to the saved TFRecord file and the metadata csv file</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L757" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="tfrecord_from_text" class="level3">
<h3 class="anchored" data-anchor-id="tfrecord_from_text">tfrecord_from_text</h3>
<blockquote class="blockquote">
<pre><code> tfrecord_from_text (p2ds, p2tfrds:pathlib.Path|None=None,
                     overwrite:bool=False)</code></pre>
</blockquote>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p2ds</td>
<td></td>
<td></td>
<td>Path to the text dataset, in the format of original CNN Virus data</td>
</tr>
<tr class="even">
<td>p2tfrds</td>
<td>Path | None</td>
<td>None</td>
<td>Path to the TFRecord file, default creates a file in savec directory</td>
</tr>
<tr class="odd">
<td>overwrite</td>
<td>bool</td>
<td>False</td>
<td>When True, overides any existing file, When False, raises an error</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>Path</strong></td>
<td></td>
<td><strong>Path to the saved TFRecord file</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L815" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_dataset_from_tfr" class="level3">
<h3 class="anchored" data-anchor-id="get_dataset_from_tfr">get_dataset_from_tfr</h3>
<blockquote class="blockquote">
<pre><code> get_dataset_from_tfr (p2tfrds:pathlib.Path, batch_size:int=1)</code></pre>
</blockquote>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p2tfrds</td>
<td>Path</td>
<td></td>
<td>Path to the TFRecord dataset</td>
</tr>
<tr class="even">
<td>batch_size</td>
<td>int</td>
<td>1</td>
<td>Desired batch side for the dataset</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>tf.data.Dataset</strong></td>
<td></td>
<td><strong>dataset, formated with the batch size</strong></td>
</tr>
</tbody>
</table>
<p>Create a dataset from an existing TFRecord file</p>
<div id="cell-137" class="cell">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="co"># </span><span class="al">TODO</span><span class="co">: Check how to define shape of the elements</span></span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>p2ds <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/ds/cov/single_1seq_50bp-10reads.tfrecords'</span></span>
<span id="cb158-3"><a href="#cb158-3" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> p2ds.exists()</span>
<span id="cb158-4"><a href="#cb158-4" aria-hidden="true" tabindex="-1"></a>ds <span class="op">=</span> get_dataset_from_tfr(p2ds)</span>
<span id="cb158-5"><a href="#cb158-5" aria-hidden="true" tabindex="-1"></a>ds.element_spec</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>(TensorSpec(shape=&lt;unknown&gt;, dtype=tf.float32, name=None),
 (TensorSpec(shape=&lt;unknown&gt;, dtype=tf.float32, name=None),
  TensorSpec(shape=&lt;unknown&gt;, dtype=tf.float32, name=None)))</code></pre>
</div>
</div>
<div id="cell-138" class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r, (l, p) <span class="kw">in</span> ds.take(<span class="dv">8</span>):</span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(r.shape, l.shape, p.shape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(1, 50, 5) (1, 187) (1, 10)
(1, 50, 5) (1, 187) (1, 10)
(1, 50, 5) (1, 187) (1, 10)
(1, 50, 5) (1, 187) (1, 10)
(1, 50, 5) (1, 187) (1, 10)
(1, 50, 5) (1, 187) (1, 10)
(1, 50, 5) (1, 187) (1, 10)
(1, 50, 5) (1, 187) (1, 10)</code></pre>
</div>
</div>
<p>We can convert this dataset into a batch dataset</p>
<div id="cell-140" class="cell">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>ds_batched <span class="op">=</span> ds.batch(<span class="dv">2</span>)</span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r, (l, p) <span class="kw">in</span> ds_batched.take(<span class="dv">8</span>):</span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(r.shape, l.shape, p.shape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(2, 1, 50, 5) (2, 1, 187) (2, 1, 10)
(2, 1, 50, 5) (2, 1, 187) (2, 1, 10)
(2, 1, 50, 5) (2, 1, 187) (2, 1, 10)
(2, 1, 50, 5) (2, 1, 187) (2, 1, 10)
(2, 1, 50, 5) (2, 1, 187) (2, 1, 10)</code></pre>
</div>
</div>
</section>
</section>
<section id="postprocessing" class="level2">
<h2 class="anchored" data-anchor-id="postprocessing">Postprocessing</h2>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L826" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="combine_predictions" class="level3">
<h3 class="anchored" data-anchor-id="combine_predictions">combine_predictions</h3>
<blockquote class="blockquote">
<pre><code> combine_predictions (labels:tensorflow.python.framework.tensor.Tensor,
                      label_probs:tensorflow.python.framework.tensor.Tenso
                      r,
                      positions:tensorflow.python.framework.tensor.Tensor)</code></pre>
</blockquote>
<p><em>Combine set of 50-mer predictions into one final prediction for label and position</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>labels</td>
<td>tf.Tensor</td>
<td>Label predictions for a set of 50-mer corresponding to a single k-mer</td>
</tr>
<tr class="even">
<td>label_probs</td>
<td>tf.Tensor</td>
<td>Probabilities for the labels</td>
</tr>
<tr class="odd">
<td>positions</td>
<td>tf.Tensor</td>
<td>Position predictions for a set of 50-mer corresponding to a single k-mer</td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L854" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="combine_prediction_batch" class="level3">
<h3 class="anchored" data-anchor-id="combine_prediction_batch">combine_prediction_batch</h3>
<blockquote class="blockquote">
<pre><code> combine_prediction_batch (probs_elements:tuple[tensorflow.python.framewor
                           k.tensor.Tensor,tensorflow.python.framework.ten
                           sor.Tensor])</code></pre>
</blockquote>
<p>*Combine a batch of 50-mer probabilities into one batch of final prediction for label and position</p>
<p>Note: the input must be reshape to (batch_size, k, n) where n is the number of labels or positions*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>probs_elements</td>
<td>tuple[tf.Tensor, tf.Tensor]</td>
<td>Tuple of label and position probabilities for a batch of 50-mer</td>
</tr>
</tbody>
</table>
</section>
</section>
</section>
<section id="original-code-refactored" class="level1">
<h1>Original Code (Refactored)</h1>
<p>Selected classes and functions refactored from the original code, coming with the paper</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L895" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="datagenerator_from_50mer" class="level3">
<h3 class="anchored" data-anchor-id="datagenerator_from_50mer">DataGenerator_from_50mer</h3>
<blockquote class="blockquote">
<pre><code> DataGenerator_from_50mer (f_matrix, f_labels, f_pos, batch_size=1024,
                           n_classes=187, shuffle=True)</code></pre>
</blockquote>
<p><em>data generator for generating batches of data from 50-mers</em></p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L955" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_params_150mer" class="level3">
<h3 class="anchored" data-anchor-id="get_params_150mer">get_params_150mer</h3>
<blockquote class="blockquote">
<pre><code> get_params_150mer ()</code></pre>
</blockquote>
<p><em>set default params for generating batches of 150-mer</em></p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L948" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_params_50mer" class="level3">
<h3 class="anchored" data-anchor-id="get_params_50mer">get_params_50mer</h3>
<blockquote class="blockquote">
<pre><code> get_params_50mer ()</code></pre>
</blockquote>
<p><em>set default params for generating batches of 50-mer</em></p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L939" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_learning_weights" class="level3">
<h3 class="anchored" data-anchor-id="get_learning_weights">get_learning_weights</h3>
<blockquote class="blockquote">
<pre><code> get_learning_weights (filepath)</code></pre>
</blockquote>
<p><em>get different learning weights for different classes, from file</em></p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L962" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_kmer_from_50mer" class="level3">
<h3 class="anchored" data-anchor-id="get_kmer_from_50mer">get_kmer_from_50mer</h3>
<blockquote class="blockquote">
<pre><code> get_kmer_from_50mer (filepath, max_seqs=None)</code></pre>
</blockquote>
<p><em>Load data from sequence file and returns three tensors, with max nbr sequences</em></p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L985" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_kmer_from_150mer" class="level3">
<h3 class="anchored" data-anchor-id="get_kmer_from_150mer">get_kmer_from_150mer</h3>
<blockquote class="blockquote">
<pre><code> get_kmer_from_150mer (filepath, max_seqs=None)</code></pre>
</blockquote>
<p><em>Load data from sequence file and returns three tensors, with max nbr sequences</em></p>
<div id="cell-152" class="cell">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co"># path for the training file</span></span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a>filepath_50mer<span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'CNN_Virus_data/50mer_ds_100_seq'</span></span>
<span id="cb172-3"><a href="#cb172-3" aria-hidden="true" tabindex="-1"></a>filepath_150mer<span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'CNN_Virus_data/150mer_ds_100_seq'</span></span>
<span id="cb172-4"><a href="#cb172-4" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> filepath_50mer.is_file(), filepath_50mer</span>
<span id="cb172-5"><a href="#cb172-5" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> filepath_150mer.is_file(), filepath_150mer</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-153" class="cell">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>f_matrix,f_labels,f_pos <span class="op">=</span> get_kmer_from_50mer(filepath_50mer, max_seqs<span class="op">=</span><span class="dv">5</span>)</span>
<span id="cb173-2"><a href="#cb173-2" aria-hidden="true" tabindex="-1"></a>f_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['AAAAAGATTTTGAGAGAGGTCGACCTGTCCTCCTAAAACGTTTACAAAAG',
 'CATGTAACGCAGCTTAGTCCGATCGTGGCTATAATCCGTCTTTCGATTTG',
 'AACAACATCTTGTTGATGATAACCGTCAAAGTGTTTTGGGTCTGGAGGGA',
 'AGTACCTGGAGAGCGTTAAGAAACACAAACGGCTGGATGTAGTGCCGCGC',
 'CCACGTCGATGAAGCTCCGACGAGAGTCGGCGCTGAGCCCGCGCACCTCC']</code></pre>
</div>
</div>
<div id="cell-154" class="cell">
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a>f_labels, f_pos</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>(['71', '1', '158', '6', '71'], ['0', '7', '6', '7', '6'])</code></pre>
</div>
</div>
<div id="cell-155" class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a>f_matrix,f_labels,f_pos <span class="op">=</span> get_kmer_from_150mer(filepath_150mer, max_seqs<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a>f_matrix[:<span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['CTACATGACCCTGACACTCAGCTACGAGATGTCAAATTTTGGGGGCAATG',
 'TACATGACCCTGACACTCAGCTACGAGATGTCAAATTTTGGGGGCAATGA',
 'ACATGACCCTGACACTCAGCTACGAGATGTCAAATTTTGGGGGCAATGAA',
 'CATGACCCTGACACTCAGCTACGAGATGTCAAATTTTGGGGGCAATGAAA',
 'ATGACCCTGACACTCAGCTACGAGATGTCAAATTTTGGGGGCAATGAAAG']</code></pre>
</div>
</div>
<div id="cell-156" class="cell">
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a>f_labels[:<span class="dv">5</span>], f_pos[:<span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>(['120', '120', '120', '120', '120'], ['3', '3', '3', '3', '3'])</code></pre>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script type="application/vnd.jupyter.widget-state+json">
{"26b205197a934fdeabb71e65ac11acba":{"model_module":"@jupyter-widgets/base","model_module_version":"1.2.0","model_name":"LayoutModel","state":{"_model_module":"@jupyter-widgets/base","_model_module_version":"1.2.0","_model_name":"LayoutModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"LayoutView","align_content":null,"align_items":null,"align_self":null,"border":null,"bottom":null,"display":null,"flex":null,"flex_flow":null,"grid_area":null,"grid_auto_columns":null,"grid_auto_flow":null,"grid_auto_rows":null,"grid_column":null,"grid_gap":null,"grid_row":null,"grid_template_areas":null,"grid_template_columns":null,"grid_template_rows":null,"height":null,"justify_content":null,"justify_items":null,"left":null,"margin":null,"max_height":null,"max_width":null,"min_height":null,"min_width":null,"object_fit":null,"object_position":null,"order":null,"overflow":null,"overflow_x":null,"overflow_y":null,"padding":null,"right":null,"top":null,"visibility":null,"width":null}},"514ad0bfcabf4df580a9a872af814af9":{"model_module":"@jupyter-widgets/controls","model_module_version":"1.5.0","model_name":"ProgressStyleModel","state":{"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"ProgressStyleModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"StyleView","bar_color":null,"description_width":""}},"55646397fc9349d3af9e98b1f2b26f5d":{"model_module":"@jupyter-widgets/base","model_module_version":"1.2.0","model_name":"LayoutModel","state":{"_model_module":"@jupyter-widgets/base","_model_module_version":"1.2.0","_model_name":"LayoutModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"LayoutView","align_content":null,"align_items":null,"align_self":null,"border":null,"bottom":null,"display":null,"flex":null,"flex_flow":null,"grid_area":null,"grid_auto_columns":null,"grid_auto_flow":null,"grid_auto_rows":null,"grid_column":null,"grid_gap":null,"grid_row":null,"grid_template_areas":null,"grid_template_columns":null,"grid_template_rows":null,"height":null,"justify_content":null,"justify_items":null,"left":null,"margin":null,"max_height":null,"max_width":null,"min_height":null,"min_width":null,"object_fit":null,"object_position":null,"order":null,"overflow":null,"overflow_x":null,"overflow_y":null,"padding":null,"right":null,"top":null,"visibility":null,"width":null}},"70f6be4247664b708b662c34e7abe3ee":{"model_module":"@jupyter-widgets/base","model_module_version":"1.2.0","model_name":"LayoutModel","state":{"_model_module":"@jupyter-widgets/base","_model_module_version":"1.2.0","_model_name":"LayoutModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"LayoutView","align_content":null,"align_items":null,"align_self":null,"border":null,"bottom":null,"display":null,"flex":null,"flex_flow":null,"grid_area":null,"grid_auto_columns":null,"grid_auto_flow":null,"grid_auto_rows":null,"grid_column":null,"grid_gap":null,"grid_row":null,"grid_template_areas":null,"grid_template_columns":null,"grid_template_rows":null,"height":null,"justify_content":null,"justify_items":null,"left":null,"margin":null,"max_height":null,"max_width":null,"min_height":null,"min_width":null,"object_fit":null,"object_position":null,"order":null,"overflow":null,"overflow_x":null,"overflow_y":null,"padding":null,"right":null,"top":null,"visibility":null,"width":null}},"7179c6cc207941648c348b1bf10cb87f":{"model_module":"@jupyter-widgets/controls","model_module_version":"1.5.0","model_name":"LabelModel","state":{"_dom_classes":[],"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"LabelModel","_view_count":null,"_view_module":"@jupyter-widgets/controls","_view_module_version":"1.5.0","_view_name":"LabelView","description":"","description_tooltip":null,"layout":"IPY_MODEL_70f6be4247664b708b662c34e7abe3ee","placeholder":"​","style":"IPY_MODEL_bdde467d943148ce9bb6355fd7582a5c","value":"0.078 MB of 0.078 MB uploaded (0.020 MB deduped)\r"}},"7849f255e99b4853bdba7f8badf1054a":{"model_module":"@jupyter-widgets/controls","model_module_version":"1.5.0","model_name":"VBoxModel","state":{"_dom_classes":[],"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"VBoxModel","_view_count":null,"_view_module":"@jupyter-widgets/controls","_view_module_version":"1.5.0","_view_name":"VBoxView","box_style":"","children":["IPY_MODEL_7179c6cc207941648c348b1bf10cb87f","IPY_MODEL_e0819a1ddcc64c08a748a2fd88350f09"],"layout":"IPY_MODEL_26b205197a934fdeabb71e65ac11acba"}},"bdde467d943148ce9bb6355fd7582a5c":{"model_module":"@jupyter-widgets/controls","model_module_version":"1.5.0","model_name":"DescriptionStyleModel","state":{"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"DescriptionStyleModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"StyleView","description_width":""}},"e0819a1ddcc64c08a748a2fd88350f09":{"model_module":"@jupyter-widgets/controls","model_module_version":"1.5.0","model_name":"FloatProgressModel","state":{"_dom_classes":[],"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"FloatProgressModel","_view_count":null,"_view_module":"@jupyter-widgets/controls","_view_module_version":"1.5.0","_view_name":"ProgressView","bar_style":"","description":"","description_tooltip":null,"layout":"IPY_MODEL_55646397fc9349d3af9e98b1f2b26f5d","max":1,"min":0,"orientation":"horizontal","style":"IPY_MODEL_514ad0bfcabf4df580a9a872af814af9","value":1}},"state":{},"version_major":2,"version_minor":0}
</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/vtecftwy\.github\.io\/metagentools");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/vtecftwy/metagentools/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>