{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#|default_exp cnn_virus.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#|hide\n",
    "from __future__ import annotations\n",
    "from ecutilities.ipython import nb_setup\n",
    "from fastcore.test import test_fail\n",
    "from nbdev import show_doc, nbdev_export\n",
    "from pprint import pprint\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set autoreload mode\n"
     ]
    }
   ],
   "source": [
    "#| hide\n",
    "nb_setup()\n",
    "\n",
    "# ON_COLAB, p2dataroot, p2data = setup_nb(_dev=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#|export\n",
    "from ecutilities.core import validate_path\n",
    "from functools import partial, partialmethod\n",
    "from metagentools.bio import q_score2prob_error\n",
    "from metagentools.core import TextFileBaseReader, JsonFileReader\n",
    "from pathlib import Path\n",
    "from typing import Any, Optional\n",
    "\n",
    "import json\n",
    "import numpy as np\n",
    "import re\n",
    "import tensorflow as tf\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tensorflow version: 2.8.2 - Expected 2.8.2\n"
     ]
    }
   ],
   "source": [
    "#|hide\n",
    "print(f\"Tensorflow version: {tf.__version__} - Expected 2.8.2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# data\n",
    "\n",
    "> Data preprocessing functions, data loaders, ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data structure for CNN Virus project"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following diagram shows the different types of data from raw data to processed data to data used by the CNN Virus model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "# TODO: update the diagram: (fq + aln); ds + read_info"
   ]
  },
  {
   "attachments": {
    "image-3.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAA4sAAAG2CAYAAADWaBM5AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAAEnQAABJ0Ad5mH3gAADL/SURBVHhe7d1drB1Vgi/2cmNoDNjQoOlujz0e2906upFa9ijDCJnRDA9W2o/gCUhRmLkSyFG4NxICZ4SUTEAto3lBE7D6ZYhkYSUj8oIV0488+IFIg4UuV7q2Jspta2Jbbri+9Ac92NBANx/hX651KGrtfXyOfT72Ofv3k0q1q3btj3N27bXWv9aq2uu++FIDAAAAPd/o5gAAADBLWAQAAKAiLAIAAFARFgEAAKgIiwAAAFSERQAAACojfzrjzJkz3S2myczMTHcLmEbK/umk7AeG1AfTZ1xdMDYsbt68uVtiGly8eFGDAaacsn/6KPuBUdQH02WuusAwVAAAACrCIgAAABVhEQAAgIqwCAAAQEVYBAAAoCIsAgAAUBEWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERYBAACoCIsAAABUhEUAAAAqwiIAAAAVYREAAICKsAgAAEBFWAQAAKCy7osvdbdnnTlzptm8eXO3xDS4ePFiMzMz0y2tfdnHYSGm4fuh7J8+01b2A/OjPpguc9UFwiKtaQyL17uPX758uZ1v3LixnTN5Fuszmpbvh7J/+kx7WMw+DwsxLd8X9cF0ERa5KmFx4YTFyScsLoyyf/oIi+qCaaAuWDj1wXSZa992ziIAAAAVYREAAICKsAgAAEBFWAQAAKAiLAIAAFARFgEAAKgIiwAAAFSERQAm1smTJ5tNmzY1x44d69asTnn/+TvOnz/frQFguZU6JdP+/fu7tcxlosPi7t27Zz/QTAcPHuzuWR7D11fJAyy+EqTKtFRlff81lvq1AKbZsFyflDL3scceaw4cONBcunSpOX78eLd28SWI9v/ufjDN/yDrkjOGDh8+3N5XJHuU5xh10DTr85ilNJFhsaT+vXv3th9mmWK5drJ8gDt37px97dOnTzf3339/dy8AiyGV36OPPtq89tprXyvrS6W4Z8+edt2DDz7YLl+vQ4cOzb5OXvPIkSMCI8ASSfu5lLmZUub2w9B8lFyQ+fVI8Dp37lxz7733dmsWXwl30f+777vvvqonM+9lVAAc59lnn+1uLa+JDItJ/QmKzz//fLfmiiwP1y2FsjM9/PDD3Zqm2b59e3Pq1KluCYDF8MYbbzQ7duxoQ2GRcn6xwuFc8pqpa06cONGtAWApJTjFWh0C+uSTT7Z12rDX8oknnvjaumyTHs4cLJ2PHOhMNlnqXsRRJi4sJmHnn/HUU091a8Yr3bhlGnbxjurezTbz3UHTiJlL3uu41y/6Q1lzf3kMAFekzB83zL8cpS1HX8tyjjD3y9dSgaacLeuutcdwWLcMK+f+OS+ZxtU1c90f/edYiQYAwEpI8MlBun653y/PM5VexJTH+/bta29nnvv6ZXv/MZnG1SUpY3ft2tXeTkDLtqVeGZbpw+fp1zvl/lFldu7P3/X00093a+b2+OOPt/P5lP9bt25tw+UzzzzTrVk+ExcW33777Xa+efPmdj5OKuJ0Zfe7eM+ePTtbKT/00ENtA6TsbEU+xH6P4SjpRcwHUrrKR+14+WCzsw1fvx8Yy+1yf7qg53sEAWAalMoylfhCwl0aDS+++GJbtr700kttBZryOuVs1pUhpqUxMEruS53wyCOPdGuulNupO0q5nSFUee5SmadOyeiXcn+m6L/33M7zlvvTcBiW/amrynkzmV5//fWxjRyAteSee+5p52+99VY7T3lYyvNMCZMlIGakScrzKKcrlFGGKfP7w1xTpo47ZSw9e9k2Umdk+4xgST2Q1yrrMuX1UycNM0S2K+8hzzf0zjvvtPMtW7a086speSN1zHzK/1JfLqSuXAwTFxYvXLjQzvMPHKck97LzFNnRSkDM8KJ08b7yyivdvV+dAzOf4U39nTM7zDA05oPNjtWXBkHeV7bLlNv9owvZsbIDAnBFyvpUvBkOWg7QzaciTPlbhq6WMj3la6nASx0wHCFSQmWmBLg0HspjSt3ywgsvtMtRKvOjR4+2y3ne4SkJee+pe4r8Hf2yPu9vWF8MZXjSXPUewFqVMrWU51HC5NUCVOqOfrmZcxH7ZfF85DzAlPH9bJA6YZghIuV6/30OzbfDq68E3x//+MftfC75W/MeUsfMJ1wulok8Z/FqSnIffhhludyfo8Wp+Is0GrJDzFd2iOyI5UhEQmMJglG6sfsNj6IcLbn77rvbOQDjJSylvE0ZnYpwqY6cpqLN62SKnF9SlLqjHCAsU95PXw48Du/PyJIoR6JLY2ecHEjM4/L4uXo/Adaqfg9cv0wtvYoXL15s5+OkPd5/XGmHLyRIJVxu27atW/pKLnI5DJ4ZCroUSgAc9mSOUg5u9uuupTZxYbFcoWg+/7CreeCBB2Z7GiMfRIanLlSSfAmMr776ajuPfpd1f3J0GODa5ChrCYxLLWV4DigO65v+sKYyld7EcgpCGYqUaSEHIYscxc5j00goBx6X80gxwEp5880323k6eUrgSzlaytThyMFRUm7nwF7/AODVRnAsteHw2vkqPZnPPfdct2Zu4+qupTJxYbH0xA27fvvKkYjhEYeyXO5PaMvwoOyU+Yfmg5ir+3g+clShhMHS3TzKuPdYhtkCMF7K66WWwNavoMeV230pw1OvjKtLhiNcinH1RRoJaeRE/2AkwFqVYf0pR9OeLuVtOR9vlFHDOksZW3rarlXqgFFt84wWWWg9VOqFl19+uZ0vRDmVLeevX82w7lpqExcWs+MkMeeo8nAYUo7oZl0+jPyTcpGBvnRbDyvxXMwmO2XCZ7mIQTmKMW6YU16nf6GaKN29ZUxzjoDk3Jd+qs/t8rhR7zHPOzxanvfgiDIwrVL+DcvRlJP9i84spXK6QsrgUm6XIVBFyu5MkeFKZfvIENJ+uV4OUvZPS8g2qS+KPLZ/hdTy9y/VECeASZEyP6P+ys9IlCDYP1g2LIOL0iMZ5eBeGcafcrVf7s5X6oCU4f16KG3zvMe5Auw46RVNHTHMEXmfw3V9yRepO/LY+ch1Wua77fWayHMW8w/LMKB8eNmpypTkX04EzZCgjCfu358AN/xdkzxXPvA8V4alzkfpDu4/d44wlKO/kfeRru/s0GWbBMP+6+c95rXL/Xn/eQwAV6Ss75ejuZ0Dhtd7tHi+Sr1QGiopt1Nhl/dTyu7yfjLP/eW8xlwcYViupx7o1yE5ytwfVpVAmQZKuT9/c56jHIwEWCuG54CXYaNF6STqX3xsOJw02+RxZZscvMvBvawrw/jzOsPHzUfK9DyuXw8lM+Q95nUXKu8rj02QK8+XKfXAMKMM9S+udjV5nWs5BeJarPviS93tWWfOnBnZ5cv1yw6enb3/RZkEGQYwMzPTLa19i7GPX758uZ1v3LixnTN5Fuszmpbvh7J/+kxb2T+kLpgO6oKFUx9Ml7n27VV5NVQAAACWlrAIAABARVhcZhkbPWlDUAEAAIaERQAAACrCIgAAABVhEQAAgIqwCAAAQEVYBAAAoCIsAgAAUBEWr8OxY8ea3bt3d0ujZZtNmza108GDB7u1483nOQFYWSdPnmzL9bmUbTLt37+/Wzu3w4cPz3tbAJZP2vELaaOfP3++Lf/Ttl/NVjQs9oNUfxqGqiyPWl8Mn6dsl0q3v37UVJTnWMhO8OyzzzaPPPJItzTao48+2hw6dKj9bcXnn3++Wzvegw8+2Jw7d27V71gAoyQIDcvhTAlWfePWF8PnKdulDO+vH079emRUHTGunhl67rnnmgMHDnRLoz322GPtNin/jx8/3q2d2wMPPNCcOHFi7N8NAMtpInoWT58+3VamZRqGqiNHjjR79+5t50MJVQlkr7322uzjy/ryA/j99SW49dfFyy+/3L5Ggtp8Kulsk21TsY9Tnueee+5p5/OVxkXeD0ySNM7TuIbrlbK2Xw5n2rNnT3fvlfJ7x44d7fTKK690a7+SffHs2bNfe/y+ffva+06dOjW77qWXXmrX9euYUr8kVD7zzDNV/ZN65moHDXO0OIHuoYce6tbUsk3qiHvvvbdbMz/bt29v/z+j/m5Ya3KAxoERmGwTPwy1NBpeeOGF2eW+N954o72/39BIYyA9dPNVKv68xrjGyVC2ybap2BdbGhd5P3lfANMmB8syaiPTqIOEKR+HozoS9OYrBz0S5BIUh2V41uW+uQ6MvPrqq+28X+8spvvuu2/k3w2LLW2qBLbV0N5wwBJWxsSHxdJoKEdbR/W4pWK/noIuFX+eO68xrnEylMZKHjNOhjKVI92Z94+eDYdPDQPw3Xff3c7feuutds61+6/+x39s/v4n57ol+vr7YL8Czv6YfTT7cLkvPS3Z59MTM9ye1eU/v/dxd2sylYN3GbVRRm4My8h4/fXXu1sLd/To0dkyf6jUNdlmnLz2XOV/vh+7du1qb2fkS74z5W8o36sy5e/NvN+7Ukaj6HFhqeXAeg60jPouwCTrH+jol6ml3OyvG9Yhw8dkGlXe9u8fd4rCsExfi+2jiQ6LpdFQKs4cbc1y1hePP/54O0/FPO6DvJo0CvLcUV5rVOOkLwF127Zt3VItvZsZGhtliGyOQqcR/vDDD88OecowqTQm+n9TKbTffvvtds71+b/+7/8kNA4k/JXhd9k/EwL7+2C+Z9m/c3+Gc2doXxrHZRh31rE6vfveJxP9fcjBuzJqI1NuDw8SZj/MPpqKeaGBKvt5yu9S5o+S18w242QIbLYZJ9+PfL8iZXy+M2mUpxGRg5H9714JlX2bN29u5++88047Z+3oNyzTHojsk1kujd/+fcXw/qsNle7L9v3vSXm9zLM+t4u8v3KAsDwuy6PeT/85y/aZ5mqL9f/+UY3qrCv3ZyrGHbAs779M/ffE3D746NPu1uqWMjTlaaacxlU6aErbO/VFv52dfSSP6Z+WlnI6j+u3/bPPpd1TtolhZ1K+F6kryjYp27OPjtq3V7OJCIv50Ppf9vJhlUZDGepTjjKXIUCRxkQ+oHyg+RDz+IWExuw0+aDLc+e1RjVO+soOt3Xr1na+ELnIQX+IbOlFvHjxYjsv8h4uXLjQLS29X13+rO1xWM7pR0f/Y9toXcqpT2j8SsJfOShRvl/9fTD7n0D4lf/hf/tFtW8tx/SP//Srkd+d65k++PhKA2Elvw8l6PWnIgfv+kNMc3t4kLAfxoYjNxbDXAcC42oHC8fJ35bGTP+7Vw4q9pX7HTBcPH/5t281r/27d7ullZEGZPbl0rBMOdvfb9PuKPdlu+FB61xUr9wf822Qpn3UP72mP5pqlDR2X3zxxfZ1Sv0wl7yPnTt3zr63cRfzy9+TdlrZbtjGyf1ZV+7PeyztuVEHLFMm5CJSZfvcl2Xm5/9758Pmsef/w4p/L65XqQuinEeecnWYHcpovVycLPtSv42TdnnWlbZ/yQZPPfVUuxzZr/sXNcs2+Z6W0+Qi36lsM9fIlNVo3Rdf6m7POnPmzOyRzaWUgiFpPx/0qEKrpPp+wZMUn6O6KThGScGSwigf1rDASoMihUl/B8n2+bD7z5eCL4XluPeVHSQNlByJmOvcyLJdf6eN/F3Do9bD58o2KXznewW96/UXP/rqy7aW3bZhffPDP/l280dbP25+MLPwsN93+fLldr5x48Z2vpqUfbOv7Kf5XqZRMvyO5buX3pjVFCIX6zOahu/HX/z57zd7/4vfLXnZX3ooRpVto8rMNAjLUeBx+14Jm8OydlQdM5/nK/VIGqBD83l8lO1K2T5cLsbVE/mbRtVji20S9u0fPfKvmj/9wV3d0tLIgZH4zp3fbP7qh3/Q7PuT77TLi9HemW85k3r96aefrtoNZd/o76fZB3NAouxj2R/6+0jaKQlW89k/yveg7M/991H2v3JfCWf9581rZeh1/zvbfz+lzTTq+9I3qg4Z/l19w9e9Wh00/Fv6Vntd8OLBP2q+t+XWbmlxnPrn95u//vt/am/nuff/2ebZ70UsVxa4VqPK97nqkFL2jsoW0c8D4/bp/jbl9UfJgaBsM67cn0TpMJiZmemWvm5iw2L5wMcZV7jEuIo+hdKwgs+6ccY1Bub74Y/aafN62UlL4TfuucbtzEvlv/tf/59m/Y3ru6Xl9W/u37FkDYXSQCghMYXhd++8eVkbCJOm7HPD/bIs53spLH7dXIXoUkiv+z//pw+6pcX14UefVcOPdn/v9mb/n29ufu+mX61oWCxl9yil8h0n+/AwXI2rY65Wvs71HmNUXTI0LNvHlfWj6omYz2sshkkIiymfb91wQ7e0NDL8uq+Exh23v7+sYTE9dsO2S9k3hmExyj7aL6NjVICbS3l89APVMGANXzeuFhajNK5jVJsu8r3KaTj9/X/4PNkmjfGi314aVQf1X7dYi2Fxqb4jw+9FPzQKi/MLi+P29xhX7k+iVRkW+x/IUAqXuY64jntsHtevfOd6/RRKc/VgDp9rlOFOW3aauXbiYj7Pv5iWuzG8XPb/L29+LSQW0xYW+9+JYcOkfA/KfpnlUWExzxHjvneTaLWGxaXUP5pcQmI5WLMcZf9cQWxcuTfcR0cZ9dhxZXz25YTSUWV/KbfnqtznczBvVNk+13ucT72wFFZy385BkQy1XglpfH/v929t/ut7bli2USbZbxbSsxhlH8u+099HFhoWy/Olt7LfIzmfsJh9NMPz5gqLRd5XhuCNajuNCnv958lrZ9RVeZ3h3zh8fPnujAu+fau1Lsgw0QwXXS75XqS9lHbTpV9eWJNhcVwd1F8/ru7INiUbjHqtoeUsy6/XXPv2xF7gJhX5I4NLoxeppNPwjRQ0+cCK3J7rsX0p/PJcw8ZCpEBKoZUPepRR5xSmoMv7GfeY8jr9cwfuv//+7tZXyuOv5ZxIvu7vD+5uey77QXHaZT9MgzUFWPbXfA+yP19NzgXIdyuPSUHK6pWQmKF/f/dvf7Dkw//mq+xT5fySvnJu95tvvtlul0Z3X2ngjnrsKKUhnO9AX8reVP6pF+aq2HOKwPBUgjRsh/XRUA5y5mh1KeOzbRokQ6m0Y8uWLe18rfrr/+b7zT/8zR8vy1SkMZz9/+/+zQ/a/f+ujUvXo5l9td8myH6VA3FF9tu59pfFVMrvBLm5fh90lOyHaXOVv6N830bpt1vyt/W/E2lX9XsBS+O8r18XDc/7GtXu6m+/Fn+bNPvpqP15MaZ8/4p8L9LTnjbTv963bU23mdK7nf25345J+Z115RzFUv4/+eST7Tyy35fsEQmI2f9SZ/TluTKtJRMZFudqNEQ+6FTUKYCS+vNBpUDKlNtJ8MMj00Mp9PKh57lGKa/dv5hOXz+wLkSOQJQGd6YcZRwqJ+GWBhLXTki8Io3j/pHefD9y9DVTjqLlvnJkLIXkqKPCub88ZtKPkDHe7u9/2VCeoJBYzHXwLutyXxqP2fdyMLCUoZlSpma/HPXYcbJ9nrP/PAmPCXTDhsRQGr3XUv7ne5jXLAdqcjGO1AlDCcUx7mj1WpEGasro5ZiGIXGxz/+aj3z+OdBQ9rdYrs+4NGzL7YXI9vlelP02vZP9kJbQV/6mHPz4yU9+0t3zdal3+t+5tL/6z5P/T799lG37hgcsUxYM/59rzZJ+R7515XsxLSGxyH6Tcjf7atl3chAjeaL/3chyyvmyTfb7HGjvS1tpWI/kgMbVMshqs6LDUFezBNUE02EX9WLI0Yv+UIzlsJaG2c3HtA1DnVaGoS6Msv+KNETTkBg1JDZysDEN57mGH81XqUv6z5XGdxrRpQd0KU3Lvp2hfKMCorpgOqgLajl3PdO4gKg+mC5z7dsT/6P8kyqVeirzcT2P1yNHzsb1eAKwtHLkOT2POdo8qoex9HQuxbC3MuploUMFmdtK9CTCJCu9lnA1wuJ1yBDSxf4tlTRMEkIN8wNYWXMNuc65LTmwt9jKb+Bdb48l06M/BG44rbVzp4DlZxgqLcNQF87Qo8ln6NHCKPunz7SV/UPqgumgLlg49cF0MQwVAACABREWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERYBAACoCIsAAABUhEUAAAAqwiIAAAAVYREAAICKsAgAAEBFWAQAAKAiLAIAAFARFgEAAKgIiwAAAFSERQAAACrCIgAAABVhEQAAgIqwCAAAQEVYBAAAoCIsAgAAUBEWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWFxDfrss8+a3/72t83HH3/cfPTRR81vfvOb5sMPP2w++OCDdrp8+XJz6dKl5v3332/+5V/+pXnvvfe6RwKwWin7Afj888+b3/3ud4tWF6z74kvd7VlnzpxpNm/e3C0xadIgGE7ZMT799NN2uuGGG5r169e362+55ZZm3bp1zTe+8Y3ZebndX5+d6o477uheYe3LPg4LMTMz091au5T9k21Y7mdS9l8fdQELNQ11QagPJlcp9/v1QLmd9Snbb7zxxrY+KGV/maLcnm9dICxOuHJ0oD/lQ80OkB3hpptumr2dKY2FsjMALISyf3Io+4GVpD6YDIlpw7og9cOwDuhPqSsWk7A4YTKEKFOODJQd4uabb26nb37zm+08OwLAYlP2rxxlPzBJ1AcrI3XAJ598MlsXZF7qgA0bNrS3ExSXk7C4wtIgyE6RHSJji3N0OMOHyk6hcQAsF2X/8lH2A5NMfbB8UhdkygHDxLJSF5RppQmLKyCNg7JTZErDYOPGje3OoYEArBRl/9JS9gOrhfpg6eTcwlIXZJ6ewltvvbW57bbb2p7DSSMsLpPsGOWqRJGdokwAk0DZv/iU/cBqpD5YfKkHEg4z5SBhwmHqgowsmWTC4hLKvzaNhEw5epAjyOUoMsCkUfYvDmU/sNqpDxZH6oCExEypA0p9sNgXoVlKwuISGO4YmzZtancMgEmm7L8+yn5grVAfXLuMKEk9kAOGuUp16oHUB7l66WokLC6i0kjIhQtuv/32dpr0rmWAQtl/bZT9wFqjPli4csAwP4Jf6oJJuEDN9fKjTNcpRw8++OCD5uc//3l78YK77rqr2bFjR3PnnXdqLACsUcp+ACIB8b333msuXbrUjir53ve+13znO99ZE0ExhMVrlKMH77//fvOLX/yibRhs3bq12bJlS3uyKgBrk7IfgAzMLAcMUy+s5QOGwuICZfzxr3/967axkCsYff/732++/e1vT+SlbgFYHMp+ADKq5PLly21IzEVqcrAwBw3X8gFDYXGeMv74l7/8ZdvVfMcddzQ7d+5svvWtb62qqxkBsDDKfgByukEZVZLfxf3DP/zD5rvf/e6aGWo6F2FxDv0u5uwkOYq8bdu29opGAKxNyn4AIr+JmFElmTZs2NCej/h7v/d7bWCcFsLiCJ9++ml7kuq7777bLpcuZj+iDLB2KfsBiHLRmhw4zIHChMScl5ifwpg2wmJPuXBBhhzliEFOVM1lg6ehixlgWin7Aciokg8//LCtC9KjmHC4ffv29hSEaTbydxZ/9rOftf+km266qa0scwL/Wk7SOSclRxD+p//93eaXlz5t/o//+b9sNt+1obsXYDpMa9mfajC/h+VcRIArpqk+yOkGqQtSJ2QkSeqCDDnlipFhMXK1n6TrXPEn/7zsLNlRMq1fv77bavUqO0am/CZKGgr//eH/t3n3vU+af/ibP26+e6cjysD0mcay3zBTgFqG5vfrg1IXJDyuhZ+HKHVB6r0MNU0P4lqo5xbb2LA4VHaWzHNkoewwaUisFtnpc5QkU9kx0lAoJ6n+5d++JSwC9ExD2Q/A3BIXcv5epgTHhMVSH6ymsjT1QH4KKfO89wREv5M7t3mHxb6k8DQcssOk4s1OkoZD5pkmaRhPv5GQI8rZITZu3DjySLKwCDDeWi37AViYBMbUBakTEiVSF6RXrtQJk6TUBQmJeY+pC1InrKaDnivpmsJiXy4MkA8gO01J6v3GQ6bl7NJNAyHvKY2DTJ9//vlsIyFDjuYiLALMT38452ov+wG4dil7UxckC6QuSBlc6oFSLyzXsNXEmrx+vz4oATFT3gsLc91hcShPl4ZDptKIyLp8OBnClJ1lOF2LHNUuU14nR7SzY+Q1clJqxlOX+XwJiwDXZjWX/QAsnhysK3VBqQ9SRie0lfJ/WC8sVOqXfn2QqQTEBNTUARlJkvlyHrhcixY9LI5ShgOVI7+Z58PMPB9u2WFS6Y8bxpT1/R2i7HDZIXI7jYPrHTctLAIsntVS9gOwtFIHlDBX6oFSJyRcplwv9cFcSl2Q+JI6IFPK/xIQUydc7TlYmGUJi1fT31mG+m+v7BSZloKwCLB8JqXsB2DlpLwv9UFuJ+yVOqBfF2R9qQsSLFkeExEWJ4WwCAAAcMXa/bVlAAAArpmwCAAAQEVYBAAAoCIsAgAAUBEWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERYBAACoCIsAAABUhEUAAAAqwiIAAAAVYREAAICKsAgAAEBFWAQAAKAiLAIAAFARFgEAAKgIiwAAAFSERQAAACrCIgAAABVhEQAAgIqw2PnR0f/YfPjRZ90SAADAdBMWO//4T79qPvjo024JAABgugmLAz965F81t21Y3y0BAABMJ2Fx4E9/cJewCAAATD1hEQAAgIqwCAAAQEVYBAAAoCIsAgAAUBEWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERYBAACoCIsAAABUhEUAAAAqwiIAAAAVYREAAICKsAgAAEBFWAQAAKAiLAIAAFARFgEAAKgIiwAAAFSERQAAACrCIgAAABVhEQAAgIqwCAAAQGXdF1/qbk+dDz76tPmrv/33s7fjtg3rm1s33NC8ePCP2tsAAADTaKp7FhMGf/gn354NipHbf/qDuwRFAABgqk39MNS/+uEfdLeuSEjc/2ebuyUAAIDpNPVhMeHwL/7897ulpu1p/O6dN3dLAAAA08kFbr5Uehf1KgIAAFwhLH6p9C7qVQQAALhi3U9/+tOpvRpq30e//aL5zSefN3dtvKFbQ9/MzEx3CwAAmAbrLl26dE1h8fLly+1848aN7ZzJs1if0cWLF4VFAACYMoahAgAAUBEWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERYBAACoCIsAAABUhEUAAAAqwiIAAAAVYREAAICKsAgAAEBFWAQAAKAiLAIAAFARFgEAAKgIiwAAAFSERQAAACrCIgAAABVhEQAAgIqwCAAAQEVYBAAAoCIsAgAAUBEWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgMnVh8dixY83u3bu7JQAAAEbRszhw+PDhZtOmTd0SAADAdBIWB5544onm0qVL3RIAAMB0mtqwWHoQMx08eLBbO3qY6v79+2e3zXT+/PnuHgAAgLVpKsPiuXPn2nl6EE+fPt0cOXKkDYmjJChGts106NChZteuXe06AACAtWoqw+KOHTva4aaxffv25sCBA80bb7zRLvelB/HEiRPNCy+80K25Mkw1Tp482c4BAADWIucsfmnbtm3drdESKPsSNt95551uCQAAYO0RFr904cKF7tZow3MUM4x1y5Yt3RIAAMDas+bDYoaL5qI0/WGjCXu5wE0kCOacxYceeqhd7kuP4t69e5snn3yyW3PlwjjpWdyzZ0+3BgAAYO2Zyp7FBMD0JiZE5mI1uWjNuPB3/Pjx5uzZs7NXQj169Ghz6tSp7l4AAIC1ad2lS5e+6G4vyOXLl9v5xo0b2zmTZ7E+o4sXLzYzMzPdEgAAMA2cswgAAEBFWAQAAKAiLAIAAFARFgEAAKgIiwAAAFSERQAAACrCIgAAABVhEQAAgIqwCAAAQEVYBAAAoCIsAgAAUBEWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERYBAACoCIsAAABUhEUAAAAqwiIAAAAVYREAAICKsAgAAEBFWAQAAKAiLAIAAFARFgEAAKgIiwAAAFSERQAAACrCIgAAABVhEQAAgIqwCAAAQGXdT3/60y+621Pvmf/zveZXlz9rDv23dzZ3bbyhW0vMzMx0twAAgGmw7osvdben3l/+7VvNu+990vzD3/xx8907b+7WAgAATB/DUAEAAKgIiwAAAFSERQAAACrCIgAAABVhEQAAgIqwCAAAQEVYBAAAoCIsAgAAUBEWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERY7p/75/e4WAAAAwmLnr//+n5p33/ukWwIAAJhuwuLA7u/d3t0CAACYXsLiwN/92x80373z5m4JAABgOgmLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERYBAACoCIsAAABUhEUAAAAqUx0WP/jo0+Y/v/dxOxXDZQAAgGm07osvdbenTsLiY8//h+bd9z7p1lxx24b1zT/8zR+3cwAAgGk01T2LCYN/9cM/6Ja+sv/PNguKAADAVJv6cxb/9Ad3Nd+585vd0pUA+Rd//vvdEgAAwHSa+rA47F3UqwgAACAstkrvol5FAACAK6b6Ajd9r/27d9sL3fzrfdu6NQAAANNrZFg8c+ZMd2t6fPTbK/+GDTeta+fTaGZmprsFAABMu7FhcfPmzd0S0+DixYvCIgAAMMs5iwAAAFSERQAAACrCIgAAABVhEQAAgIqwCAAAQEVYBAAAoCIsAgAAUBEWAQAAqAiLAAAAVIRFAAAAKsIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERYBAACoCIsAAABUhEUAAAAqwiIAAAAVYREAAICKsAgAAEBFWAQAAKAy0WFx9+7dzaZNm2angwcPdvcsj+Hrnz9/vrsHAABgbZvIsHjy5Mk2nO3du7e5dOnS7BTLFRgTFHfu3Dn72qdPn27uv//+7l4AAIC1bd0XX+puzzpz5kyzefPmbmn5laB2/Pjxbs3ySg/irl27mpdeeql58MEHu7Vr28WLF5uZmZluCQAAmHYT17N47Nix5ty5c81TTz3VrRkvvYz9YaL79+/v7rlyX0LnULbpbzeXN954o7s1Wt7ruNcv+kNZc395DAAAwCSbuLD49ttvt/Or9WwmeB05cuRrw1TPnj07GxAfeuihNnRmSGvfiRMnmocffrhbGm379u3NgQMH2udPsBt1ruLhw4ebRx99tHr9fmAst8v99913X/sYAACASTdxYfHChQvtPIFtnATAhL7XXnutW3PFiy++OBsQ9+zZ0+zYsaN55ZVXunuv9ATGfIaWPv/887PPnyGpw9D4zDPPtMNU+55++un2fWW7TLmddcUTTzzRHDp0qFsCAACYXBN9NdRx3nnnnXY+7H0sy+X+Rx55pA1sRYaVpsdwvhI40yOYi9tEQmMJgpFewjLENFO/1/Ctt95q53fffXc7BwAAWE0mLizee++97Xw4fPRaPPDAA18bipphpRmeulDp5SyB8dVXX23nkZ7FMsS0P83VKwoAALAaTFxYLD1x/eGjQ1u2bGnnuYJnX1ku9ye05ec33nzzzTYwZlhqeguvx9atW2fDYDm/cpRx77EMswUAAJhkExcWE8TSY5dewOFvKuaiMllXzkd87LHHunuu2LdvXxsO+4EwF7M5evRoGz4zLDUyjDTDRsf9ZmNeZ3hl0yeffLKdl/MdM5w15y32e0Bzuzxu1HvM8+bv6itXdC1DWwEAACbBRJ6zmECWYZ/laqRlSq9cLjwTp06dan+LsX9/AtzwtxnzXBmKmufKsNT5yIVoEvT6z50rnWaIaZH3kYvVJKCWbRIM+6+f95jXLvfn/bvADQAAsBpM5I/yr2XpXUyPZD94TgI/yg8AAPStyquhAgAAsLSERQAAACrC4jLL+ZCTNgQVAABgSFgEAACgIiwCAABQERYBAACoCIsAAABUhEUAAAAqwiIAAAAVYfEa7N+/v50AAADWqokJiwcPHmw2bdrUTnPZvXt3u42wBgAAsHQmqmdxx44d7fzw4cPtfOjkyZPNuXPnZrdbiITMBNKVkL/naiEYAABgkkzcMNQDBw40R48e7Za+7rnnnmv27t3b7Ny5s1sDAADAUpi4sPjQQw+1vYfHjh3r1lxx/vz55sSJE81TTz3Vrfm6/jDWTKV3Mr2RWc5zHjlyZPb+YtzjirzuXPcX/W0y5XGRHs1nnnmmvV3uK3/b8LkzAQAATIKJC4t79uxpew9ffvnlbs0VP/7xj9vhp7l/KOcvJgxeunSpnU6fPt0GtAS7bJ91eWx6Lcs2kaC4bdu22XWvvfZa+7gEzGLXrl1fe9zrr7/ehta+hLy8Ztkm299///3tfadOnWoOHTrU3i73P/jgg21QzHOXdZnydzsXEwAAmAQTFxbj4YcfbgNZ6Z2L9Ao+/fTT3dJXEuyy7QsvvNCtaZrt27fPOZy1eP7555snnniiW7oSVOOdd95p56UH8PHHH2/ncfz48TbU9SXo5TWLe++9tw2vc8n2eVzffffd15w9e7ZbAgAAWDkTGRbT8xbpTYwS2sr6vhLs0kvXH86ZcDkf6cnrPy7efvvtdv7GG2+0PZL9IDjKcDjpo48+Ort+LmWIbJnSq3m1kAkAALAcJjIsRnoGS+B79tln2+W59IeBlilDQOeS8wnTk9d/zEIl8CWoZqhpeY6XXnqpu3e8BOB9+/a125bHleGqAAAAK21iw2IZ+pnzCtPb1h8K2rdly5Z2fvHixXY+zqgrqOZ5Rw1tLXI+46ievv5Q0dKz2R/OOrR169bu1lfSe5ley1G9pQAAACttYsNihn7m3MD0LmY+bihozjNM6EovXV8ubjO8cunwwjR5XP9COsOLyzzwwAPtvP/7jOViOkUJq/0rnJZhqEP9C+ckQOZ5yrrMy1VTAQAAVtrEhsXIhW7683Ey3DSBsn/+34ULF77W25cL4CSc5b4MP42f/OQnbYAsj8kFZhIgiwTUXCG1/5MbeS/9IbEJqxk+moCY+zMkdTgMNb2HeX8JtNkmwTDr8jxlXRmSCgAAMAnWffGl7vasM2fONJs3b+6WmAYZxjszM9MtAQAA026iexYBAABYGcIiAAAAFWERAACAirAIAABARVgEAACgIiwCAABQERYBAACoCIsAAABU1kxY3LRpU3Py5MluqWn279/f7N69u1taHnn9vA8AAIDVbk2ExcOHDzc7duxo9uzZ061ZGXn9vI+8HwAAgNVsTYTFo0ePNo888ki3dMXx48ebU6dOdUvLJ+8j7wcAAGA1W/VhMUM/z50719xzzz3dmpWV95H30x8SCwAAsNqs+rD45ptvtvPhENSDBw9W5yzmfMIyDYeKHjt27Gv3Z+or958/f/5r22R9X3kf5X0BAACsRqs+LF64cKE9T/BqEhwPHDjQXLp0qZ1ef/31NvhFguOjjz7avPbaa7P3Z9sSDvt27do1u82hQ4faxw23yfvJ+wIAAFitVn1YzJDPnTt3dkvzl3Mat2/f3t5+5pln2uDX7518/vnn2/mrr77azovTp093t5rmgQceaOdvvfVWOy/yfvK+AAAAVqtVHxbPnj3b3Zrb008/3Rw5cqQaOlp6Bbdu3drO+67WQ1jC5ttvv93O++b7vgAAACbRqg+L8+1VfPDBB782dHTUENPFdC29nQAAAJNi1YfF9P4tpBfviSeeaENjZIjpXL2DGUq6bdu2bmn+8n7mcx4lAADApFr1YTFh7mrnB6YHsX9l1PKzFmXoaS5mk/MW+z2N+/fvb+cJlwt1rSETAABgUqz6sFh+X3Gu3zVM72F+LL/83MW+ffva4agZmhq5mE2Wc6XTsk16B0sP5EKU9zEpv/sIAABwLdZ98aXu9qwzZ840mzdv7pYmX3oNEwb7vYD5ncUTJ040p06d6tYsj/wMx9GjR5f9da/XxYsXm5mZmW4JAACYdqu+ZzESFBPQ+hIU9+7d2y0tn7yPvB8AAIDVbE30LEaGjuZH9d988832/MPIbyKWC9gshwxBzRDXaxm+utL0LAIAAH1rJixyfYRFAACgb00MQwUAAGBxCYsAAABUhEUAAAAqwiIAAAAVYREAAICKsAgAAEBFWAQAAKAiLAIAAFARFgEAAKgIiwAAAFSERQAAACrCIgAAABVhEQAAgIqwCAAAQEVYBAAAoCIsAgAAUBEWAQAAqAiLAAAAVNZ98aXu9qwzZ840mzdv7pZYbT777LN2+vzzz5t8vP0phuuy3SeffNLMzMy09wMAAAiLq1AJg/0pge/TTz9tpxtuuKFZv359u/6WW25p1q1b13zjG9+YnZfb/fW/+93vmjvuuKN7BQAAYNoJixMuITBBrj8l3CUM3njjjc1NN900eztTgmICIAAAwPUQFifMb3/723ZKD2GCYcLizTff3E7f/OY323lCIQAAwFISFldYOV8wwfDjjz9uewYzdDShcMOGDYIhAACwIoTFFZBgmIBYehETCjdu3NiGROEQAACYBMLiMsnFZtJz+NFHH7XLt9566+wEAAAwaYTFJZR/bQJipvQgpvew9CACAABMMmFxCSQYpgcxU4Lhpk2b2pAIAACwWgiLi6gExFy05vbbb2+nXLAGAABgtfGDfNcp5yJ+8MEHzc9//vP2wjV33XVXs2PHjubOO+8UFAEAgFVLWLxGGWr6/vvvN7/4xS/aULh169Zmy5YtzW233dZtAQAAsHoJiwuUi9X8+te/boNirmT6/e9/v/n2t7/d/mA+AADAWiEsztNvfvOb5pe//GV7TuIdd9zR7Ny5s/nWt77VrFu3rtsCAABg7RAW55Br//TPR0wP4rZt29qrmwIAAKxlwuIIn376aXPp0qXm3XffbZdzLmLOSfQD+gAAwLQQFnvKRWsy3PTGG29sr2qanxC5+eabuy0AAACmw8jfWfzZz37WfPLJJ81NN93UBqVcvOUb31i7uTLnI+ZcxPwr8tuIzkUEAACm3ciwGPn9wA8//LC5fPlyG6YSHBMaM61fv77bavXKOYjlR/RvueWWNiQaZgoAAHDF2LA4VIJj5ullLMExIXK1yLmI6THNlDCcC9UkJGbIKQAAAF+Zd1jsS29cQmOuFJrQlbCV0Jh5pkkawtkPiOlNzI/mb9y4US8iAADAHK4pLPblojAJYhmqmh+sz+1+cMy0nMNWEw7znhIMM33++eezATHDTQEAALi66w6LQ3m6hMZM6YHMPOsSGjN89YYbbqima5EezTLlddKbmZCY19iwYUN7YZ4yBwAAYGEWPSyOUoaCll6/zNPrl3nCXgmRCXzjhrBmfT8gprcyj0kvZm4nGOYcyoRSAAAArs+yhMWrKeExQ0aH+m8vobBMAAAALJ2JCIsAAABMlrX7S/sAAABco6b5/wEeINyrhHvnQwAAAABJRU5ErkJggg=="
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image-3.png](attachment:image-3.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Corona Virus Sequence Files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All corona virus sequences are available as FASTA files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Examples of corona virus sequence files (fasta) with different number of sequences used in the project:\n",
    "```\n",
    "data\n",
    "  |--cov_data\n",
    "  |    |--cov_virus_sequences_hundred.fa\n",
    "  |    |--cov_virus_sequences_twenty_five.fa\n",
    "  |    |--cov_virus_sequences_ten.fa\n",
    "  |    |--cov_virus_sequences.fa\n",
    "```\n",
    "\n",
    "Those fasta files have the following format (showing 5 sequences):\n",
    "```\n",
    ">2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-BetaCoV/YN2018D\tscient\n",
    "TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTAGCTGTCGCTCGGCTGCATGCCTA ...\n",
    ">11128:ncbi:2 [LC494191]\n",
    "CATCCCGCTTCACTGATCTCTTGTTAGATCTTTTCATAATCTAAACTTTATAAAAACATCCACTCCCTGTAGTCTATGCCTATGGGCGTAGATTTTTCATAGTGGTGTCT ...\n",
    ">31631:ncbi:3 [KY967361]\t31631\tncbi\t3 [KY967361] 31631\tHuman coronavirus OC43\tscientific name\n",
    "ATCTCTTGTTAGATCTTTTTGTAATCTAAACTTTATAAAAACATCCACTCCCTGTAATCTATGCTTGTGGGCGTAGATTTTTCATAGTGGTGTTTATATTCATTTCTGCT ...\n",
    ">277944:ncbi:4 [LC654455]\t277944\tncbi\t4 [LC654455] 277944\tHuman coronavirus NL63\tscientific name\n",
    "ATTTTCTTATTTAGACTTTGTGTCTACTCTTCTCAACTAAACGAAATTTTTCTAGTGCTGTCATTTGTTATGGCAGTCCTAGTGTAATTGAAATTTCGTCAAGTTTGTAA ...\n",
    ">11120:ncbi:5 [MN987231]\t11120\tncbi\t5 [MN987231] 11120\tInfectious bronchitis virus\tscientific name\n",
    "TCCTAAGTGTGATATAAATATATATCATACACACTAGCCTTGCGCTAGATTTCTAACTTAACAAAACGGACTTAAATACCTACAGCTGGTCCCTATAGGTGTTCCATTGC ...\n",
    "....\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metadata can be parsed from the definition line for further use.\n",
    "\n",
    "Example 1:\n",
    "\n",
    "- Definition Line:\n",
    "```ascii\n",
    "    >2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-BetaCoV/YN2018D\t\tscientific name\n",
    "```\n",
    "- Metadata:\n",
    "    - `seqid` = `2591237:ncbi:1`\n",
    "    - `taxonomyid` = `2591237`\n",
    "    - `source` = `ncbi`\n",
    "    - `seqnb` = `1`\n",
    "    - `accession` = `MK211378`\n",
    "    - `species` = `Coronavirus BtRs-BetaCoV/YN2018D`\n",
    "\n",
    "Example 2:\n",
    "\n",
    "- Definition Line\n",
    "```ascii\n",
    "    >11128:ncbi:2 [LC494191]\n",
    "```\n",
    "\n",
    "- Metadata:\n",
    "    - `seqid` = `11128:ncbi:2`\n",
    "    - `taxonomyid` = `11128`\n",
    "    - `source` = `ncbi`\n",
    "    - `seqnb` = `2`\n",
    "    - `accession` = `LC494191`\n",
    "    - `species` = `''`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulated Read Files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Simulated reads information is split between two files:\n",
    "\n",
    "- FASTQ (`.fq`) files providing the read sequences and their ASCII quality scores\n",
    "- ALN (`.aln`) files with alignment information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Example of read sequence FASTA and ALN files used in this project\n",
    "\n",
    "```ascii\n",
    "data\n",
    "  |--cov_simreads\n",
    "  |    |--single_25seq_50bp\n",
    "  |    |    |--single_25seq_50bp.aln\n",
    "  |    |    |--single_25seq_50bp.fq\n",
    "  |    |--single_25seq_150bp\n",
    "  |    |    |--single_25seq_150bp.fq\n",
    "  |    |    |--single_25seq_150bp.aln\n",
    "  |    |--single_100seq_50bp\n",
    "  |    |    |--single_100seq_50bp.fq\n",
    "  |    |    |--single_100seq_50bp.aln\n",
    "  |    |--single_100seq_150bp\n",
    "  |    |    |--single_100seq_150bp.fq\n",
    "  |    |    |--single_100seq_150bp.aln\n",
    "  ```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FASTQ** (`.fq`) files generated by ART Illumina have the following structure (showing 5 reads):\n",
    "\n",
    "```ascii\n",
    "@2591237:ncbi:1-60400\n",
    "ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG\n",
    "+\n",
    "CCCBCGFGBGGGGGGGBGGGGGGGGG>GGG1G=/GGGGGGGGGGGGGGGG\n",
    "@2591237:ncbi:1-60399\n",
    "GATCAATGTGGCATCTACAATACAGACAGCATGAAGCACCACCAAAGGAC\n",
    "+\n",
    "BCBCCFGGGGGGGG1CGGGG<GGBGGGGGFGCGGGGGGDGGG/GG1GGGG\n",
    "@2591237:ncbi:1-60398\n",
    "ATCTACCAGTGGTAGATGGGTTCTTAATAATGAACATTATAGAGCTCTAC\n",
    "+\n",
    "CCCCCGGGEGG1GGF1G/GGEGGGGGGGGGGGGFFGGGGGGGGGGDGGDG\n",
    "@2591237:ncbi:1-60397\n",
    "CGTAAAGTAGAGGCTGTATGGTAGCTAGCACAAATGCCAGCACCAATAGG\n",
    "+\n",
    "BCCCCGGGFGGGGGGFGGGGFGG1GGGGGGG>GG1GGGGGGGGGGE<GGG\n",
    "@2591237:ncbi:1-60396\n",
    "GGTATCGGGTATCTCCTGCATCAATGCAAGGTCTTACAAAGATAAATACT\n",
    "+\n",
    "CBCCCGGG@CGGGGGGGGGGGG=GFGGGGDGGGFG1GGGGGGGG@GGGGG\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following information can be parsed from the each read sequence in the FASTQ file:\n",
    "\n",
    "- Line 1: `readid`, a unique ID for the read, under for format `@readid` \n",
    "- Line 2: `readseq`, the sequence of the read\n",
    "- Line 3: a separator `+`\n",
    "- Line 4: `read_qscores`, the base quality scores encoded in ASCII \n",
    "\n",
    "Example:\n",
    "```\n",
    "@2591237:ncbi:1-60400\n",
    "ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG\n",
    "+\n",
    "CCCBCGFGBGGGGGGGBGGGGGGGGG>GGG1G=/GGGGGGGGGGGGGGGG\n",
    "```\n",
    "- `readid` = `2591237:ncbi:1-60400`\n",
    "- `readseq` = `ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG`, a 50 bp read\n",
    "- `read_qscores` = `CCCBCGFGBGGGGGGGBGGGGGGGGG>GGG1G=/GGGGGGGGGGGGGGGG`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**ALN** (`.aln`) files generated by ART Illumina have the following structure (showing 5 reads).\n",
    "\n",
    "ALN files consist of :\n",
    "\n",
    "- a header with the ART-Ilumina command used for the simulation (`@CM`) and info on each of the reference sequences used for the simulations (`@SQ`). Header always starts with `##ART_Illumina` and ends with `##Header End` :\n",
    "- the body with three lines for each read:\n",
    "    - definition line with `readid`, \n",
    "        - reference sequence identification number `refseqid`, \n",
    "        - the position in the read in the reference sequence `aln_start_pos` \n",
    "        - the strand the read was taken from `ref_seq_strand`. `+` for coding strand and `-` for template strand\n",
    "    - aligned reference sequence, that is the sequence segment in the original reference corresponding to the read\n",
    "    - aligned read sequence, that is the simmulated read sequence, where each bp corresponds to the reference sequence bp in the same position."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```ascii\n",
    "##ART_Illumina\tread_length\t50\n",
    "@CM\t/bin/art_illumina -i /home/vtec/projects/bio/metagentools/data/cov_data/cov_virus_sequences_ten.fa -ss HS25 -l 50 -f 100 -o /home/vtec/projects/bio/metagentools/data/cov_simreads/single_10seq_50bp/single_10seq_50bp -rs 1674660835\n",
    "@SQ\t2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-BetaCoV/YN2018D\t30213\n",
    "@SQ\t11128:ncbi:2 [LC494191]\t11128\tncbi\t2 [LC494191] 11128\tBovine coronavirus\t30942\n",
    "@SQ\t31631:ncbi:3 [KY967361]\t31631\tncbi\t3 [KY967361] 31631\tHuman coronavirus OC43\t\t30661\n",
    "@SQ\t277944:ncbi:4 [LC654455]\t277944\tncbi\t4 [LC654455] 277944\tHuman coronavirus NL63\t27516\n",
    "@SQ\t11120:ncbi:5 [MN987231]\t11120\tncbi\t5 [MN987231] 11120\tInfectious bronchitis virus\t27617\n",
    "@SQ\t28295:ncbi:6 [KU893866]\t28295\tncbi\t6 [KU893866] 28295\tPorcine epidemic diarrhea virus\t28043\n",
    "@SQ\t28295:ncbi:7 [KJ645638]\t28295\tncbi\t7 [KJ645638] 28295\tPorcine epidemic diarrhea virus\t27998\n",
    "@SQ\t28295:ncbi:8 [KJ645678]\t28295\tncbi\t8 [KJ645678] 28295\tPorcine epidemic diarrhea virus\t27998\n",
    "@SQ\t28295:ncbi:9 [KR873434]\t28295\tncbi\t9 [KR873434] 28295\tPorcine epidemic diarrhea virus\t28038\n",
    "@SQ\t1699095:ncbi:10 [KT368904]\t1699095\tncbi\t10 [KT368904] 1699095\tCamel alphacoronavirus\t27395\n",
    "##Header End\n",
    ">2591237:ncbi:1\t2591237:ncbi:1-60400\t14770\t+\n",
    "ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG\n",
    "ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG\n",
    ">2591237:ncbi:1\t2591237:ncbi:1-60399\t17012\t-\n",
    "GATCAATGTGGCATCTACAATACAGACAGCATGAAGCACCACCAAAGGAC\n",
    "GATCAATGTGGCATCTACAATACAGACAGCATGAAGCACCACCAAAGGAC\n",
    ">2591237:ncbi:1\t2591237:ncbi:1-60398\t9188\t+\n",
    "ATCTACCAGTGGTAGATGGGTTCTTAATAATGAACATTATAGAGCTCTAC\n",
    "ATCTACCAGTGGTAGATGGGTTCTTAATAATGAACATTATAGAGCTCTAC\n",
    ".....\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing sequence files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following classes make it easier to read and parse files of different format into their underlying components to generated the training, validation, testing and inference datasets for the model.\n",
    "\n",
    "Each class inherits from `TextFileBaseReader` and adds:\n",
    "\n",
    "- One or several text parsing method(s) to parse metadata according to a specific format\n",
    "- A file parsing method to extract metadata from all elements in the file, returning it as a key:value dictionary and optionally save the metadata as a json file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTA file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extension of `TextFileBaseReader` class for fasta sequence files.\n",
    "\n",
    "Structure of a FASTA sequence file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-Be\n",
      "TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT\n",
      ">11128:ncbi:2 [LC494191]\n",
      "CATCCCGCTTCACTGATCTCTTGTTAGATCTTTTCATAATCTAAACTTTATAAAAACATCCACTCCCTGTAGTCTATGCC\n"
     ]
    }
   ],
   "source": [
    "#| echo: false\n",
    "p2fasta = Path('data_dev/cov_virus_sequences_two.fa').resolve()\n",
    "\n",
    "it = TextFileBaseReader(p2fasta, nlines=1)\n",
    "for i, t in enumerate(it):\n",
    "    txt = t.replace('\\n', '')[:80]\n",
    "    print(f\"{txt}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "class FastaFileReader(TextFileBaseReader):\n",
    "    \"\"\"Wrap a FASTA file and retrieve its content in raw format and parsed format\"\"\"\n",
    "    def __init__(\n",
    "        self,\n",
    "        path: str|Path,  # path to the Fasta file\n",
    "    ):\n",
    "        super().__init__(path, nlines=1)\n",
    "        self.text_to_parse_key = 'definition line'\n",
    "        self.set_parsing_rules(verbose=False)\n",
    "        \n",
    "    def __next__(self)-> dict[str, str]:   # `{'definition line': text in dfn line, 'sequence': full sequence as str}` \n",
    "        \"\"\"Return one definition line and the corresponding sequence\"\"\"\n",
    "        lines = []\n",
    "        for i in range(2):\n",
    "            lines.append(self._safe_readline())\n",
    "        dfn_line = lines[0]\n",
    "        sequence = lines[1].replace('\\n', '')\n",
    "        return {'definition line':dfn_line, 'sequence':f\"{sequence}\"}\n",
    "    \n",
    "    def print_first_chunks(\n",
    "        self, \n",
    "        nchunks:int=3,  # number of chunks to print out\n",
    "    ):\n",
    "        \"\"\"Print the first `nchunks` chunks of text from the file\"\"\"\n",
    "        self.reset_iterator()\n",
    "        for i, seq_dict in enumerate(self.__iter__()):\n",
    "            print(f\"\\nSequence {i+1}:\")\n",
    "            print(seq_dict['definition line'])\n",
    "            print(f\"{seq_dict['sequence'][:80]} ...\")\n",
    "            if i >= nchunks: break\n",
    "        self.reset_iterator()\n",
    "            \n",
    "    def parse_file(\n",
    "        self,\n",
    "        add_seq :bool=False,     # When True, add the full sequence to the parsed metadata dictionary\n",
    "        save_json: bool=False    # When True, save the file metadata as a json file of same stem name\n",
    "    )-> dict[str]:               # Metadata as Key/Values pairs\n",
    "        \"\"\"Read fasta file and return a dictionary with definition line metadata and optionally sequences\"\"\"\n",
    "    \n",
    "        self.reset_iterator()\n",
    "        parsed = {}\n",
    "        for d in self:\n",
    "            dfn_line = d['definition line']\n",
    "            seq = d['sequence']\n",
    "            metadata = self._parse_text_fn(dfn_line, self.re_pattern, self.re_keys)\n",
    "            if add_seq: metadata['sequence'] = seq         \n",
    "            parsed[metadata['seqid']] = metadata\n",
    "                        \n",
    "        if save_json:\n",
    "            p2json = self.path.parent / f\"{self.path.stem}_metadata.json\"\n",
    "            with open(p2json, 'w') as fp:\n",
    "                json.dump(parsed, fp, indent=4)\n",
    "                print(f\"Metadata for '{self.path.name}'> saved as <{p2json.name}> in  \\n{p2json.parent.absolute()}\\n\")\n",
    "\n",
    "        return parsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L21){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastaFileReader\n",
       "\n",
       ">      FastaFileReader (path:str|pathlib.Path)\n",
       "\n",
       "Wrap a FASTA file and retrieve its content in raw format and parsed format\n",
       "\n",
       "|    | **Type** | **Details** |\n",
       "| -- | -------- | ----------- |\n",
       "| path | str \\| Path | path to the Fasta file |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L21){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastaFileReader\n",
       "\n",
       ">      FastaFileReader (path:str|pathlib.Path)\n",
       "\n",
       "Wrap a FASTA file and retrieve its content in raw format and parsed format\n",
       "\n",
       "|    | **Type** | **Details** |\n",
       "| -- | -------- | ----------- |\n",
       "| path | str \\| Path | path to the Fasta file |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(FastaFileReader)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As an iterator, `FastaFileReader` returns a `dict` at each step, as follows:\n",
    "```python\n",
    "{\n",
    "    'definition line': 'string in file as the definition line for the sequence',\n",
    "    'sequence': 'the full sequence'\n",
    "}\n",
    "```\n",
    "\n",
    "Illustration:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-Be ...\n",
      "TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'>2591237:ncbi:1 [MK211378]\\t2591237\\tncbi\\t1 [MK211378] 2591237\\tCoronavirus BtRs-BetaCoV/YN2018D\\t\\tscientific name\\n'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "p2fasta = Path('data_dev/cov_virus_sequences_two.fa')\n",
    "it = FastaFileReader(p2fasta)\n",
    "iteration_output = next(it)\n",
    "\n",
    "print(iteration_output['definition line'][:80], '...')\n",
    "print(iteration_output['sequence'][:80], '...')\n",
    "\n",
    "display(iteration_output['definition line'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output type :     <class 'dict'>\n",
      "keys :            dict_keys(['definition line', 'sequence'])\n",
      "definition line : >2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-Be ...'\n",
      "sequence :       'TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTAGCTGTCGCTCGGC ...'\n"
     ]
    }
   ],
   "source": [
    "print(f\"output type :     {type(iteration_output)}\")\n",
    "print(f\"keys :            {iteration_output.keys()}\")\n",
    "print(f\"definition line : {iteration_output['definition line'][:80]} ...'\")\n",
    "print(f\"sequence :       '{iteration_output['sequence'][:100]} ...'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L40){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastaFileReader.print_first_chunks\n",
       "\n",
       ">      FastaFileReader.print_first_chunks (nchunks:int=3)\n",
       "\n",
       "Print the first `nchunks` chunks of text from the file\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| nchunks | int | 3 | number of chunks to print out |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L40){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastaFileReader.print_first_chunks\n",
       "\n",
       ">      FastaFileReader.print_first_chunks (nchunks:int=3)\n",
       "\n",
       "Print the first `nchunks` chunks of text from the file\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| nchunks | int | 3 | number of chunks to print out |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(FastaFileReader.print_first_chunks)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is convenient to quickly discover and explore new fasta files in raw text format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Sequence 1:\n",
      ">2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-BetaCoV/YN2018D\t\tscientific name\n",
      "\n",
      "TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...\n",
      "\n",
      "Sequence 2:\n",
      ">11128:ncbi:2 [LC494191]\n",
      "\n",
      "CATCCCGCTTCACTGATCTCTTGTTAGATCTTTTCATAATCTAAACTTTATAAAAACATCCACTCCCTGTAGTCTATGCC ...\n"
     ]
    }
   ],
   "source": [
    "it = FastaFileReader(p2fasta)\n",
    "it.print_first_chunks(nchunks=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parsing metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The class also provides methods to parse metadata from the file content."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For instance, we can use regex pattern for parsing metadata fom the definition lines in the reference sequence fasta file (rule `fasta_cov_ncbi`):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sequence 1:\n",
    "\n",
    "- Definition Line:\n",
    "```ascii\n",
    ">2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus YN2018D\t\tscientific name\n",
    "```\n",
    "- Metadata:\n",
    "    - `seqid` = `2591237:ncbi:1`\n",
    "    - `taxonomyid` = `2591237`\n",
    "    - `source` = `ncbi`\n",
    "    - `seqnb` = `1`\n",
    "    - `accession` = `MK211378`\n",
    "    - `species` = `Coronavirus BtRs-BetaCoV/YN2018D`\n",
    "\n",
    "Sequence 2:\n",
    "\n",
    "- Definition Line\n",
    "```ascii\n",
    "    >11128:ncbi:2 [LC494191]\n",
    "```\n",
    "\n",
    "- Metadata:\n",
    "    - `seqid` = `11128:ncbi:2`\n",
    "    - `taxonomyid` = `11128`\n",
    "    - `source` = `ncbi`\n",
    "    - `seqnb` = `2`\n",
    "    - `accession` = `LC494191`\n",
    "    - `species` = `''`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To do this, we have the method `parse_text` as well as the option of setting a default \"parsing rule\" for one instance with `set_parsing_rules`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/core.py#LNone){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### TextFileBaseReader.parse_text\n",
       "\n",
       ">      TextFileBaseReader.parse_text (txt:str, pattern:str=None,\n",
       ">                                     keys:list[str]=None)\n",
       "\n",
       "Parse text using regex pattern and key. Return a metadata dictionary\n",
       "\n",
       "The passed text is parsed using the regex pattern. The method return a dictionary in the format:\n",
       "    {\n",
       "        'key_1': 'metadata 1',\n",
       "        'key_2': 'metadata 2',\n",
       "        ...\n",
       "    }\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| txt | str |  | text to parse |\n",
       "| pattern | str | None | If None, uses standard regex pattern to extract metadata, otherwise, uses passed regex |\n",
       "| keys | list | None | If None, uses standard regex list of keys, otherwise, uses passed list of keys (str) |\n",
       "| **Returns** | **dict** |  | **parsed metadata in key/value format** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/core.py#LNone){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### TextFileBaseReader.parse_text\n",
       "\n",
       ">      TextFileBaseReader.parse_text (txt:str, pattern:str=None,\n",
       ">                                     keys:list[str]=None)\n",
       "\n",
       "Parse text using regex pattern and key. Return a metadata dictionary\n",
       "\n",
       "The passed text is parsed using the regex pattern. The method return a dictionary in the format:\n",
       "    {\n",
       "        'key_1': 'metadata 1',\n",
       "        'key_2': 'metadata 2',\n",
       "        ...\n",
       "    }\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| txt | str |  | text to parse |\n",
       "| pattern | str | None | If None, uses standard regex pattern to extract metadata, otherwise, uses passed regex |\n",
       "| keys | list | None | If None, uses standard regex list of keys, otherwise, uses passed list of keys (str) |\n",
       "| **Returns** | **dict** |  | **parsed metadata in key/value format** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(FastaFileReader.parse_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running the parser function with specifically defined `pattern` and `keys`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-BetaCoV/YN2018D\t\tscientific name\n"
     ]
    }
   ],
   "source": [
    "it = FastaFileReader(p2fasta)\n",
    "dfn_line, sequence = next(it).values()\n",
    "print(dfn_line.replace('\\n', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern = r\"^>(?P<seqid>(?P<taxonomyid>\\d+):(?P<source>ncbi):(?P<seqnb>\\d*))[\\s\\t]*\\[(?P<accession>[\\w\\d]*)\\]([\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P=source)[\\s\\t]*(?P=seqnb)[\\s\\t]*\\[(?P=accession)\\][\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P<species>[\\w\\s\\-\\_\\/]*))?\"\n",
    "\n",
    "keys = 'seqid taxonomyid accession source seqnb species'.split(' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'accession': 'MK211378',\n",
       " 'seqid': '2591237:ncbi:1',\n",
       " 'seqnb': '1',\n",
       " 'source': 'ncbi',\n",
       " 'species': 'Coronavirus BtRs-BetaCoV/YN2018D  scientific name',\n",
       " 'taxonomyid': '2591237'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "it.parse_text(dfn_line, pattern=pattern, keys=keys)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When a `FastaFileReader` instance is created, all existing rules in `parsing_rules.json` file are tested on the first definition line of the file and the one rule that parses the most matches will be selected automatically and saved in instance attributes `re_rule_name`, `re_pattern` and `re_keys`. Finally, `parse_file` extract metadata from each definition line in the fasta file and return a dictionary with all metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fasta_cov_ncbi\n",
      "^>(?P<seqid>(?P<taxonomyid>\\d+):(?P<source>ncbi):(?P<seqnb>\\d*))[\\s\\t]*\\[(?P<accession>[\\w\\d]*)\\]([\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P=source)[\\s\\t]*(?P=seqnb)[\\s\\t]*\\[(?P=accession)\\][\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P<species>[\\w\\s\\-\\_\\/]*))?\n",
      "['seqid', 'taxonomyid', 'source', 'accession', 'seqnb', 'species']\n"
     ]
    }
   ],
   "source": [
    "print(it.re_rule_name)\n",
    "print(it.re_pattern)\n",
    "print(it.re_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'accession': 'MK211378',\n",
       " 'seqid': '2591237:ncbi:1',\n",
       " 'seqnb': '1',\n",
       " 'source': 'ncbi',\n",
       " 'species': 'Coronavirus BtRs-BetaCoV/YN2018D  scientific name',\n",
       " 'taxonomyid': '2591237'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "it.parse_text(dfn_line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When another fasta file, which has another definition line structure, is used, another parsing rule is selected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">1 dna_rm:primary_assembly primary_assembly:mRhiFer1_v1.p:1:1:124933378:1 REF\n"
     ]
    }
   ],
   "source": [
    "p2ncov = Path('../data/ncov_data/rhinolophus_ferrumequinum/dna/Rhinolophus_ferrumequinum.mRhiFer1_v1.p.dna_rm.primary_assembly.1.clean.fa')\n",
    "assert p2ncov.is_file()\n",
    "\n",
    "it2 = FastaFileReader(path=p2ncov)\n",
    "\n",
    "dfn_line, sequence = next(it2).values()\n",
    "print(dfn_line.replace('\\n', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fasta_rhinolophus_ferrumequinum\n",
      "^>\\d[\\s\\t](?P<seq_type>dna_rm):(?P<id_type>[\\w\\_]*)[\\s\\w](?P=id_type):(?P<assy>[\\w\\d\\_]*)\\.(?P<seq_level>[\\w]*):\\d*:\\d*:(?P<taxonomy>\\d*):(?P<id>\\d*)[\\s\t]REF$\n",
      "['seq_type', 'id_type', 'assy', 'seq_level', 'taxonomy', 'id']\n"
     ]
    }
   ],
   "source": [
    "print(it2.re_rule_name)\n",
    "print(it2.re_pattern)\n",
    "print(it2.re_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'assy': 'mRhiFer1_v1',\n",
      " 'id': '1',\n",
      " 'id_type': 'primary_assembly',\n",
      " 'seq_level': 'p',\n",
      " 'seq_type': 'dna_rm',\n",
      " 'taxonomy': '124933378'}\n"
     ]
    }
   ],
   "source": [
    "pprint(it2.parse_text(dfn_line))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This rule selection is performed by the class method `set_parsing_rule`. The method can also be called with specific `pattern` and `keys` to force parsing rule not yet saved in the json file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/core.py#LNone){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### TextFileBaseReader.set_parsing_rules\n",
       "\n",
       ">      TextFileBaseReader.set_parsing_rules (pattern:str|bool=None,\n",
       ">                                            keys:list[str]=None,\n",
       ">                                            verbose:bool=False)\n",
       "\n",
       "Set the standard regex parsing rule for the file.\n",
       "\n",
       "Rules can be set:\n",
       "\n",
       "1. manually by passing specific custom values for `pattern` and `keys`\n",
       "2. automatically, by testing all parsing rules saved in `parsing_rule.json` \n",
       "\n",
       "Automatic selection of parsing rules works by testing each rule saved in `parsing_rule.json` on the first \n",
       "definition line of the fasta file, and selecting the one rule that generates the most metadata matches.\n",
       "\n",
       "Rules consists of two parameters:\n",
       "\n",
       "- The regex pattern including one `group` for each metadata item, e.g `(?P<group_name>regex_code)`\n",
       "- The list of keys, i.e. the list with the name of each regex groups, used as key in the metadata dictionary\n",
       "\n",
       "This method updates the three following class attributes: `re_rule_name`, `re_pattern`, `re_keys`\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| pattern | str \\| bool | None | regex pattern to apply to parse the text, search in parsing rules json if None |\n",
       "| keys | list | None | list of keys/group for regex, search in parsing rules json if None |\n",
       "| verbose | bool | False | when True, provides information on each rule |\n",
       "| **Returns** | **None** |  |  |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/core.py#LNone){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### TextFileBaseReader.set_parsing_rules\n",
       "\n",
       ">      TextFileBaseReader.set_parsing_rules (pattern:str|bool=None,\n",
       ">                                            keys:list[str]=None,\n",
       ">                                            verbose:bool=False)\n",
       "\n",
       "Set the standard regex parsing rule for the file.\n",
       "\n",
       "Rules can be set:\n",
       "\n",
       "1. manually by passing specific custom values for `pattern` and `keys`\n",
       "2. automatically, by testing all parsing rules saved in `parsing_rule.json` \n",
       "\n",
       "Automatic selection of parsing rules works by testing each rule saved in `parsing_rule.json` on the first \n",
       "definition line of the fasta file, and selecting the one rule that generates the most metadata matches.\n",
       "\n",
       "Rules consists of two parameters:\n",
       "\n",
       "- The regex pattern including one `group` for each metadata item, e.g `(?P<group_name>regex_code)`\n",
       "- The list of keys, i.e. the list with the name of each regex groups, used as key in the metadata dictionary\n",
       "\n",
       "This method updates the three following class attributes: `re_rule_name`, `re_pattern`, `re_keys`\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| pattern | str \\| bool | None | regex pattern to apply to parse the text, search in parsing rules json if None |\n",
       "| keys | list | None | list of keys/group for regex, search in parsing rules json if None |\n",
       "| verbose | bool | False | when True, provides information on each rule |\n",
       "| **Returns** | **None** |  |  |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(FastaFileReader.set_parsing_rules)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "definition line: '>2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-BetaCoV/YN2018D\t\tscientific name'\n"
     ]
    }
   ],
   "source": [
    "it = FastaFileReader(p2fasta)\n",
    "dfn_line, sequence = next(it).values()\n",
    "print(f\"definition line: '{dfn_line[:-1]}'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Automatic parsing works by testing each saved rule for the value of `definition line` in the first sequence in the fasta file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "key for text to parse: definition line\n",
      "\n",
      "Text to parse for testing (extracted from first iteration):\n",
      ">2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-BetaCoV/YN2018D\t\tscientific name\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <fasta_cov_ncbi> generated 6 matches\n",
      "--------------------------------------------------------------------------------\n",
      "^>(?P<seqid>(?P<taxonomyid>\\d+):(?P<source>ncbi):(?P<seqnb>\\d*))[\\s\\t]*\\[(?P<accession>[\\w\\d]*)\\]([\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P=source)[\\s\\t]*(?P=seqnb)[\\s\\t]*\\[(?P=accession)\\][\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P<species>[\\w\\s\\-\\_\\/]*))?\n",
      "['seqid', 'taxonomyid', 'source', 'accession', 'seqnb', 'species']\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <fasta_rhinolophus_ferrumequinum> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <fastq_art_illumina> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <aln_art_illumina> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <aln_art_illumina-refseq> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Selected rule with most matches: fasta_cov_ncbi\n"
     ]
    }
   ],
   "source": [
    "print(f\"key for text to parse: {it.text_to_parse_key}\\n\")\n",
    "it.reset_iterator()\n",
    "print('Text to parse for testing (extracted from first iteration):')\n",
    "print(next(it)[it.text_to_parse_key])\n",
    "print()\n",
    "it.set_parsing_rules(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If no saved rule generates a match, `re_rule_name`, `re_pattern` and `re_keys` remain `None` and a warning message is issued to ask user to add a parsing rule manually. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vtec/projects/bio/metagentools/metagentools/core.py:208: UserWarning: \n",
      "        None of the saved parsing rules were able to extract metadata from the first line in this file.\n",
      "        You must set a custom rule (pattern + keys) before parsing text, by using:\n",
      "            `self.set_parsing_rules(custom_pattern, custom_list_of_keys)`\n",
      "                \n",
      "  warnings.warn(msg, category=UserWarning)\n"
     ]
    }
   ],
   "source": [
    "it2 = FastaFileReader('data_dev/sequences_two_no_matching_rule.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "it2.re_rule_name is None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But we still can set a standard rule manually, by passing a re pattern and the corresponding list of keys."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Custom Rule\n",
      "^>(?P<seqid>(?P<taxonomyid>\\d+):(?P<source>ncbi):(?P<seqnb>\\d*))\\s*(?P<text>[\\w\\s]*)$\n",
      "['seqid', 'taxonomyid', 'source', 'seqnb', 'text']\n"
     ]
    }
   ],
   "source": [
    "pat = r\"^>(?P<seqid>(?P<taxonomyid>\\d+):(?P<source>ncbi):(?P<seqnb>\\d*))\\s*(?P<text>[\\w\\s]*)$\"\n",
    "keys = \"seqid taxonomyid source seqnb text\".split()\n",
    "it2.set_parsing_rules(pattern=pat, keys=keys)\n",
    "\n",
    "print(it2.re_rule_name)\n",
    "print(it2.re_pattern)\n",
    "print(it2.re_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "definition line: '>2591237:ncbi:1 this sequence does not match any saved parsing rule'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'seqid': '2591237:ncbi:1',\n",
       " 'seqnb': '1',\n",
       " 'source': 'ncbi',\n",
       " 'taxonomyid': '2591237',\n",
       " 'text': 'this sequence does not match any saved parsing rule'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "it2.reset_iterator()\n",
    "dfn_line, sequence = next(it2).values()\n",
    "print(f\"definition line: '{dfn_line[:-1]}'\")\n",
    "it2.parse_text(dfn_line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L53){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastaFileReader.parse_file\n",
       "\n",
       ">      FastaFileReader.parse_file (add_seq:bool=False, save_json:bool=False)\n",
       "\n",
       "Read fasta file and return a dictionary with definition line metadata and optionally sequences\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| add_seq | bool | False | When True, add the full sequence to the parsed metadata dictionary |\n",
       "| save_json | bool | False | When True, save the file metadata as a json file of same stem name |\n",
       "| **Returns** | **dict[str]** |  | **Metadata as Key/Values pairs** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L53){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastaFileReader.parse_file\n",
       "\n",
       ">      FastaFileReader.parse_file (add_seq:bool=False, save_json:bool=False)\n",
       "\n",
       "Read fasta file and return a dictionary with definition line metadata and optionally sequences\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| add_seq | bool | False | When True, add the full sequence to the parsed metadata dictionary |\n",
       "| save_json | bool | False | When True, save the file metadata as a json file of same stem name |\n",
       "| **Returns** | **dict[str]** |  | **Metadata as Key/Values pairs** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(FastaFileReader.parse_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'11128:ncbi:2': {'accession': 'LC494191',\n",
      "                  'seqid': '11128:ncbi:2',\n",
      "                  'seqnb': '2',\n",
      "                  'source': 'ncbi',\n",
      "                  'species': None,\n",
      "                  'taxonomyid': '11128'},\n",
      " '2591237:ncbi:1': {'accession': 'MK211378',\n",
      "                    'seqid': '2591237:ncbi:1',\n",
      "                    'seqnb': '1',\n",
      "                    'source': 'ncbi',\n",
      "                    'species': 'Coronavirus BtRs-BetaCoV/YN2018D  scientific '\n",
      "                               'name',\n",
      "                    'taxonomyid': '2591237'}}\n"
     ]
    }
   ],
   "source": [
    "it = FastaFileReader(p2fasta)\n",
    "pprint(it.parse_file())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metadata for 'cov_virus_sequences_two.fa'> saved as <cov_virus_sequences_two_metadata.json> in  \n",
      "/home/vtec/projects/bio/metagentools/nbs-dev/data_dev\n",
      "\n"
     ]
    }
   ],
   "source": [
    "it.parse_file(save_json=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'aln_art_illumina': {'keys': 'refseqid '\n",
      "                              'reftaxonomyid '\n",
      "                              'refsource '\n",
      "                              'refseqnb '\n",
      "                              'readid '\n",
      "                              'readnb '\n",
      "                              'aln_start_pos '\n",
      "                              'refseq_strand',\n",
      "                      'pattern': '^>(?P<refseqid>(?P<reftaxonomyid>\\\\d*):(?P<refsource>\\\\w*):(?P<refseqnb>\\\\d*))(\\\\s|\\t'\n",
      "                                 ')*(?P<readid>(?P=reftaxonomyid):(?P=refsource):(?P=refseqnb)-(?P<readnb>\\\\d*(\\\\/\\\\d(-\\\\d)?)?))(\\\\s|\\t'\n",
      "                                 ')(?P<aln_start_pos>\\\\d*)(\\\\s|\\t'\n",
      "                                 ')(?P<refseq_strand>(-|\\\\+))$'},\n",
      " 'aln_art_illumina-refseq': {'keys': 'refseqid '\n",
      "                                     'reftaxonomyid '\n",
      "                                     'refsource '\n",
      "                                     'refseqnb '\n",
      "                                     'refseq_accession '\n",
      "                                     'species '\n",
      "                                     'refseq_length',\n",
      "                             'pattern': '^@SQ[\\\\t\\\\s]*(?P<refseqid>(?P<reftaxonomyid>\\\\d*):(?P<refsource>\\\\w*):(?P<refseqnb>\\\\d*))[\\\\t\\\\s]*\\\\[(?P<refseq_accession>[\\\\d\\\\w]*)\\\\][\\\\t\\\\s]*(?P=reftaxonomyid)[\\\\s\\\\t]*(?P=refsource)[\\\\s\\\\t]*(?P=refseqnb)[\\\\s\\\\t]*\\\\[(?P=refseq_accession)\\\\][\\\\s\\\\t]*(?P=reftaxonomyid)[\\\\s\\\\t]*(?P<species>\\\\w[\\\\w\\\\d\\\\/\\\\s\\\\-\\\\.]*)[\\\\s\\\\t](?P<refseq_length>\\\\d*)$'},\n",
      " 'fasta_cov_ncbi': {'keys': 'seqid '\n",
      "                            'taxonomyid '\n",
      "                            'source '\n",
      "                            'accession '\n",
      "                            'seqnb '\n",
      "                            'species',\n",
      "                    'pattern': '^>(?P<seqid>(?P<taxonomyid>\\\\d+):(?P<source>ncbi):(?P<seqnb>\\\\d*))[\\\\s\\\\t]*\\\\[(?P<accession>[\\\\w\\\\d]*)\\\\]([\\\\s\\\\t]*(?P=taxonomyid)[\\\\s\\\\t]*(?P=source)[\\\\s\\\\t]*(?P=seqnb)[\\\\s\\\\t]*\\\\[(?P=accession)\\\\][\\\\s\\\\t]*(?P=taxonomyid)[\\\\s\\\\t]*(?P<species>[\\\\w\\\\s\\\\-\\\\_\\\\/]*))?'},\n",
      " 'fasta_rhinolophus_ferrumequinum': {'keys': 'seq_type '\n",
      "                                             'id_type '\n",
      "                                             'assy '\n",
      "                                             'seq_level '\n",
      "                                             'taxonomy '\n",
      "                                             'id',\n",
      "                                     'pattern': '^>\\\\d[\\\\s\\\\t](?P<seq_type>dna_rm):(?P<id_type>[\\\\w\\\\_]*)[\\\\s\\\\w](?P=id_type):(?P<assy>[\\\\w\\\\d\\\\_]*)\\\\.(?P<seq_level>[\\\\w]*):\\\\d*:\\\\d*:(?P<taxonomy>\\\\d*):(?P<id>\\\\d*)[\\\\s\\t'\n",
      "                                                ']REF$'},\n",
      " 'fastq_art_illumina': {'keys': 'readid '\n",
      "                                'reftaxonomyid '\n",
      "                                'refsource '\n",
      "                                'refseqnb '\n",
      "                                'readnb',\n",
      "                        'pattern': '^@(?P<readid>(?P<reftaxonomyid>\\\\d*):(?P<refsource>\\\\w*):(?P<refseqnb>\\\\d*)-(?P<readnb>\\\\d*))$'}}\n"
     ]
    }
   ],
   "source": [
    "with open('../parsing_rules.json', 'r') as fp:\n",
    "    pprint(json.load(fp), width=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metadata for 'cov_virus_sequence_one.fa'> saved as <cov_virus_sequence_one_metadata.json> in  \n",
      "/home/vtec/projects/bio/metagentools/nbs-dev/data_dev\n",
      "\n",
      "{'2591237:ncbi:1': {'accession': 'MK211378',\n",
      "                    'seqid': '2591237:ncbi:1',\n",
      "                    'seqnb': '1',\n",
      "                    'source': 'ncbi',\n",
      "                    'species': 'Coronavirus BtRs-BetaCoV/YN2018D  scientific '\n",
      "                               'name',\n",
      "                    'taxonomyid': '2591237'}}\n"
     ]
    }
   ],
   "source": [
    "p2fasta = Path('data_dev/cov_virus_sequence_one.fa').resolve()\n",
    "it = FastaFileReader(p2fasta)\n",
    "fasta_meta = it.parse_file(save_json=True)\n",
    "pprint(fasta_meta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQ file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extension of `TextFileBaseReader` class for fastq sequence files.\n",
    "\n",
    "Structure of a FASTQ sequence file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@2591237:ncbi:1-20100\n",
      "GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGT\n",
      "+\n",
      "CCC1=GGGGGGGGJJJGCGJJJGCJCJJJGGGG=GGJJJJCGGGJJCJ1GJJGJC1J8GGCJGGJJGGJGGJGGGGGGGG\n",
      "@2591237:ncbi:1-20099\n",
      "TCTTAAAATAGTTGAACTGTGTTTTTACTGATTCTCCAATTAATGTGACTCCATTGACACTAGCTTGTGCTGGTCCTTTT\n",
      "+\n",
      "CCCGGCGCGGGGGJJCJJGGCCJJGJJJGCJJGJJG1GGGJJCJJ(JGJGGJGJCGJGGCJJG=JGGJGGCGCCGGCG8G\n",
      "@2591237:ncbi:1-20098\n",
      "AAAGAAATTGACCGCCTCAATGAGGTTGCCAAAAATCTAAATGAATCGCTCATTGACCTCCAAGAACTTGGTAAATATGA\n",
      "+\n",
      "CCC1GG=CGGGGGCCGJGJJJJJJJJJJJ=J1JGCGJCGGJJCGJGGGGCJGJGCGGJJJJGJGGJGGJJGGGGJGJGJG\n"
     ]
    }
   ],
   "source": [
    "#| echo: false\n",
    "p2fastq = Path('data_dev/single_1seq_150bp/single_1seq_150bp.fq').resolve()\n",
    "\n",
    "it = TextFileBaseReader(p2fastq, nlines=1)\n",
    "for i, t in enumerate(it):\n",
    "    txt = t.replace('\\n', '')[:80]\n",
    "    print(f\"{txt}\")\n",
    "    if i >= 11: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "class FastqFileReader(TextFileBaseReader):\n",
    "    \"\"\"Iterator going through a fastq file's sequences and return each section + prob error as a dict\"\"\"\n",
    "    def __init__(\n",
    "        self,\n",
    "        path:str|Path,   # path to the fastq file\n",
    "    )-> dict:           # key/value with keys: definition line; sequence; q score; prob error\n",
    "        self.nlines = 4\n",
    "        super().__init__(path, nlines=self.nlines)\n",
    "        self.text_to_parse_key = 'definition line'\n",
    "        self.set_parsing_rules(verbose=False)        \n",
    "    \n",
    "    def __next__(self):\n",
    "        \"\"\"Return definition line, sequence and quality scores\"\"\"\n",
    "        lines = []\n",
    "        for i in range(self.nlines):\n",
    "            lines.append(self._safe_readline().replace('\\n', ''))\n",
    "        \n",
    "        output = {\n",
    "            'definition line':lines[0], \n",
    "            'sequence':f\"{lines[1]}\", \n",
    "            'read_qscores': f\"{lines[3]}\",\n",
    "        }\n",
    "        output['probs error'] = np.array([q_score2prob_error(q) for q in output['read_qscores']])\n",
    "        \n",
    "        return output\n",
    "    \n",
    "    def print_first_chunks(\n",
    "        self, \n",
    "        nchunks:int=3,  # number of chunks to print out\n",
    "    ):\n",
    "        \"\"\"Print the first `nchunks` chunks of text from the file\"\"\"\n",
    "        for i, seq_dict in enumerate(self.__iter__()):\n",
    "            print(f\"\\nSequence {i+1}:\")\n",
    "            print(seq_dict['definition line'])\n",
    "            print(f\"{seq_dict['sequence'][:80]} ...\")\n",
    "            if i >= nchunks: break\n",
    "            \n",
    "    def parse_file(\n",
    "        self,\n",
    "        add_readseq :bool=False,    # When True, add the full sequence to the parsed metadata dictionary\n",
    "        add_qscores:bool=False,     # Add the read ASCII Q Scores to the parsed dictionary when True\n",
    "        add_probs_error:bool=False, # Add the read probability of error to the parsed dictionary when True\n",
    "        save_json: bool=False       # When True, save the file metadata as a json file of same stem name\n",
    "    )-> dict[str]:                  # Metadata as Key/Values pairs\n",
    "        \"\"\"Read fastq file, return a dict with definition line metadata and optionally read sequence and q scores, ...\"\"\"\n",
    "    \n",
    "        self.reset_iterator()\n",
    "        parsed = {}\n",
    "        for d in self:\n",
    "            dfn_line = d['definition line']\n",
    "            seq, q_scores, prob_e = d['sequence'], d['read_qscores'], d['probs error']\n",
    "            metadata = self._parse_text_fn(dfn_line, self.re_pattern, self.re_keys)\n",
    "            if add_readseq: metadata['readseq'] = seq         \n",
    "            if add_qscores: metadata['read_qscores'] = q_scores\n",
    "            if add_probs_error: metadata['probs error'] = prob_e\n",
    "            parsed[metadata['readid']] = metadata \n",
    "                        \n",
    "        if save_json:\n",
    "            p2json = self.path.parent / f\"{self.path.stem}_metadata.json\"\n",
    "            with open(p2json, 'w') as fp:\n",
    "                json.dump(parsed, fp, indent=4)\n",
    "                print(f\"Metadata for '{self.path.name}'> saved as <{p2json.name}> in  \\n{p2json.parent.absolute()}\\n\")\n",
    "\n",
    "        return parsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L78){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastqFileReader\n",
       "\n",
       ">      FastqFileReader (path:str|pathlib.Path)\n",
       "\n",
       "Iterator going through a fastq file's sequences and return each section + prob error as a dict\n",
       "\n",
       "|    | **Type** | **Details** |\n",
       "| -- | -------- | ----------- |\n",
       "| path | str \\| Path | path to the fastq file |\n",
       "| **Returns** | **dict** | **key/value with keys: definition line; sequence; q score; prob error** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L78){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastqFileReader\n",
       "\n",
       ">      FastqFileReader (path:str|pathlib.Path)\n",
       "\n",
       "Iterator going through a fastq file's sequences and return each section + prob error as a dict\n",
       "\n",
       "|    | **Type** | **Details** |\n",
       "| -- | -------- | ----------- |\n",
       "| path | str \\| Path | path to the fastq file |\n",
       "| **Returns** | **dict** | **key/value with keys: definition line; sequence; q score; prob error** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(FastqFileReader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n",
      "dict_keys(['definition line', 'sequence', 'read_qscores', 'probs error'])\n",
      "Definition line:  @2591237:ncbi:1-20100\n",
      "Read sequence:    GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGTATCTACTGGTTTTGTACTCCAAAGACAACGTAAACACCAAGTATTTGGTTTAAACGTTGTCTTGGTTGTA\n",
      "Q scores (ASCII): CCC1=GGGGGGGGJJJGCGJJJGCJCJJJGGGG=GGJJJJCGGGJJCJ1GJJGJC1J8GGCJGGJJGGJGGJGGGGGGGGG8G=GGCGGGCGCGGCGCGGJCGGC=CGGGGGGCGGCGGGCGGGGCGGGGCG81CC=G=GGGCCGGGCGC\n",
      "Prob error:       0.0004,0.0004,0.0004,0.0251,0.0016,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0001,0.0001,0.0001,0.0002,0.0004,0.0002,0.0001,0.0001,0.0001,0.0002,0.0004,0.0001,0.0004,0.0001,0.0001,0.0001,0.0002,0.0002,0.0002,0.0002,0.0016,0.0002,0.0002,0.0001,0.0001,0.0001,0.0001,0.0004,0.0002,0.0002,0.0002,0.0001,0.0001,0.0004,0.0001,0.0251,0.0002,0.0001,0.0001,0.0002,0.0001,0.0004,0.0251,0.0001,0.0050,0.0002,0.0002,0.0004,0.0001,0.0002,0.0002,0.0001,0.0001,0.0002,0.0002,0.0001,0.0002,0.0002,0.0001,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0050,0.0002,0.0016,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0004,0.0002,0.0004,0.0002,0.0002,0.0004,0.0002,0.0004,0.0002,0.0002,0.0001,0.0004,0.0002,0.0002,0.0004,0.0016,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0004,0.0002,0.0050,0.0251,0.0004,0.0004,0.0016,0.0002,0.0016,0.0002,0.0002,0.0002,0.0004,0.0004,0.0002,0.0002,0.0002,0.0004,0.0002,0.0004\n"
     ]
    }
   ],
   "source": [
    "it = FastqFileReader(p2fastq)\n",
    "iteration_output = next(it)\n",
    "\n",
    "print(type(iteration_output))\n",
    "print(iteration_output.keys())\n",
    "print(f\"Definition line:  {iteration_output['definition line']}\")\n",
    "print(f\"Read sequence:    {iteration_output['sequence']}\")\n",
    "print(f\"Q scores (ASCII): {iteration_output['read_qscores']}\")\n",
    "print(f\"Prob error:       {','.join([f'{p:.4f}' for p in iteration_output['probs error']])}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Five largest probabilities of error:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.00501187, 0.02511886, 0.02511886, 0.02511886, 0.02511886])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sort(iteration_output['probs error'])[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 81,  48,   3,  55, 133])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.argsort(iteration_output['probs error'])[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'readid': '2591237:ncbi:1-20100',\n",
       " 'readnb': '20100',\n",
       " 'refseqnb': '1',\n",
       " 'refsource': 'ncbi',\n",
       " 'reftaxonomyid': '2591237'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfn_line = iteration_output['definition line']\n",
    "meta = it.parse_text(dfn_line)\n",
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['definition line', 'sequence', 'read_qscores', 'probs error'])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq = FastqFileReader(p2fastq)\n",
    "next(fastq).keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L115){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastqFileReader.parse_file\n",
       "\n",
       ">      FastqFileReader.parse_file (add_readseq:bool=False,\n",
       ">                                  add_qscores:bool=False,\n",
       ">                                  add_probs_error:bool=False,\n",
       ">                                  save_json:bool=False)\n",
       "\n",
       "Read fastq file, return a dict with definition line metadata and optionally read sequence and q scores, ...\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| add_readseq | bool | False | When True, add the full sequence to the parsed metadata dictionary |\n",
       "| add_qscores | bool | False | Add the read ASCII Q Scores to the parsed dictionary when True |\n",
       "| add_probs_error | bool | False | Add the read probability of error to the parsed dictionary when True |\n",
       "| save_json | bool | False | When True, save the file metadata as a json file of same stem name |\n",
       "| **Returns** | **dict[str]** |  | **Metadata as Key/Values pairs** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L115){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### FastqFileReader.parse_file\n",
       "\n",
       ">      FastqFileReader.parse_file (add_readseq:bool=False,\n",
       ">                                  add_qscores:bool=False,\n",
       ">                                  add_probs_error:bool=False,\n",
       ">                                  save_json:bool=False)\n",
       "\n",
       "Read fastq file, return a dict with definition line metadata and optionally read sequence and q scores, ...\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| add_readseq | bool | False | When True, add the full sequence to the parsed metadata dictionary |\n",
       "| add_qscores | bool | False | Add the read ASCII Q Scores to the parsed dictionary when True |\n",
       "| add_probs_error | bool | False | Add the read probability of error to the parsed dictionary when True |\n",
       "| save_json | bool | False | When True, save the file metadata as a json file of same stem name |\n",
       "| **Returns** | **dict[str]** |  | **Metadata as Key/Values pairs** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(FastqFileReader.parse_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2591237:ncbi:1-20100\n",
      "{'readid': '2591237:ncbi:1-20100',\n",
      " 'readnb': '20100',\n",
      " 'refseqnb': '1',\n",
      " 'refsource': 'ncbi',\n",
      " 'reftaxonomyid': '2591237'}\n",
      "2591237:ncbi:1-20099\n",
      "{'readid': '2591237:ncbi:1-20099',\n",
      " 'readnb': '20099',\n",
      " 'refseqnb': '1',\n",
      " 'refsource': 'ncbi',\n",
      " 'reftaxonomyid': '2591237'}\n",
      "2591237:ncbi:1-20098\n",
      "{'readid': '2591237:ncbi:1-20098',\n",
      " 'readnb': '20098',\n",
      " 'refseqnb': '1',\n",
      " 'refsource': 'ncbi',\n",
      " 'reftaxonomyid': '2591237'}\n",
      "2591237:ncbi:1-20097\n",
      "{'readid': '2591237:ncbi:1-20097',\n",
      " 'readnb': '20097',\n",
      " 'refseqnb': '1',\n",
      " 'refsource': 'ncbi',\n",
      " 'reftaxonomyid': '2591237'}\n"
     ]
    }
   ],
   "source": [
    "parsed = fastq.parse_file(add_readseq=False, add_qscores=False, add_probs_error=False)\n",
    "for i, (k, v) in enumerate(parsed.items()):\n",
    "    print(k)\n",
    "    pprint(v)\n",
    "    if i >=3: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>readid</th>\n",
       "      <th>readnb</th>\n",
       "      <th>refseqnb</th>\n",
       "      <th>refsource</th>\n",
       "      <th>reftaxonomyid</th>\n",
       "      <th>readseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20100</th>\n",
       "      <td>2591237:ncbi:1-20100</td>\n",
       "      <td>20100</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20099</th>\n",
       "      <td>2591237:ncbi:1-20099</td>\n",
       "      <td>20099</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>TCTTAAAATAGTTGAACTGTGTTTTTACTGATTCTCCAATTAATGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20098</th>\n",
       "      <td>2591237:ncbi:1-20098</td>\n",
       "      <td>20098</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>AAAGAAATTGACCGCCTCAATGAGGTTGCCAAAAATCTAAATGAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20097</th>\n",
       "      <td>2591237:ncbi:1-20097</td>\n",
       "      <td>20097</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>ACATTGACTTACCCGCGCGAGCGAGAAAATCTTTGATGCAATCAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20096</th>\n",
       "      <td>2591237:ncbi:1-20096</td>\n",
       "      <td>20096</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>TGAGTAAAAATTCCTCTGTGTAATGAACCAAGAAGTGCCATTAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20095</th>\n",
       "      <td>2591237:ncbi:1-20095</td>\n",
       "      <td>20095</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>GATCACCAACAGTGTCATTATTATGGCATATGTAACTGGTGGTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20094</th>\n",
       "      <td>2591237:ncbi:1-20094</td>\n",
       "      <td>20094</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>GAATGCAGTATACAAGAATGTTGTGAAAATCAACCATTCCAACCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20093</th>\n",
       "      <td>2591237:ncbi:1-20093</td>\n",
       "      <td>20093</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>AATTCTCAGAATAACCACCAATTTGGTAGTCTTCTTTGAGTTTTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20092</th>\n",
       "      <td>2591237:ncbi:1-20092</td>\n",
       "      <td>20092</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>GCAGACAATGTATGTACTGGCATAACGATGTAACAGACCAATGCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591237:ncbi:1-20091</th>\n",
       "      <td>2591237:ncbi:1-20091</td>\n",
       "      <td>20091</td>\n",
       "      <td>1</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>2591237</td>\n",
       "      <td>GTTGAATTCACTTTGAATTTGTCAAAACACTCTACGCGCGCACGCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    readid readnb refseqnb refsource  \\\n",
       "2591237:ncbi:1-20100  2591237:ncbi:1-20100  20100        1      ncbi   \n",
       "2591237:ncbi:1-20099  2591237:ncbi:1-20099  20099        1      ncbi   \n",
       "2591237:ncbi:1-20098  2591237:ncbi:1-20098  20098        1      ncbi   \n",
       "2591237:ncbi:1-20097  2591237:ncbi:1-20097  20097        1      ncbi   \n",
       "2591237:ncbi:1-20096  2591237:ncbi:1-20096  20096        1      ncbi   \n",
       "2591237:ncbi:1-20095  2591237:ncbi:1-20095  20095        1      ncbi   \n",
       "2591237:ncbi:1-20094  2591237:ncbi:1-20094  20094        1      ncbi   \n",
       "2591237:ncbi:1-20093  2591237:ncbi:1-20093  20093        1      ncbi   \n",
       "2591237:ncbi:1-20092  2591237:ncbi:1-20092  20092        1      ncbi   \n",
       "2591237:ncbi:1-20091  2591237:ncbi:1-20091  20091        1      ncbi   \n",
       "\n",
       "                     reftaxonomyid  \\\n",
       "2591237:ncbi:1-20100       2591237   \n",
       "2591237:ncbi:1-20099       2591237   \n",
       "2591237:ncbi:1-20098       2591237   \n",
       "2591237:ncbi:1-20097       2591237   \n",
       "2591237:ncbi:1-20096       2591237   \n",
       "2591237:ncbi:1-20095       2591237   \n",
       "2591237:ncbi:1-20094       2591237   \n",
       "2591237:ncbi:1-20093       2591237   \n",
       "2591237:ncbi:1-20092       2591237   \n",
       "2591237:ncbi:1-20091       2591237   \n",
       "\n",
       "                                                                readseq  \n",
       "2591237:ncbi:1-20100  GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGT...  \n",
       "2591237:ncbi:1-20099  TCTTAAAATAGTTGAACTGTGTTTTTACTGATTCTCCAATTAATGT...  \n",
       "2591237:ncbi:1-20098  AAAGAAATTGACCGCCTCAATGAGGTTGCCAAAAATCTAAATGAAT...  \n",
       "2591237:ncbi:1-20097  ACATTGACTTACCCGCGCGAGCGAGAAAATCTTTGATGCAATCAAG...  \n",
       "2591237:ncbi:1-20096  TGAGTAAAAATTCCTCTGTGTAATGAACCAAGAAGTGCCATTAGAT...  \n",
       "2591237:ncbi:1-20095  GATCACCAACAGTGTCATTATTATGGCATATGTAACTGGTGGTCTT...  \n",
       "2591237:ncbi:1-20094  GAATGCAGTATACAAGAATGTTGTGAAAATCAACCATTCCAACCTG...  \n",
       "2591237:ncbi:1-20093  AATTCTCAGAATAACCACCAATTTGGTAGTCTTCTTTGAGTTTTGG...  \n",
       "2591237:ncbi:1-20092  GCAGACAATGTATGTACTGGCATAACGATGTAACAGACCAATGCAG...  \n",
       "2591237:ncbi:1-20091  GTTGAATTCACTTTGAATTTGTCAAAACACTCTACGCGCGCACGCG...  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = it.parse_file(add_readseq=True)\n",
    "df = pd.DataFrame(metadata).T\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------------------------------\n",
      "Rule <fasta_cov_ncbi> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <fasta_rhinolophus_ferrumequinum> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <fastq_art_illumina> generated 5 matches\n",
      "--------------------------------------------------------------------------------\n",
      "^@(?P<readid>(?P<reftaxonomyid>\\d*):(?P<refsource>\\w*):(?P<refseqnb>\\d*)-(?P<readnb>\\d*))$\n",
      "['readid', 'reftaxonomyid', 'refsource', 'refseqnb', 'readnb']\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <aln_art_illumina> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <aln_art_illumina-refseq> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Selected rule with most matches: fastq_art_illumina\n"
     ]
    }
   ],
   "source": [
    "fastq.set_parsing_rules(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ALN Alignment Files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extension of `TextFileBaseReader` class for ALN read/sequence alignment files.\n",
    "\n",
    "Structure of a ALN sequence file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##ART_Illumina\tread_length\t150\n",
      "@CM\t/bin/art_illumina -i /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/c\n",
      "@SQ\t2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs\n",
      "@SQ\t7321952:ncbi:2 [MK873112]\t7321952\tncbi\t2 [MK873112] 7321952\tNot a real virus\n",
      "##Header End\n",
      ">2591237:ncbi:1\t2591237:ncbi:1-20100\t23878\t-\n",
      "GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGT\n",
      "GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGT\n",
      ">2591237:ncbi:1\t2591237:ncbi:1-20099\t10122\t-\n",
      "TCTTAAAATAGTTGAACTGTGTTTTTACTGATTCTCCAATTAATGTGACTCCATTGACACTAGCTTGTGCTGGTCCTTTT\n",
      "TCTTAAAATAGTTGAACTGTGTTTTTACTGATTCTCCAATTAATGTGACTCCATTGACACTAGCTTGTGCTGGTCCTTTT\n",
      ">2591237:ncbi:1\t2591237:ncbi:1-20098\t24935\t+\n",
      "AAAGAAATTGACCGCCTCAATGAGGTTGCCAAAAATCTAAATGAATCGCTCATTGACCTCCAAGAACTTGGTAAATATGA\n"
     ]
    }
   ],
   "source": [
    "#| echo: false\n",
    "p2aln = Path('data_dev/single_1seq_150bp/single_1seq_150bp.aln').resolve()\n",
    "assert p2aln.is_file()\n",
    "\n",
    "it = TextFileBaseReader(p2aln, nlines=1)\n",
    "for i, t in enumerate(it):\n",
    "    txt = t.replace('\\n', '')[:80]\n",
    "    print(f\"{txt}\")\n",
    "    if i >= 12: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "class AlnFileReader(TextFileBaseReader):\n",
    "    \"\"\"Iterator going through an ALN file\"\"\"\n",
    "    def __init__(\n",
    "        self,\n",
    "        path:str|Path,   # path to the aln file\n",
    "    )-> dict:            # key/value with keys: \n",
    "        self.nlines = 1\n",
    "        super().__init__(path, nlines=self.nlines)\n",
    "        self.header = self.read_header()\n",
    "        self.nlines = 3\n",
    "        self.text_to_parse_key = 'definition line'\n",
    "        self.set_parsing_rules(verbose=False)\n",
    "        self.set_header_parsing_rules(verbose=False)\n",
    "\n",
    "    def __next__(self):\n",
    "        \"\"\"Return definition line, sequence and quality scores\"\"\"\n",
    "        lines = []\n",
    "        for i in range(self.nlines):\n",
    "            lines.append(self._safe_readline().replace('\\n', ''))\n",
    "\n",
    "        output = {\n",
    "            'definition line':lines[0], \n",
    "            'ref_seq_aligned':f\"{lines[1]}\", \n",
    "            'read_seq_aligned': f\"{lines[2]}\",\n",
    "        }   \n",
    "        return output\n",
    "    \n",
    "    def read_header(self):\n",
    "        \"\"\"Read ALN file Header and return each section parsed in a dictionary\"\"\"\n",
    "        \n",
    "        header = {}\n",
    "        if self.fp is not None:\n",
    "            self.fp.close()\n",
    "        self.fp = open(self.path, 'r')\n",
    "        \n",
    "        line = self._safe_readline().replace('\\n', '')\n",
    "        if not line.startswith('##ART_Illumina'): \n",
    "            raise ValueError(f\"Header of this file does not start with ##ART_Illumina\")\n",
    "        line = self._safe_readline().replace('\\n', '')\n",
    "        if not line.startswith('@CM'): \n",
    "            raise ValueError(f\"First header line should start with @CM\")\n",
    "        else: \n",
    "            header['command'] = line[3:].replace('\\t', '').strip()\n",
    "\n",
    "        refseqs = []\n",
    "        while True:\n",
    "            line = self._safe_readline().replace('\\n', '')\n",
    "            if line.startswith('##Header End'): break\n",
    "            else:\n",
    "                refseqs.append(line)\n",
    "        header['reference sequences'] = refseqs\n",
    "        \n",
    "        return header\n",
    "    \n",
    "    def reset_iterator(self):\n",
    "        \"\"\"Reset the iterator to point to the first line in the file, by recreating a new file handle.\n",
    "        \n",
    "        `AlnFileReader` requires a specific `reset_iterator` method, in order to skip the header every time it is reset\n",
    "        \"\"\"\n",
    "        if self.fp is not None:\n",
    "            self.fp.close()\n",
    "        self.fp = open(self.path, 'r')\n",
    "        while True:\n",
    "            line = self._safe_readline().replace('\\n', '')\n",
    "            if line.startswith('##Header End'): break\n",
    "    \n",
    "    def parse_file(\n",
    "        self, \n",
    "        add_ref_seq_aligned:bool=False,   # Add the reference sequence aligned to the parsed dictionary when True\n",
    "        add_read_seq_aligned:bool=False,  # Add the read sequence aligned to the parsed dictionary when True\n",
    "    )-> dict[str]: \n",
    "        # Key/Values. Keys: \n",
    "        # `readid`,`seqid`,`seq_nbr`,`read_nbr`,`aln_start_pos`,`ref_seq_strand`\n",
    "        # optionaly `ref_seq_aligned`,`read_seq_aligned`\n",
    "        \"\"\"Read ALN file, return a dict w/ alignment info for each read and optionaly aligned reference sequence & read\"\"\"\n",
    "        self.reset_iterator()\n",
    "        parsed = {}\n",
    "        for d in self:\n",
    "            dfn_line = d['definition line']\n",
    "            ref_seq_aligned, read_seq_aligned = d['ref_seq_aligned'], d['read_seq_aligned']\n",
    "            metadata = self.parse_text(dfn_line)\n",
    "            if add_ref_seq_aligned: metadata['ref_seq_aligned'] = ref_seq_aligned         \n",
    "            if add_read_seq_aligned: metadata['read_seq_aligned'] = read_seq_aligned\n",
    "            parsed[metadata['readid']] = metadata \n",
    "        return parsed\n",
    "\n",
    "    def parse_header_reference_sequences(\n",
    "        self,\n",
    "        pattern:str|None=None,     # regex pattern to apply to parse the reference sequence info\n",
    "        keys:list[str]|None=None,  # list of keys: keys are both regex match group names and corresponding output dict keys \n",
    "        )->dict[str]:                  # parsed metadata in key/value format\n",
    "        \"\"\"Extract metadata from all header reference sequences\"\"\"\n",
    "        if pattern is None and keys is None:\n",
    "            pattern, keys = self.re_header_pattern, self.re_header_keys\n",
    "        parsed = {}\n",
    "        for seq_dfn_line in self.header['reference sequences']:\n",
    "            metadata = self.parse_text(seq_dfn_line, pattern, keys)\n",
    "            parsed[metadata['refseqid']] = metadata\n",
    "            \n",
    "        return parsed\n",
    "            \n",
    "        \n",
    "    def set_header_parsing_rules(\n",
    "        self,\n",
    "        pattern: str|bool=None,   # regex pattern to apply to parse the text, search in parsing rules json if None\n",
    "        keys: list[str]=None,     # list of keys/group for regex, search in parsing rules json if None\n",
    "        verbose: bool=False       # when True, provides information on each rule\n",
    "    )-> None:\n",
    "        \"\"\"Set the regex parsing rule for reference sequence in ALN header.\n",
    "               \n",
    "        Updates 3 class attributes: `re_header_rule_name`, `re_header_pattern`, `re_header_keys`\n",
    "        \n",
    "        TODO: refactor this and the method in Core: to use a single function for the common part and a parameter for               the text_to_parse \n",
    "        \"\"\"\n",
    "        \n",
    "        P2JSON = Path('../parsing_rules.json')\n",
    "        \n",
    "        self.re_header_rule_name = None\n",
    "        self.re_header_pattern = None\n",
    "        self.re_header_keys = None\n",
    "        \n",
    "        # get the first reference sequence definition line in header\n",
    "        text_to_parse = self.header['reference sequences'][0]\n",
    "        divider_line = f\"{'-'*80}\"\n",
    "\n",
    "        if pattern is not None and keys is not None:  # When specific pattern and keys are passed\n",
    "            try:\n",
    "                metadata_dict = self.parse_text(text_to_parse, pattern, keys)\n",
    "                self.re_header_rule_name = 'Custom Rule'\n",
    "                self.re_header_pattern = pattern\n",
    "                self.re_header_keys = keys\n",
    "                if verbose:\n",
    "                    print(divider_line)\n",
    "                    print(f\"Custom rule was set for header in this instance.\")\n",
    "            except Exception as err: \n",
    "                raise ValueError(f\"The pattern generates the following error:\\n{err}\")\n",
    "                \n",
    "        else:  # automatic rule selection among rules saved in json file\n",
    "            # Load all existing rules from json file\n",
    "            with open(P2JSON, 'r') as fp:\n",
    "                parsing_rules = json.load(fp)\n",
    "                \n",
    "            # test all existing rules and keep the one with highest number of matches\n",
    "            max_nbr_matches = 0\n",
    "            for k, v in parsing_rules.items():\n",
    "                re_header_pattern = v['pattern']\n",
    "                re_header_keys = v['keys'].split(' ')\n",
    "                try:\n",
    "                    metadata_dict = self.parse_text(text_to_parse, re_header_pattern, re_header_keys)\n",
    "                    nbr_matches = len(metadata_dict)\n",
    "                    if verbose:\n",
    "                        print(divider_line)\n",
    "                        print(f\"Rule <{k}> generated {nbr_matches:,d} matches\")\n",
    "                        print(divider_line)\n",
    "                        print(re_header_pattern)\n",
    "                        print(re_header_keys)\n",
    "\n",
    "                    if len(metadata_dict) > max_nbr_matches:\n",
    "                        self.re_header_pattern = re_header_pattern\n",
    "                        self.re_header_keys = re_header_keys\n",
    "                        self.re_header_rule_name = k    \n",
    "                except Exception as err:\n",
    "                    if verbose:\n",
    "                        print(divider_line)\n",
    "                        print(f\"Rule <{k}> generated an error\")\n",
    "                        print(err)\n",
    "                    else:\n",
    "                        pass\n",
    "            if self.re_header_rule_name is None:\n",
    "                msg = \"\"\"\n",
    "        None of the saved parsing rules were able to extract metadata from the first line in this file.\n",
    "        You must set a custom rule (pattern + keys) before parsing text, by using:\n",
    "            `self.set_parsing_rules(custom_pattern, custom_list_of_keys)`\n",
    "                \"\"\"\n",
    "                warnings.warn(msg, category=UserWarning)\n",
    "            \n",
    "            if verbose:\n",
    "                print(divider_line)\n",
    "                print(f\"Selected rule with most matches: {self.re_header_rule_name}\")\n",
    "\n",
    "            # We used the iterator, now we need to reset it to make all lines available\n",
    "            self.reset_iterator()\n",
    "    \n",
    "    @property\n",
    "    def ref_sequences(self):\n",
    "        refseq_metadata = {}\n",
    "        for refseq in self.header['reference sequences']:\n",
    "            \n",
    "            meta = parse_art_read_aln_refseqs(refseq)\n",
    "            refseq_metadata[meta['refseqid']] = meta\n",
    "        \n",
    "        return refseq_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L144){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### AlnFileReader\n",
       "\n",
       ">      AlnFileReader (path:str|pathlib.Path)\n",
       "\n",
       "Iterator going through an ALN file\n",
       "\n",
       "|    | **Type** | **Details** |\n",
       "| -- | -------- | ----------- |\n",
       "| path | str \\| Path | path to the aln file |\n",
       "| **Returns** | **dict** | **key/value with keys:** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L144){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### AlnFileReader\n",
       "\n",
       ">      AlnFileReader (path:str|pathlib.Path)\n",
       "\n",
       "Iterator going through an ALN file\n",
       "\n",
       "|    | **Type** | **Details** |\n",
       "| -- | -------- | ----------- |\n",
       "| path | str \\| Path | path to the aln file |\n",
       "| **Returns** | **dict** | **key/value with keys:** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(AlnFileReader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "it = AlnFileReader(p2aln)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`AlnFileReader` iterator returns elements one by one, as dictionaries with each data line related to the read, accessible through the following keys: \n",
    "\n",
    "- key `'definition line'`: **read definition line**, including read metadata \n",
    "- key `'ref_seq_aligned'`: **aligned reference sequence**, that is the sequence segment in the original reference corresponding to the read\n",
    "- key `'read_seq_aligned'`: **aligned read**, that is the simmulated read sequence, where each bp corresponds to the reference sequence bp in the same position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['definition line', 'ref_seq_aligned', 'read_seq_aligned'])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "one_iteration = next(it)\n",
    "one_iteration.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'definition line': '>2591237:ncbi:1\\t2591237:ncbi:1-20100\\t23878\\t-',\n",
      " 'read_seq_aligned': 'GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGTATCTACTGGTTTTGTACTCCAAAGACAACGTAAACACCAAGTATTTGGTTTAAACGTTGTCTTGGTTGTA',\n",
      " 'ref_seq_aligned': 'GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGTATCTACTGGTTTTGTACTCCAAAGACAACGTAAACACCAAGTATTTGGTTTAAACGTTGTCTTGGTTGTA'}\n"
     ]
    }
   ],
   "source": [
    "pprint(one_iteration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfn_line, ref_seq_aligned, read_seq_aligned = one_iteration.values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'>2591237:ncbi:1\\t2591237:ncbi:1-20100\\t23878\\t-'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfn_line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGTATCTACTGGTTTTGTACTCC'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seq_aligned[:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGTATCTACTGGTTTTGTACTCC'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_seq_aligned[:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'definition line': '>2591237:ncbi:1\\t2591237:ncbi:1-20099\\t10122\\t-',\n",
      " 'read_seq_aligned': 'TCTTAAAATAGTTGAACTGTGTTTTTACTGATTCTCCAATTAATGTGACTCCATTGACACTAGCTTGTGCTGGTCCTTTTGAAGGTGTTAAACCTTTAACTGAACCTTCTGTTATTAAAACACCATTACGGGCATTTCTATAAAGGTCTA',\n",
      " 'ref_seq_aligned': 'TCTTAAAATAGTTGAACTGTGTTTTTACTGATTCTCCAATTAATGTGACTCCATTGACACTAGCTTGTGCTGGTCCTTTTGAAGGTGTTAAACCTTTAACTGAACCTTCTGTTATTAAAACACCATTACGGGCATTTCTAAAAAGGTCTA'}\n"
     ]
    }
   ],
   "source": [
    "another_iteration = next(it)\n",
    "pprint(another_iteration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">2591237:ncbi:1\t2591237:ncbi:1-20100\t23878\t-\n",
      "GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGT ...\n",
      "GTGGGTTGTTGACTTTCACAAGCAAGATTGTCCATTCCTTGTGTGTCTTCTACTGCCAGAACTTCAAATGAATTTGAAGT ...\n",
      "\n",
      ">2591237:ncbi:1\t2591237:ncbi:1-20099\t10122\t-\n",
      "TCTTAAAATAGTTGAACTGTGTTTTTACTGATTCTCCAATTAATGTGACTCCATTGACACTAGCTTGTGCTGGTCCTTTT ...\n",
      "TCTTAAAATAGTTGAACTGTGTTTTTACTGATTCTCCAATTAATGTGACTCCATTGACACTAGCTTGTGCTGGTCCTTTT ...\n",
      "\n",
      ">2591237:ncbi:1\t2591237:ncbi:1-20098\t24935\t+\n",
      "AAAGAAATTGACCGCCTCAATGAGGTTGCCAAAAATCTAAATGAATCGCTCATTGACCTCCAAGAACTTGGTAAATATGA ...\n",
      "AAAGAAATTGACCGCCTCAATGAGGTTGCCAAAAATCTAAATGAATCGCTCATTGACCTCCAAGAACTTGGTAAATATGA ...\n",
      "\n",
      ">2591237:ncbi:1\t2591237:ncbi:1-20097\t29312\t-\n",
      "ACATTGACTTACCCGCGCGAGCGAGAAAATCTTTGATGCAATCAAGAGGGTACCCATCTGGGCCACAGAAGTTGTTGTCG ...\n",
      "ACATTGACTTACCCGCGCGAGCGAGAAAATCTTTGATGCAATCAAGAGGGTACCCATCTGGGCCACAGAAGTTGTTGTCG ...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "it.reset_iterator()\n",
    "for i, d in enumerate(it):\n",
    "    print(d['definition line'])\n",
    "    print(d['ref_seq_aligned'][:80], '...')\n",
    "    print(d['read_seq_aligned'][:80], '...\\n')\n",
    "    if i >= 3: break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once instantiated, the `AlnFileReader`iterator gives access to the file's header information through `header` instance attribute. It is a dictionary with two keys: `'command'` and `'reference sequences'`:\n",
    "\n",
    "```\n",
    "    {'command':             'art-illumina command used to create the reads',\n",
    "     'reference sequences': ['@SQ metadata on reference sequence 1 used for the reads',\n",
    "                             '@SQ metadata on reference sequence 2 used for the reads', \n",
    "                             ...\n",
    "                            ]\n",
    "    }\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/art_illumina -i /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/cov_virus_sequence_one.fa -ss HS25 -l 150 -f 100 -o /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/single_1seq_150bp/single_1seq_150bp -rs 1674645489\n"
     ]
    }
   ],
   "source": [
    "print(it.header['command'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SQ\t2591237:ncbi:1 [MK211378]\t2591237\tncbi\t1 [MK211378] 2591237\tCoronavirus BtRs-BetaCoV/YN2018D\t\tscientific name\t30213\n",
      "@SQ\t7321952:ncbi:2 [MK873112]\t7321952\tncbi\t2 [MK873112] 7321952\tNot a real virus just here for code testing\tscientific name\t31203 \n"
     ]
    }
   ],
   "source": [
    "for seq_info in it.header['reference sequences']:\n",
    "    print(seq_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **read definition line** includes key metadata, which need to be parsed using the appropriate parsing rule."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/core.py#LNone){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### TextFileBaseReader.parse_text\n",
       "\n",
       ">      TextFileBaseReader.parse_text (txt:str, pattern:str=None,\n",
       ">                                     keys:list[str]=None)\n",
       "\n",
       "Parse text using regex pattern and key. Return a metadata dictionary\n",
       "\n",
       "The passed text is parsed using the regex pattern. The method return a dictionary in the format:\n",
       "    {\n",
       "        'key_1': 'metadata 1',\n",
       "        'key_2': 'metadata 2',\n",
       "        ...\n",
       "    }\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| txt | str |  | text to parse |\n",
       "| pattern | str | None | If None, uses standard regex pattern to extract metadata, otherwise, uses passed regex |\n",
       "| keys | list | None | If None, uses standard regex list of keys, otherwise, uses passed list of keys (str) |\n",
       "| **Returns** | **dict** |  | **parsed metadata in key/value format** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/core.py#LNone){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### TextFileBaseReader.parse_text\n",
       "\n",
       ">      TextFileBaseReader.parse_text (txt:str, pattern:str=None,\n",
       ">                                     keys:list[str]=None)\n",
       "\n",
       "Parse text using regex pattern and key. Return a metadata dictionary\n",
       "\n",
       "The passed text is parsed using the regex pattern. The method return a dictionary in the format:\n",
       "    {\n",
       "        'key_1': 'metadata 1',\n",
       "        'key_2': 'metadata 2',\n",
       "        ...\n",
       "    }\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| txt | str |  | text to parse |\n",
       "| pattern | str | None | If None, uses standard regex pattern to extract metadata, otherwise, uses passed regex |\n",
       "| keys | list | None | If None, uses standard regex list of keys, otherwise, uses passed list of keys (str) |\n",
       "| **Returns** | **dict** |  | **parsed metadata in key/value format** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(AlnFileReader.parse_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "pattern, keys = it.re_pattern, it.re_keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aln_start_pos': '23878',\n",
       " 'readid': '2591237:ncbi:1-20100',\n",
       " 'readnb': '20100',\n",
       " 'refseq_strand': '-',\n",
       " 'refseqid': '2591237:ncbi:1',\n",
       " 'refseqnb': '1',\n",
       " 'refsource': 'ncbi',\n",
       " 'reftaxonomyid': '2591237'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "it.parse_text(dfn_line, pattern, keys)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Upon instance creation, `AlnFileReader` automatically checks the `parsing_rules.json` file for a workable rule among saved rules. Saved rules include the rule for ART Illumina ALN files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'aln_art_illumina'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "it.re_rule_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is therefore not required to pass a specific `pattern` and `keys` parameter.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aln_start_pos': '23878',\n",
       " 'readid': '2591237:ncbi:1-20100',\n",
       " 'readnb': '20100',\n",
       " 'refseq_strand': '-',\n",
       " 'refseqid': '2591237:ncbi:1',\n",
       " 'refseqnb': '1',\n",
       " 'refsource': 'ncbi',\n",
       " 'reftaxonomyid': '2591237'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "it.parse_text(dfn_line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ART Ilumina ALN files definition lines consist of:\n",
    "\n",
    "- The **read** ID: `readid`, e.g. `2591237:ncbi:1-20100`\n",
    "- the **read** number (order in the file): `readnb`, e.g. `20100`\n",
    "- The **read** start position in the reference sequence: `aln_start_pos`, e.g. `23878`\n",
    "- The **reference sequence** ID: `readid`, e.g. `2591237:ncbi:1-20100`\n",
    "- The **reference sequence** number: `refseqnb`, e.g. `1`\n",
    "- The **reference sequence** source: `refsource`, e.g. `ncbi`\n",
    "- The **reference sequence** taxonomy: `reftaxonomyid`, e.g. `2591237`\n",
    "- The **reference sequence** strand:  `refseq_strand` wich is either `+` or  `-`,\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L210){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### AlnFileReader.parse_file\n",
       "\n",
       ">      AlnFileReader.parse_file (add_ref_seq_aligned:bool=False,\n",
       ">                                add_read_seq_aligned:bool=False)\n",
       "\n",
       "Read ALN file, return a dict w/ alignment info for each read and optionaly aligned reference sequence & read\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| add_ref_seq_aligned | bool | False | Add the reference sequence aligned to the parsed dictionary when True |\n",
       "| add_read_seq_aligned | bool | False | Add the read sequence aligned to the parsed dictionary when True |\n",
       "| **Returns** | **dict[str]** |  |  |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L210){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### AlnFileReader.parse_file\n",
       "\n",
       ">      AlnFileReader.parse_file (add_ref_seq_aligned:bool=False,\n",
       ">                                add_read_seq_aligned:bool=False)\n",
       "\n",
       "Read ALN file, return a dict w/ alignment info for each read and optionaly aligned reference sequence & read\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| add_ref_seq_aligned | bool | False | Add the reference sequence aligned to the parsed dictionary when True |\n",
       "| add_read_seq_aligned | bool | False | Add the read sequence aligned to the parsed dictionary when True |\n",
       "| **Returns** | **dict[str]** |  |  |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(AlnFileReader.parse_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2591237:ncbi:1-20100\n",
      "{'aln_start_pos': '23878',\n",
      " 'readid': '2591237:ncbi:1-20100',\n",
      " 'readnb': '20100',\n",
      " 'refseq_strand': '-',\n",
      " 'refseqid': '2591237:ncbi:1',\n",
      " 'refseqnb': '1',\n",
      " 'refsource': 'ncbi',\n",
      " 'reftaxonomyid': '2591237'}\n",
      "2591237:ncbi:1-20099\n",
      "{'aln_start_pos': '10122',\n",
      " 'readid': '2591237:ncbi:1-20099',\n",
      " 'readnb': '20099',\n",
      " 'refseq_strand': '-',\n",
      " 'refseqid': '2591237:ncbi:1',\n",
      " 'refseqnb': '1',\n",
      " 'refsource': 'ncbi',\n",
      " 'reftaxonomyid': '2591237'}\n",
      "2591237:ncbi:1-20098\n",
      "{'aln_start_pos': '24935',\n",
      " 'readid': '2591237:ncbi:1-20098',\n",
      " 'readnb': '20098',\n",
      " 'refseq_strand': '+',\n",
      " 'refseqid': '2591237:ncbi:1',\n",
      " 'refseqnb': '1',\n",
      " 'refsource': 'ncbi',\n",
      " 'reftaxonomyid': '2591237'}\n",
      "2591237:ncbi:1-20097\n",
      "{'aln_start_pos': '29312',\n",
      " 'readid': '2591237:ncbi:1-20097',\n",
      " 'readnb': '20097',\n",
      " 'refseq_strand': '-',\n",
      " 'refseqid': '2591237:ncbi:1',\n",
      " 'refseqnb': '1',\n",
      " 'refsource': 'ncbi',\n",
      " 'reftaxonomyid': '2591237'}\n",
      "2591237:ncbi:1-20096\n",
      "{'aln_start_pos': '5485',\n",
      " 'readid': '2591237:ncbi:1-20096',\n",
      " 'readnb': '20096',\n",
      " 'refseq_strand': '-',\n",
      " 'refseqid': '2591237:ncbi:1',\n",
      " 'refseqnb': '1',\n",
      " 'refsource': 'ncbi',\n",
      " 'reftaxonomyid': '2591237'}\n"
     ]
    }
   ],
   "source": [
    "parsed = it.parse_file()\n",
    "\n",
    "for i, (k, v) in enumerate(parsed.items()):\n",
    "    print(k)\n",
    "    pprint(v)\n",
    "    if i > 3: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L230){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### AlnFileReader.parse_header_reference_sequences\n",
       "\n",
       ">      AlnFileReader.parse_header_reference_sequences (pattern:str|None=None,\n",
       ">                                                      keys:list[str]|None=None)\n",
       "\n",
       "Extract metadata from all header reference sequences\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| pattern | str \\| None | None | regex pattern to apply to parse the reference sequence info |\n",
       "| keys | list[str] \\| None | None | list of keys: keys are both regex match group names and corresponding output dict keys |\n",
       "| **Returns** | **dict[str]** |  | **parsed metadata in key/value format** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L230){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### AlnFileReader.parse_header_reference_sequences\n",
       "\n",
       ">      AlnFileReader.parse_header_reference_sequences (pattern:str|None=None,\n",
       ">                                                      keys:list[str]|None=None)\n",
       "\n",
       "Extract metadata from all header reference sequences\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| pattern | str \\| None | None | regex pattern to apply to parse the reference sequence info |\n",
       "| keys | list[str] \\| None | None | list of keys: keys are both regex match group names and corresponding output dict keys |\n",
       "| **Returns** | **dict[str]** |  | **parsed metadata in key/value format** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(AlnFileReader.parse_header_reference_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'2591237:ncbi:1': {'refseq_accession': 'MK211378',\n",
      "                    'refseq_length': '30213',\n",
      "                    'refseqid': '2591237:ncbi:1',\n",
      "                    'refseqnb': '1',\n",
      "                    'refsource': 'ncbi',\n",
      "                    'reftaxonomyid': '2591237',\n",
      "                    'species': 'Coronavirus BtRs-BetaCoV/YN2018D  scientific '\n",
      "                               'name'},\n",
      " '7321952:ncbi:2': {'refseq_accession': 'MK873112',\n",
      "                    'refseq_length': '',\n",
      "                    'refseqid': '7321952:ncbi:2',\n",
      "                    'refseqnb': '2',\n",
      "                    'refsource': 'ncbi',\n",
      "                    'reftaxonomyid': '7321952',\n",
      "                    'species': 'Not a real virus just here for code testing '\n",
      "                               'scientific name 31203'}}\n"
     ]
    }
   ],
   "source": [
    "pprint(it.parse_header_reference_sequences())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L246){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### AlnFileReader.set_header_parsing_rules\n",
       "\n",
       ">      AlnFileReader.set_header_parsing_rules (pattern:str|bool=None,\n",
       ">                                              keys:list[str]=None,\n",
       ">                                              verbose:bool=False)\n",
       "\n",
       "Set the regex parsing rule for reference sequence in ALN header.\n",
       "\n",
       "Updates 3 class attributes: `re_header_rule_name`, `re_header_pattern`, `re_header_keys`\n",
       "\n",
       "TODO: refactor this and the method in Core: to use a single function for the common part and a parameter for               the text_to_parse\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| pattern | str \\| bool | None | regex pattern to apply to parse the text, search in parsing rules json if None |\n",
       "| keys | list[str] | None | list of keys/group for regex, search in parsing rules json if None |\n",
       "| verbose | bool | False | when True, provides information on each rule |\n",
       "| **Returns** | **None** |  |  |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L246){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### AlnFileReader.set_header_parsing_rules\n",
       "\n",
       ">      AlnFileReader.set_header_parsing_rules (pattern:str|bool=None,\n",
       ">                                              keys:list[str]=None,\n",
       ">                                              verbose:bool=False)\n",
       "\n",
       "Set the regex parsing rule for reference sequence in ALN header.\n",
       "\n",
       "Updates 3 class attributes: `re_header_rule_name`, `re_header_pattern`, `re_header_keys`\n",
       "\n",
       "TODO: refactor this and the method in Core: to use a single function for the common part and a parameter for               the text_to_parse\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| pattern | str \\| bool | None | regex pattern to apply to parse the text, search in parsing rules json if None |\n",
       "| keys | list[str] | None | list of keys/group for regex, search in parsing rules json if None |\n",
       "| verbose | bool | False | when True, provides information on each rule |\n",
       "| **Returns** | **None** |  |  |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(AlnFileReader.set_header_parsing_rules)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------------------------------\n",
      "Rule <fasta_cov_ncbi> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <fasta_rhinolophus_ferrumequinum> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <fastq_art_illumina> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <aln_art_illumina> generated an error\n",
      "No match on this line\n",
      "--------------------------------------------------------------------------------\n",
      "Rule <aln_art_illumina-refseq> generated 7 matches\n",
      "--------------------------------------------------------------------------------\n",
      "^@SQ[\\t\\s]*(?P<refseqid>(?P<reftaxonomyid>\\d*):(?P<refsource>\\w*):(?P<refseqnb>\\d*))[\\t\\s]*\\[(?P<refseq_accession>[\\d\\w]*)\\][\\t\\s]*(?P=reftaxonomyid)[\\s\\t]*(?P=refsource)[\\s\\t]*(?P=refseqnb)[\\s\\t]*\\[(?P=refseq_accession)\\][\\s\\t]*(?P=reftaxonomyid)[\\s\\t]*(?P<species>\\w[\\w\\d\\/\\s\\-\\.]*)[\\s\\t](?P<refseq_length>\\d*)$\n",
      "['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'refseq_accession', 'species', 'refseq_length']\n",
      "--------------------------------------------------------------------------------\n",
      "Selected rule with most matches: aln_art_illumina-refseq\n"
     ]
    }
   ],
   "source": [
    "it.set_header_parsing_rules(verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aln_art_illumina-refseq\n",
      "^@SQ[\\t\\s]*(?P<refseqid>(?P<reftaxonomyid>\\d*):(?P<refsource>\\w*):(?P<refseqnb>\\d*))[\\t\\s]*\\[(?P<refseq_accession>[\\d\\w]*)\\][\\t\\s]*(?P=reftaxonomyid)[\\s\\t]*(?P=refsource)[\\s\\t]*(?P=refseqnb)[\\s\\t]*\\[(?P=refseq_accession)\\][\\s\\t]*(?P=reftaxonomyid)[\\s\\t]*(?P<species>\\w[\\w\\d\\/\\s\\-\\.]*)[\\s\\t](?P<refseq_length>\\d*)$\n",
      "['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'refseq_accession', 'species', 'refseq_length']\n"
     ]
    }
   ],
   "source": [
    "print(it.re_header_rule_name)\n",
    "print(it.re_header_pattern)\n",
    "print(it.re_header_keys)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def create_infer_ds_from_fastq(\n",
    "    p2fastq: str|Path,             # Path to the fastq file (aln file path is inferred)\n",
    "    output_dir:str|Path|None=None, # Path to directory where ds file will be saved\n",
    "    overwrite_ds:bool=False,       # If True, overwrite existing ds file. If False, error is raised if ds file exists\n",
    "    nsamples:int|None=None         # Used to limit the number of reads to use for inference, use all if None\n",
    ")-> (Path, np.ndarray):      # Path to the dataset file, Array with additional read information\n",
    "    \"\"\"Build a dataset file for inference only, from simreads fastq to text format ready for the CNN Virus model\n",
    "    \n",
    "    > Note: currently also return additional read information as an array. \n",
    "    >\n",
    "    > TODO: consider to save as a file\n",
    "    \"\"\"\n",
    "    fastq = FastqFileReader(p2fastq)\n",
    "    aln = AlnFileReader(p2fastq.parent / f\"{p2fastq.stem}.aln\")\n",
    "    \n",
    "    if output_dir is None:\n",
    "        p2dir = Path()\n",
    "    else:\n",
    "        validate_path(output_dir, path_type='dir', raise_error=True)\n",
    "        p2outdir = output_dir if isinstance(output_dir, Path) else Path(output_dir)\n",
    "    \n",
    "    p2dataset = p2outdir / f\"{p2fastq.stem}_ds\"\n",
    "    if p2dataset.is_file():\n",
    "        if overwrite_ds: \n",
    "            p2dataset.unlink()\n",
    "        else:\n",
    "            raise ValueError(f\"{p2dataset.name} already exists in {p2dataset.absolute()}\")\n",
    "    p2dataset.touch()\n",
    "    \n",
    "    read_ids = []\n",
    "    read_refseqs = []\n",
    "    read_start_pos = []\n",
    "    read_strand = []\n",
    "    \n",
    "    with open(p2dataset, 'a') as fp:\n",
    "        i = 1\n",
    "        for fastq_chunk, aln_chunk in zip(fastq.it, aln.it):\n",
    "            seq = fastq_chunk['sequence']\n",
    "            \n",
    "            aln_meta = parse_metadata_art_read_aln(aln_chunk['definition line'])\n",
    "#             print(aln_meta.keys())\n",
    "            read_ids.append(aln_meta['readid'])\n",
    "            read_refseqs.append(aln_meta['refseqid'])\n",
    "            read_start_pos.append(aln_meta['aln_start_pos'])\n",
    "            read_strand.append(aln_meta['refseq_strand'])\n",
    "\n",
    "            fp.write(f\"{seq}\\t{0}\\t{0}\\n\")\n",
    "#             print(f\"{seq}\\t{0}\\t{0}\\n\")\n",
    "\n",
    "            i += 1\n",
    "            if nsamples:\n",
    "                if i > nsamples: break\n",
    "                    \n",
    "    print(f\"Dataset with {i-1:,d} reads\")    \n",
    "    return p2dataset, np.array(list(zip(read_ids, read_refseqs, read_start_pos, read_strand)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L338){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### create_infer_ds_from_fastq\n",
       "\n",
       ">      create_infer_ds_from_fastq (p2fastq:str|pathlib.Path,\n",
       ">                                  output_dir:str|pathlib.Path|None=None,\n",
       ">                                  overwrite_ds:bool=False,\n",
       ">                                  nsamples:int|None=None)\n",
       "\n",
       "Build a dataset file for inference only, from simreads fastq to text format ready for the CNN Virus model\n",
       "\n",
       "> Note: currently also return additional read information as an array. \n",
       ">\n",
       "> TODO: consider to save as a file\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| p2fastq | str \\| Path |  | Path to the fastq file (aln file path is inferred) |\n",
       "| output_dir | str \\| Path \\| None | None | Path to directory where ds file will be saved |\n",
       "| overwrite_ds | bool | False | If True, overwrite existing ds file. If False, error is raised if ds file exists |\n",
       "| nsamples | int \\| None | None | Used to limit the number of reads to use for inference, use all if None |\n",
       "| **Returns** | **(Path, np.ndarray)** |  | **Path to the dataset file, Array with additional read information** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/cnn_virus/data.py#L338){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### create_infer_ds_from_fastq\n",
       "\n",
       ">      create_infer_ds_from_fastq (p2fastq:str|pathlib.Path,\n",
       ">                                  output_dir:str|pathlib.Path|None=None,\n",
       ">                                  overwrite_ds:bool=False,\n",
       ">                                  nsamples:int|None=None)\n",
       "\n",
       "Build a dataset file for inference only, from simreads fastq to text format ready for the CNN Virus model\n",
       "\n",
       "> Note: currently also return additional read information as an array. \n",
       ">\n",
       "> TODO: consider to save as a file\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| p2fastq | str \\| Path |  | Path to the fastq file (aln file path is inferred) |\n",
       "| output_dir | str \\| Path \\| None | None | Path to directory where ds file will be saved |\n",
       "| overwrite_ds | bool | False | If True, overwrite existing ds file. If False, error is raised if ds file exists |\n",
       "| nsamples | int \\| None | None | Used to limit the number of reads to use for inference, use all if None |\n",
       "| **Returns** | **(Path, np.ndarray)** |  | **Path to the dataset file, Array with additional read information** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_doc(create_infer_ds_from_fastq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessors for CNN model data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def strings_to_tensors(\n",
    "    b: tf.Tensor        # batch of strings \n",
    "    ):\n",
    "    \"\"\"Function converting a batch of bp strings into three tensors: (x_seqs, (y_labels, y_pos))\"\"\"\n",
    "    \n",
    "    # Split the string in three : returns a ragged tensor which needs to be converted into a normal tensor using .to_tensor()\n",
    "    t = tf.strings.split(b, '\\t').to_tensor(default_value = '', shape=[None, 3])\n",
    "\n",
    "    # Split each sequence string into a list of single base strings:\n",
    "    # 'TCAAAATAATCA' -> ['T','C','A','A','A','A','T','A','A','T','C','A']\n",
    "    seqs = tf.strings.bytes_split(t[:, 0]).to_tensor(shape=(None, 50))\n",
    "\n",
    "\n",
    "    # BHE sequences\n",
    "    # Each base letter (A, C, G, T, N) is replaced by a OHE vector\n",
    "    #     \"A\" converted into [1,0,0,0,0]\n",
    "    #     \"C\" converted into [0,1,0,0,0]\n",
    "    #     \"G\" converted into [0,0,1,0,0]\n",
    "    #     \"T\" converted into [0,0,0,1,0]\n",
    "    #     \"N\" converted into [0,0,0,0,1]\n",
    "    # \n",
    "    # Technical Notes:\n",
    "    # a. The batch of sequence `seqs` has a shape (batch_size, 50) after splitting each byte. \n",
    "    #    Must flatten it first, then apply the transform on each base, then reshape to original shape\n",
    "    # b. We need to map each letter to one vector/tensor. \n",
    "    #    1. Cast bytes seqs into integer sequence (uint8 to work byte by byte)\n",
    "    #    2. For each base letter (A, C, G, T, N) create one tensor (batch_size, 50) (seqs_A, _C, _G, _T, _N)\n",
    "    #    3. Value is 1 if it is the base in the sequence, otherwise 0\n",
    "    #    4. Concatenate these 5 tensors into a tensor of shape (batch_size, 50, 5)\n",
    " \n",
    "    seqs_uint8 = tf.io.decode_raw(seqs, out_type=tf.uint8)\n",
    "    # note: tf.io.decode_raw adds one dimension at the end in the process\n",
    "    #       [b'C', b'A', b'T'] will return [[67], [65], [84]] and not [67, 65, 84]\n",
    "    #       this is actually what we want to contatenate the values for each base letter\n",
    "\n",
    "    A, C, G, T, N = 65, 67, 71, 84, 78\n",
    "\n",
    "    seqs_A = tf.cast(seqs_uint8 == A, tf.float32)\n",
    "    seqs_C = tf.cast(seqs_uint8 == C, tf.float32)\n",
    "    seqs_G = tf.cast(seqs_uint8 == G, tf.float32)\n",
    "    seqs_T = tf.cast(seqs_uint8 == T, tf.float32)\n",
    "    seqs_N = tf.cast(seqs_uint8 == N , tf.float32)\n",
    "\n",
    "    x_seqs = tf.concat([seqs_A, seqs_C, seqs_G, seqs_T, seqs_N], axis=2)\n",
    "\n",
    "    # OHE labels\n",
    "    n_labels = 187\n",
    "    y_labels = tf.strings.to_number(t[:, 1], out_type=tf.int32)\n",
    "    y_labels = tf.gather(tf.eye(n_labels), y_labels)\n",
    "\n",
    "    # OHE positions\n",
    "    n_pos = 10\n",
    "    y_pos = tf.strings.to_number(t[:, 2], out_type=tf.int32)\n",
    "    y_pos= tf.gather(tf.eye(n_pos), y_pos)\n",
    "\n",
    "    return (x_seqs, (y_labels, y_pos))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "26b205197a934fdeabb71e65ac11acba": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "514ad0bfcabf4df580a9a872af814af9": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    },
    "55646397fc9349d3af9e98b1f2b26f5d": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "70f6be4247664b708b662c34e7abe3ee": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "7179c6cc207941648c348b1bf10cb87f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "LabelModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "LabelModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "LabelView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_70f6be4247664b708b662c34e7abe3ee",
      "placeholder": "​",
      "style": "IPY_MODEL_bdde467d943148ce9bb6355fd7582a5c",
      "value": "0.078 MB of 0.078 MB uploaded (0.020 MB deduped)\r"
     }
    },
    "7849f255e99b4853bdba7f8badf1054a": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "VBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "VBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "VBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_7179c6cc207941648c348b1bf10cb87f",
       "IPY_MODEL_e0819a1ddcc64c08a748a2fd88350f09"
      ],
      "layout": "IPY_MODEL_26b205197a934fdeabb71e65ac11acba"
     }
    },
    "bdde467d943148ce9bb6355fd7582a5c": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "e0819a1ddcc64c08a748a2fd88350f09": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_55646397fc9349d3af9e98b1f2b26f5d",
      "max": 1,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_514ad0bfcabf4df580a9a872af814af9",
      "value": 1
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
